
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
knitr::opts_knit$set(root.dir="/storage1/fs1/leyao.wang/Active/Users/bquinn")
getwd()
```

# Package Loading

```{r}
# if(!require("BiocManager", quietly=TRUE))
#   install.packages("BiocManager")
# BiocManager::install("DirichletMultinomial", version="3.14", force=TRUE)


setwd("/storage1/fs1/leyao.wang/Active/Users/bquinn")
.libPaths(c("/storage1/fs1/leyao.wang/Active/R_libraries/CampPontanezen", "/storage1/fs1/leyao.wang/Active/R_libraries/Camp Pontanezen.Camp Pontanezen", "/storage1/fs1/leyao.wang/Active/R_libraries/x86_64, linux-gnu.x86_64, linux-gnu", "/storage1/fs1/leyao.wang/Active/R_libraries/Camp Pontanezen.Camp Pontanezen/Bailey"))
package.list <- c("devtools", "dada2", "ape", "scales", "rlang", "gtools", "ggplot2", "dplyr", "ggpubr", "remotes", "plyr", 
                  "microbiome", "xtable", "kableExtra", "DescTools", "glue", "paletteer", "ggplot2", "magrittr",
                  "ggrepel", "grid", "rstatix", "data.table", "tidyverse", "ggpubr", "ggprism", "patchwork", 
                  "dplyr", "decontam", "Biostrings", "DESeq2", "phyloseqCompanion", "qiime2R", 
                  "phyloseq", "biomformat", "ggtext", "extrafont") #c("lemon", "RColorBrewer", "randomcoloR", "viridis", "mmtable2", "ShortRead", )
# install.packages(package.list, lib.loc="/storage1/fs1/leyao.wang/Active/R_libraries/Camp Pontanezen.Camp Pontanezen")
lapply(package.list, require, character.only=TRUE)
# lapply(uln(sapply(package.list, function(x) paste0("package:", x))), detach, character.only=TRUE, unload=TRUE, force=TRUE)

library(FEAST, lib.loc="/storage1/fs1/leyao.wang/Active/R_libraries/Camp Pontanezen.Camp Pontanezen/Bailey")

options(scipen=999)

Pheno.path <- "/storage1/fs1/leyao.wang/Active/Users/bquinn/Infant.Nasal.Microbiome/General.files/infant.nasal.csv"

load("/storage1/fs1/leyao.wang/Active/Users/bquinn/Functions.RData")
```


# My Function

```{r}
my.function <- function(path, group.id, group.title, processing=FALSE, grouped=FALSE, plots=NULL, pheno=FALSE){
  file.creation(path)
  if(processing==TRUE){
    if(grouped==FALSE){
      decontam.preprocessing(group.id, Initial.path, Decontam.path)
      # data.set.number <- as.numeric(readline(prompt="data.set.number: "))
      asv.prune <- identify.contaminants(group.id, Decontam.path, 1)
      decontam(group.id, asv.prune, 0.1, Initial.path, Decontam.path)
      # decontam(group.id, NULL, 0.1, Initial.path, Decontam.path)
      if(pheno==FALSE){
        phyloseq.processing(group.id, Initial.path, Decontam.path, Phyloseq.path, NULL, FALSE, FALSE)
      }else{
        phyloseq.processing(group.id, Initial.path, Decontam.path, Phyloseq.path, Pheno.path, FALSE, FALSE)
      }
    }else{
      if(group.title!="1-37"){
        phyloseq.processing(group.id, Initial.path, Decontam.path, Phyloseq.path, Pheno.path, TRUE, FALSE)
      }else{
        phyloseq.processing(group.id, Initial.path, Decontam.path, Phyloseq.path, Pheno.path, TRUE, TRUE)
      }
    }
  }

  taxa.level <- c("ASV")
  # taxa.level <- c("Genus")
  # taxa.level <- c("Species")

  for(j in 1:length(taxa.level)){
    cat("\n", taxa.level[j])
    assign(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j])), readRDS(paste0(Phyloseq.path, "/phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]), ".rds"))) #, envir=.GlobalEnv
    assign(paste0("phyloseq.pruned.control.", group.id, ".", tolower(taxa.level[j])), readRDS(paste0(Phyloseq.path, "/phyloseq.pruned.control.", group.id, ".", tolower(taxa.level[j]), ".rds")))
    assign(paste0("phyloseq.pruned.both.", group.id, ".", tolower(taxa.level[j])), readRDS(paste0(Phyloseq.path, "/phyloseq.pruned.both.", group.id, ".", tolower(taxa.level[j]), ".rds")))

    sample.types <- unname(unlist(unique(as.data.frame(sample_data(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j])))))[, "type"])))

    if(length(unname(unlist(unique(as.data.frame(sample_data(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j])))))[, "group"]))))==1){
      if(unname(unlist(unique(as.data.frame(sample_data(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j])))))[, "group"])))=="mother"){
        pheno.group.list <- c("day", "conditions", "medications", "antibiotics", "pets", "family", "race", "exposures")
      }else if(unname(unlist(unique(as.data.frame(sample_data(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j])))))[, "group"])))=="baby"){
        pheno.group.list <- c("time", "day", "birth_type", "sex", "conditions", "medications", "antibiotics", "pets", "family", "race", "exposures")
      }
    }else{
      pheno.group.list <- c("type", "birth_type", "sex", "antibiotics", "pets", "family", "race", "grouping")
    }

    # pheno.counts(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), group.id, pheno.group.list[c(1:6, 8)], paste0("phyloseq.pruned.", group.id), Phyloseq.path)
    # pheno.counts(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), group.id, pheno.group.list[1:2], paste0("phyloseq.pruned.", group.id), Phyloseq.path)
    
    if(taxa.level[j]=="Family"||taxa.level[j]=="Genus"||taxa.level[j]=="Species"){
      if(plots=="bar"){ 
        for(i in 1:length(pheno.group.list)){
          cat("\n", pheno.group.list[i])
          if(pheno.group.list[i]=="type"){
            phyloseq.barplot(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 20, pheno.group.list[i], "family", Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
          }else{
            phyloseq.barplot(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 20, pheno.group.list[i], NULL, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
            # plot.top.pie(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 5, pheno.group.list[i], NULL, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
          }
        }
      }
    }
    
    
    # if(taxa.level[j]=="Genus"||taxa.level[j]=="Species"){
    if(taxa.level[j]=="Species"){
      if(plots=="correlation"){
        # phyloseq.taxa(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 0, "type", sample.types[1], sample.types[2:length(sample.types)], Comparisons.path, 
        # paste0("phyloseq.pruned.", group.id))
        family.comparison.processing(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], sample.types[1], sample.types[2:length(sample.types)], 0, Correlation.path, paste0("phyloseq.pruned.", group.id))
      }
      if(plots=="bar"){
        phyloseq.barplot(get(paste0("phyloseq.pruned.control.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 20, "Sample", "group", 
                         Taxa.path, paste0("phyloseq.pruned.control.", group.id), group.title)
      }
    }

    if(taxa.level[j]=="Genus"||taxa.level[j]=="Species"||taxa.level[j]=="ASV"){
      # top.genus(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), c("Streptococcus", "Staphylococcus", "Corynebacterium", "Pseudomonas", "Veillonella"), Taxa.path, paste0("phyloseq.pruned.", group.id))
      if(plots=="diversity"){
        # pheno.group.list <- "birth_type"
        for(i in 1:length(pheno.group.list)){
          if(pheno.group.list[i]!="family"){
            # richness.alpha(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], pheno.group.list[i], path, paste0("phyloseq.pruned.", group.id), reads=TRUE)
            richness.alpha(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], pheno.group.list[i], path, paste0("phyloseq.pruned.", group.id), reads=FALSE)
            plotBeta(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], pheno.group.list[i], path, paste0("phyloseq.pruned.", group.id), TRUE, TRUE, grouped.title)
          }
        }
      }

      # if(plots=="bar"){
      #   phyloseq.barplot(get(paste0("phyloseq.pruned.control.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], 20, "Sample", "group", 
      #                    Taxa.path, paste0("phyloseq.pruned.control.", group.id), group.title)
      # }

      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "epidermidis", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "mitis", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "salivarius", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "parainfluenzae", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "johnsonii", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "aureus", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Bergeyella Unclassified Species asv8", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "tuberculostearicum", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "parvum", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "acnes", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "aeruginosa", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)

    }else if(taxa.level[j]=="Genus"){
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Staphylococcus", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Pseudomonas", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Corynebacterium", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Bergeyella", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
      # species.correlation(get(paste0("phyloseq.pruned.", group.id, ".", tolower(taxa.level[j]))), taxa.level[j], n=0, "Streptococcus", "grouping", "BNA.1-50", "MNA.1-50", FALSE, Taxa.path, paste0("phyloseq.pruned.", group.id), group.title)
    }
  }
}
```

# File Creation

```{r File Creation}
file.creation <- function(path){
  dir.create(path)
  file.list <- c("Sequence", "Output")
  for(i in 1:length(file.list)){
    dir.create(paste0(path, "/", file.list[i]))
    assign(paste0(file.list[i], ".path"), paste0(path, "/", file.list[i]), envir=parent.frame())
  }
  
  output.file.list <- c("Correlation", "Diversity", "Taxa")
  for(i in 1:length(output.file.list)){
    dir.create(paste0(path, "/Output/", output.file.list[i]))
    assign(paste0(output.file.list[i], ".path"), paste0(path, "/Output/", output.file.list[i]), envir=parent.frame())
  }
  
  taxa.file.list <- c("Family", "Genus", "Species")
  for(i in 1:length(taxa.file.list)){
    dir.create(paste0(path, "/Output/Taxa/", taxa.file.list[i]))
    assign(paste0("Taxa.", taxa.file.list[i], ".path"), paste0(path, "/Output/Taxa/", taxa.file.list[i]), envir=parent.frame())
  }
  
  dir.create(paste0(path, "/Output/Correlation/Comparisons"))
  assign(paste0("Comparisons.path"), paste0(path, "/Output/Correlation/Comparisons"), envir=parent.frame())
  
  processing.file.list <- c("Initial", "Decontam", "Phyloseq")
  dir.create(paste0(path, "/Output/Processing"))
  for(i in 1:length(processing.file.list)){
    dir.create(paste0(path, "/Output/Processing/", processing.file.list[i]))
    assign(paste0(processing.file.list[i], ".path"), paste0(path, "/Output/Processing/", processing.file.list[i]), envir=parent.frame())
  }
}
```

# Processing

## Initial Dada Processing {#initial-dada-processing}

```{r Initial Dada Processing, warning=TRUE}
initial.dada.processing <- function(sample.id.local, group.id, min.reads, path, plots=TRUE, multithread=TRUE){
  sequence.path <- paste0(path, "/Sequence")
  output.path <- paste0(path, "/Output/Processing/Initial")
  cat("\n", paste("Initial dada processing for", group.id))
  ### File and sample names ###
        cat("\n  ", paste("File and sample names"))
  # fnFs <- sort(list.files(sequence.path, pattern="_R1_001.fastq", full.names=TRUE))
  # fnRs <- sort(list.files(sequence.path, pattern="_R2_001.fastq", full.names=TRUE))
  fnFs <- sort(list.files(sequence.path, pattern="_R1_001.fq.gz", full.names=TRUE))
  fnRs <- sort(list.files(sequence.path, pattern="_R2_001.fq.gz", full.names=TRUE))

  # get.sample.name <- function(fname) strsplit(basename(fname), "_")[[1]][sample.id.local]
  
  get.sample.name <- function(fname) paste(strsplit(basename(fname), "_")[[1]][sample.id.local], strsplit(basename(fname), "_")[[1]][3], sep="_")
  sample.names <- unname(sapply(fnFs, get.sample.name))
  sample.names.r <- unname(sapply(fnRs, get.sample.name))
        cat("\n     ", paste("Sample names (head)"))
        cat("\n      ", paste(head(sample.names)))
        # cat("\n      ", paste((sample.names)))
        
        # print(sample.names)
        # print(sample.names.r)


  ### Plot quality profile ###
        cat("\n  ", paste("File and sample names"))
        cat("\n     ", paste("Number of plots"))
        cat("\n      ", paste(length(fnFs)))
        cat("\n     ", paste("Generating plots"))
  quality.plots.per.image <- function(plot.number, rows, columns, fnFs, fnRs){
    direction <- list(fnFs, fnRs)
    for(a in 1:length(direction)){
      cycles=ceiling((length(direction[[a]]))/plot.number)
      for (i in 1:cycles){
        k=(i-1)*plot.number
        if(a==1){
          filename <- paste0(output.path, "/quality.plot.", group.id, ".fnFs.")
        }else{
          filename <- paste0(output.path, "/quality.plot.", group.id, ".fnRs.")
        }
        if((plot.number+k)<length(direction[[a]])){
                cat("\n      ", paste0(1+k, "-", plot.number+k))
          l <- (1+k):(plot.number+k)
          filename <- paste0(filename, 1+k, "-", plot.number+k, ".png")
        }else{
                cat("\n      ", paste0(1+k, "-", length(direction[[a]])))
          l <- (1+k):length(direction[[a]])
          filename <- paste0(filename, 1+k, "-", length(direction[[a]]), ".png")
        }
        if(file.exists(filename)==FALSE){
            plots <- lapply(l, function(x){
            plotQualityProfile(direction[[a]][x]) +
              theme(strip.background=element_blank(), strip.text.x=element_blank(), 
                    plot.title=element_text(size=15, hjust=0.5), text=element_text(size=15)) +
              labs(title=sample.names[x])
          })
          quality.grouped <- ggarrange(plotlist=plots, ncol=columns, nrow=rows)
          ggsave(filename=filename, plot=quality.grouped, height=(rows*3), width=((columns-1))*3)
        }
      }
    }
  }
  if(plots==TRUE){
    quality.plots.per.image(32, 4, 8, fnFs, fnRs)
  }
  


  ### Filter and trim ###
        cat("\n  ", paste("Filter and trim"))
        # cat("\n     ")
  filtFs <- file.path(sequence.path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
  duped <- which(duplicated(filtFs))
  filtRs <- file.path(sequence.path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
  duped <- which(duplicated(filtRs))
  if(length(duped)!=0){
    filtFs <- filtFs[-c(duped)]
    fnFs <- fnFs[-c(duped)]
    filtRs <- filtRs[-c(duped)]
    fnRs <- fnRs[-c(duped)]
  
    sample.names <- sample.names[-c(duped)]
          cat("\n    ", paste("Duplicated"))
          cat("\n      ", paste(duped))
  }
  
  # print(fnFs)
  # print(filtFs)
  # print(fnRs)
  # print(filtRs)
  
  if(file.exists(paste0(output.path, "/out.", group.id, ".rds"))){
          cat("\n  ", paste("Read Filtered"))
    out <- readRDS(paste0(output.path, "/out.", group.id, ".rds"))
  }else{
          cat("\n  ", paste("Assign Filtered"))
    
    trimleft1 <- readline(prompt="TrimLeft1 (19): ")
    trimleft2 <- readline(prompt="TrimLeft2 (16): ")
    truncfor <- readline(prompt="Trunc Forward: ")
    truncrev <- readline(prompt="Trunc Reverse: ")
    
    out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, trimLeft=c(as.numeric(trimleft1), as.numeric(trimleft2)), truncLen=c(as.numeric(truncfor), as.numeric(truncrev)),
                         maxN=0, maxEE=c(2, 2), truncQ=2, rm.phix=TRUE,
                         compress=TRUE, multithread=multithread)
    
    # out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, trimLeft=c(19, 16), truncLen=c(240, 160),
    #                      maxN=0, maxEE=c(2, 2), truncQ=2, rm.phix=TRUE,
    # compress=TRUE, multithread=FALSE)
    saveRDS(out, paste0(output.path, "/out.", group.id, ".rds")) 
  }
        cat("\n    ", paste("Reads in and out"))
        cat("\n      ", paste(head(out)))


  ### Error Rates ###
        cat("\n  ", paste("Error Rates"))
        cat("\n     ")
        print(filtFs)
  if(file.exists(paste0(output.path, "/errF.rds"))==FALSE){
    errF <- learnErrors(filtFs, multithread=TRUE)
    saveRDS(errF, paste0(output.path, "/errF.rds"))
  }else{
    errF <- readRDS(paste0(output.path, "/errF.rds"))
  }
        cat("\n    ")
  if(file.exists(paste0(output.path, "/errR.rds"))==FALSE){
    errR <- learnErrors(filtRs, multithread=TRUE)
    saveRDS(errR, paste0(output.path, "/errR.rds"))
  }else{
    errR <- readRDS(paste0(output.path, "/errR.rds"))
  }
  if(plots==TRUE){
    if(file.exists(paste0(output.path, "/error.plot.png"))==FALSE){
      error.plot <- plotErrors(errF, nominalQ=TRUE)
            cat("\n    ")
      filename <- paste0(output.path, "/error.plot.png")
      ggsave(filename, error.plot, height=7, width=7)
    }
  }
  
  
  ### Sample Inference ###
        cat("\n  ", paste("Sample Inference"))
        cat("\n    ")
  if(file.exists(paste0(output.path, "/derepFs.rds"))==FALSE){
    derepFs <- derepFastq(filtFs, verbose=TRUE)
    saveRDS(derepFs, paste0(output.path, "/derepFs.rds"))
  }else{
    derepFs <- readRDS(paste0(output.path, "/derepFs.rds"))
  }
        
  if(file.exists(paste0(output.path, "/derepRs.rds"))==FALSE){
    derepRs <- derepFastq(filtRs, verbose=TRUE)
    saveRDS(derepRs, paste0(output.path, "/derepRs.rds"))
  }else{
    derepRs <- readRDS(paste0(output.path, "/derepRs.rds"))
  }
  
  # Name the derep-class objects by the sample names
  names(derepFs) <- sample.names
  names(derepRs) <- sample.names

  if(file.exists(paste0(output.path, "/dadaFs.rds"))==FALSE){
    dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
    saveRDS(dadaFs, paste0(output.path, "/dadaFs.rds"))
  }else{
    dadaFs <- readRDS(paste0(output.path, "/dadaFs.rds"))
  }
  
  if(file.exists(paste0(output.path, "/dadaRs.rds"))==FALSE){
    dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
    saveRDS(dadaRs, paste0(output.path, "/dadaRs.rds"))
  }else{
    dadaRs <- readRDS(paste0(output.path, "/dadaRs.rds"))
  }
        cat("\n    ", paste("DADA2 denoising results"))
        # cat("\n      ", paste(dadaFs[[1]]))


  ### Merged paired reads ###
        cat("\n  ", paste("Merged paired reads"))
        cat("\n    ")
  if(file.exists(paste0(output.path, "/mergers.rds"))==FALSE){
    mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE, justConcatenate=TRUE)
    saveRDS(mergers, paste0(output.path, "/mergers.rds"))
  }else{
    mergers <- readRDS(paste0(output.path, "/mergers.rds"))
  }
  
  # mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE, justConcatenate=TRUE)
        cat("\n    ", paste("Mergers object (head)"))
        cat("\n      ", paste(head(mergers[[1]])))
        # View(mergers)


  ### Sequence table ###
        cat("\n  ", paste("Sequence table"))
  seqtab <- makeSequenceTable(mergers)
        cat("\n    ", paste("Dimentions"))
        cat("\n      ", paste(dim(seqtab)))
        # cat("\n     ", paste("Sequences"))
        # cat("\n      ", paste(table(nchar(getSequences(seqtab)))))


  ### Remove chimeras ###
        cat("\n  ", paste("Remove chimeras"))
        cat("\n    ")
  seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
  dim(seqtab.nochim)
        cat("\n    ", paste("Dimentions"))
        cat("\n      ", paste(dim(seqtab.nochim)))
  sum(seqtab.nochim)/sum(seqtab)
        cat("\n    ", paste("Percentag of reads kept"))
        cat("\n      ", paste0(round((sum(seqtab.nochim)/sum(seqtab))*100, 1), "%"))


  ### Track reads ###
        cat("\n  ", paste("Track reads"))
  getN <- function(x) sum(getUniques(x))
  track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
  # If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
  colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
  # track[, c("Sample")] <- sample.names[1:44]
  
  # for(i in 1:length(rownames(track))){
  #   rownames(track)[i] <- strsplit(rownames(track)[i], "_")[[1]][sample.id.local]
  # }
  
  for(i in 1:length(rownames(track))){
    rownames(track)[i] <- paste(strsplit(rownames(track)[i], "_")[[1]][sample.id.local], strsplit(rownames(track)[i], "_")[[1]][3], sep="_")
  }
  
  track <- rownames_to_column(as.data.frame(track))
  # head(track)
  write_csv(as.data.frame(track), file=paste0(output.path, "/track.", group.id, ".csv"))
  
  
        cat("\n    ", paste("Samples with reads under ", min.reads))
        cat("\n      ", paste(track[which(track[, 7]<=min.reads), 1]))
  assign(paste0("over.min.reads"), track[which(track[, 7]> min.reads), 1])
  saveRDS(over.min.reads, paste0(output.path, "/over.min.reads.", group.id, ".rds"))
  
  
  
  ## if there are more than x samples for loop this and break it up into chunks of ... then save to assign(paste(taxa, a))
  ## when for loop is completed (or during for loop) combine the taxa objects to one then save
  
  ### Assign Taxonomy ###
  if(file.exists(paste0(output.path, "/taxa.", group.id, ".rds"))){
          cat("\n  ", paste("Read Taxonomy"))
    taxa <- readRDS(paste0(output.path, "/taxa.", group.id, ".rds"))
  }else{
          cat("\n  ", paste("Assign Taxonomy"))
    taxa <- assignTaxonomy(seqtab.nochim, 
                           "/storage1/fs1/leyao.wang/Active/Users/bquinn/Infant.Nasal.Microbiome/General.files/silva_nr99_v138.1_wSpecies_train_set.fa.gz", 
                           multithread=TRUE)
    saveRDS(taxa, paste0(output.path, "/taxa.", group.id, ".rds"))
  }

  
  ### Phyloseq creation ###
        cat("\n  ", paste("Phyloseq creation"))
  # samdf=data.frame(samID=rownames(seqtab.nochim), 
  #                    family=gsub('[a-z]*', "", gsub('[A-Z]*', "", rownames(seqtab.nochim))), 
  #                    type=substring(gsub('[0-9]*', "", rownames(seqtab.nochim)), 2), 
  #                    region=substring(gsub('[0-9]*', "", rownames(seqtab.nochim)), 1, 1), 
  #                    reads=as.numeric(rowSums(seqtab.nochim)))
  
  samdf.names <- function(fname) strsplit(fname, "_")[[1]][1]
  samdf.form <- function(fname) strsplit(fname, "_")[[1]][2]
  samdf=data.frame(samID=rownames(seqtab.nochim), 
                   family=as.numeric(gsub('[a-z]*', "", gsub('[A-Z]*', "", sapply(row.names(seqtab.nochim), samdf.names)))), 
                   type=substring(gsub('[0-9]*', "", sapply(row.names(seqtab.nochim), samdf.names)), 2), 
                   region=substring(gsub('[0-9]*', "", sapply(row.names(seqtab.nochim), samdf.names)), 1, 1), 
                   form=sapply(row.names(seqtab.nochim), samdf.form), 
                   reads=as.numeric(rowSums(seqtab.nochim)))
  # print(samdf)
  rownames(samdf) <- samdf$samID
  
  samdf <- form.reassign(samdf)

  if("HO" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="HO"), "type"] <- "H2O"
    samdf[which(samdf[, "type"]=="H2O"), "family"] <- 0
  }
  if("pos" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="pos"), "type"] <- "POS"
  }
  if("Pos" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="Pos"), "type"] <- "POS"
  }
  if("POS" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="POS"), "family"] <- 0
  }
  
  if("cnt" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="cnt"), "type"] <- "CNT"
  }
  if("Cnt" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="Cnt"), "type"] <- "CNT"
  }
  if("CNT" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="CNT"), "family"] <- 0
  }
  
  if("neg" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="neg"), "type"] <- "NEG"
  }
  if("Neg" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="Neg"), "type"] <- "NEG"
  }
  if("NEG" %in% samdf[, "type"]==TRUE){
    samdf[which(samdf[, "type"]=="NEG"), "family"] <- 0
  }
  
  
  phyloseq <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
                       tax_table(taxa), 
                       sample_data(samdf))
  
  
  
  saveRDS(phyloseq, paste0(output.path, "/phyloseq.rds"))
  
  save(list=ls(all.names=TRUE), file=file.path(paste0(output.path, "/envi.", group.id, ".RData")))
}


merge.phyloseq <- function(group.id.list, path.list, group.id, path){
  output.path <- paste0(path, "/Output/Processing/Initial")
  phyloseq <- list()
  over.min.reads <- list()
  for(a in 1:length(path.list)){
    path.list[a] <- paste0(path.list[a], "/Output/Processing/Initial")
    phyloseq[[a]] <- readRDS(paste0(path.list[a], "/phyloseq.rds"))
    over.min.reads[[a]] <- readRDS(paste0(path.list[a], "/over.min.reads.", group.id.list[a], ".rds"))
    if(a!=1){
      phyloseq[[1]] <- merge_phyloseq(phyloseq[[1]], phyloseq[[a]])
      over.min.reads[[1]] <- c(over.min.reads[[1]], over.min.reads[[a]])
    }
  }
  saveRDS(phyloseq[[1]], paste0(output.path, "/phyloseq.rds"))
  saveRDS(over.min.reads[[1]], paste0(output.path, "/over.min.reads.", group.id, ".rds"))
}

phyloseq.creation <- function(group.id, path){
  output.path <- paste0(path, "/Output/Processing/Initial")
  phyloseq <- readRDS(paste0(output.path, "/phyloseq.rds"))


  dna <- Biostrings::DNAStringSet(taxa_names(phyloseq))
  names(dna) <- taxa_names(phyloseq)
  phyloseq <- merge_phyloseq(phyloseq, dna)
  taxa_names(phyloseq) <- paste0("asv", seq(ntaxa(phyloseq)))
  
  
  phyloseq <- short.taxa.name(phyloseq, "Genus", "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium")
  phyloseq <- short.taxa.name(phyloseq, "Genus", "Methylobacterium-Methylorubrum")

  phyloseq <- na.to.unclassified(phyloseq)
  
  
  ### Save ###
        cat("\n  ", paste("Save"))
  saveRDS(phyloseq, paste0(output.path, "/phyloseq.", group.id, ".rds"))
  # save(list=ls(all.names=TRUE), file=file.path(paste0(output.path, "/envi.", group.id, ".RData")))
} 
```

## Decontam Processing

```{r Decontam Preprocessing}
### Function for decontamination ###
decontam.preprocessing <- function(group.id, phyloseq.path, output.path){
  defaultW <- getOption("warn")
  options(warn=-1)
  
  # if(file.exists(paste0(phyloseq.path, "/track.csv"))==TRUE){
  #   track <- read_csv(paste0(phyloseq.path, "/track.csv"))
  # }else if(file.exists(paste0(phyloseq.path, "/track.", group.id, ".csv"))==TRUE){
  #   track <- read_csv(paste0(phyloseq.path, "/track.", group.id, ".csv"))
  # }
  
  # track <- read_csv(paste0(phyloseq.path, "/track.", group.id, ".csv"))
  phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
  # sample.types <- unname(unlist(unique(as.data.frame(sample_data(phyloseq))[, "region"])))
  # 
  # phyloseq.list <- list()
  # phyloseq.reset <- phyloseq
  # for(a in 1:length(sample.types)){
  #       cat("\n\n", paste("Decontam for", group.id, "region", sample.types[a]))
  #   phyloseq <- phyloseq.reset
  #   if(length(sample.types > 1)){
  #     sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("region")]==c(sample.types[a]))), ]
  #   }

    ### Assign to sample or control ###
          cat("\n  ", paste("Assign to sample or control"))
    for (i in 1:nrow(sample_data(phyloseq))) {
      x <- sample_data(phyloseq)[i, c("type")]
      if(x=="MSA"||x=="MNA"||x=="MBM"||x=="MAS"||x=="BST"||x=="BSA"||x=="BNA"){
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "True Sample"
      }else if(x=="POS"||x=="Pos"||x=="pos"){
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "Positive Control Sample"
      }else if(x=="CNT"||x=="Cnt"||x=="cnt"){
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "Control Sample"
      }else if(x=="NEG"||x=="Neg"||x=="neg"){
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "Control Sample"
      }else if(x=="H2O"||x=="HO"){
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "Control Sample"
      }else{
        sample_data(phyloseq)[i, c("Sample_or_Control")] <- "True Sample"
      }
    }
          
    # phyloseq <- rarefy_even_depth(phyloseq, rngseed=98, sample.size=3000)
    phyloseq.prune <- phyloseq
    saveRDS(phyloseq, paste0(output.path, "/phyloseq.prune.", group.id, ".rds"))
    # sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, "Sample_or_Control"]=="True Sample"), 
    #                                                  which(sample_data(phyloseq)[, "Sample_or_Control"]=="Control Sample")), ]
    saveRDS(phyloseq, paste0(output.path, "/phyloseq.", group.id, ".pre.processing.rds"))
    
    phyloseq.reset <- phyloseq
    threshold.list <- c(0.1, 0.5, 0.55, 0.75)
    for(a in 1:length(threshold.list)){
      phyloseq <- phyloseq.reset
      sample_data(phyloseq)$is.neg <- sample_data(phyloseq)$Sample_or_Control=="Control Sample"
      contamdf.prev <- isContaminant(phyloseq, method="prevalence", neg="is.neg", threshold=threshold.list[a])
      phyloseq <- prune_taxa(!contamdf.prev$contaminant, phyloseq.prune)
      col.name <- paste("Decontam", threshold.list[a])
      # if(exists("track")){
      #   track[, col.name] <- sample_sums(phyloseq)
      # }
    }
    # if(exists("track")){
    #   write_csv(as.data.frame(track), file=paste0(output.path, "/track.", group.id, ".csv"))
    # }
    options(warn=defaultW)
}
```

## Decontam

```{r Decontam}
decontam <- function(group.id, asv.prune=NULL, threshold, phyloseq.path, output.path){
  defaultW <- getOption("warn")
  options(warn=-1)    
  phyloseq <- readRDS(paste0(output.path, "/phyloseq.", group.id, ".pre.processing.rds"))
  
  phyloseq.prune <- readRDS(paste0(output.path, "/phyloseq.prune.", group.id, ".rds"))
  
    ### Plot the library size of the samples and controls ###
          cat("\n  ", paste("Plot the library size of the samples and controls"))
    df <- as.data.frame(sample_data(phyloseq))
    df$LibrarySize <- sample_sums(phyloseq)
    df <- df[order(df$LibrarySize), ]
    df$Index <- seq(nrow(df))
    library.size <- ggplot(data=df, aes(x=Index, y=LibrarySize, color=Sample_or_Control, label=samID)) +
      geom_point() +
      geom_text(aes(label=ifelse(Sample_or_Control=="Control Sample", as.character(samID), '')))
                #, hjust=-0.5, vjust=-2)
    # if(length(sample.types > 0)){
    #   filename <- paste0(output.path, "/", group.id, ".", sample.types[a], ".library.size.plot.png")
    #   ggsave(filename, library.size, height=7, width=7)
    # }else{
      filename <- paste0(output.path, "/", group.id, ".library.size.plot.png")
      ggsave(filename, library.size, height=7, width=7)
    # }
     
      # print(sum(sample_sums(subset.object(phyloseq.prune, "True Sample"))))
      # print(prune_taxa(taxa_sums(subset.object(phyloseq.prune, "True Sample"))!=0, subset.object(phyloseq.prune, "True Sample")))
      
      
    ### Identifying contaminants ###
          cat("\n  ", paste("Identifying contaminants"))
          cat("")
    sample_data(phyloseq)$is.neg <- sample_data(phyloseq)$Sample_or_Control=="Control Sample"
    contamdf.prev <- isContaminant(phyloseq, method="prevalence", neg="is.neg", threshold=threshold)
    contams <- contamdf.prev[which(contamdf.prev[, 6]), ]
    
    if(length(rownames(contams))!=0){
      contams.ps <- prune_taxa(rownames(contams), phyloseq)
      contams.df <- merge(as.data.frame(tax_table(contams.ps)), contams, by='row.names')
      write_csv(as.data.frame(contams.df), file=file.path(output.path, paste0("/", group.id, ".contaminants.csv")))
    }
    
    

    ### Prevalence plot ###
          cat("\n  ", paste("Prevalence plot"))
    phyloseq.pa <- transform_sample_counts(phyloseq, function(abund) 1*(abund>0))
    phyloseq.pa.neg <- prune_samples(sample_data(phyloseq.pa)$Sample_or_Control=="Control Sample", phyloseq.pa)
    phyloseq.pa.pos <- prune_samples(sample_data(phyloseq.pa)$Sample_or_Control=="True Sample", phyloseq.pa)
    # Make data.frame of prevalence in positive and negative samples
    df.pa <- data.frame(pa.pos=taxa_sums(phyloseq.pa.pos), pa.neg=taxa_sums(phyloseq.pa.neg), 
                          contaminant=contamdf.prev$contaminant)
    prevalence.plot <- ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
      xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")
    # if(length(sample.types > 0)){
    #   filename <- paste0(output.path, "/", group.id, ".", sample.types[a], ".prevalence.plot.png")
    #   ggsave(filename, prevalence.plot, height=7, width=7)
    # }else{
      filename <- paste0(output.path, "/", group.id, ".prevalence.plot.png")
      ggsave(filename, prevalence.plot, height=7, width=7)
    # }
      
    phyloseq.control <- phyloseq
    ### Prune the dataframe to exclude contaminants ###
        cat("\n  ", paste(""))
    phyloseq <- prune_taxa(!contamdf.prev$contaminant, phyloseq.prune)
    # phyloseq <- phyloseq.prune
    
    
# print(sum(sample_sums(subset.object(phyloseq, "True Sample"))))
# print(prune_taxa(taxa_sums(subset.object(phyloseq, "True Sample"))!=0, subset.object(phyloseq, "True Sample")))
  #   if(length(sample.types > 1)){
  #     phyloseq.list[[a]] <- phyloseq
  #     if(a > 1){
  #       phyloseq.list[[1]] <- merge_phyloseq(phyloseq.list[[1]], phyloseq.list[[a]])
  #     }
  #   }
  # }
  # 
  # if(length(sample.types > 1)){
  #   phyloseq <- phyloseq.list[[1]]
  # }
     
    
    # print(sum(sample_sums(subset.object(phyloseq, "True Sample"))))
    # print(prune_taxa(taxa_sums(subset.object(phyloseq, "True Sample"))!=0, subset.object(phyloseq, "True Sample")))
    ### Assign samples mother or baby ###
        cat("\n  ", paste("Assign samples mother or baby"))
  for (i in 1:nrow(sample_data(phyloseq.control))) {
    x <- sample_data(phyloseq.control)[i, c("type")]
    if(x=="MSA"||x=="MNA"||x=="MBM"||x=="MAS"){
      sample_data(phyloseq.control)[i, c("group")] <- "mother"
    } else if(x=="BST"||x=="BSA"||x=="BNA"){
      sample_data(phyloseq.control)[i, c("group")] <- "baby"
    }else if(x=="POS"||x=="Pos"||x=="pos"){
      sample_data(phyloseq.control)[i, c("group")] <- "positive control"
    # }else if(x=="H2O"){
      # sample_data(phyloseq)[i, c("group")] <- "water"
    }else{
      sample_data(phyloseq.control)[i, c("group")] <- "negative control"
    }
  }
  for (i in 1:nrow(sample_data(phyloseq))) {
    x <- sample_data(phyloseq)[i, c("type")]
    if(x=="MSA"||x=="MNA"||x=="MBM"||x=="MAS"){
      sample_data(phyloseq)[i, c("group")] <- "mother"
    } else if(x=="BST"||x=="BSA"||x=="BNA"){
      sample_data(phyloseq)[i, c("group")] <- "baby"
    }else if(x=="POS"||x=="Pos"||x=="pos"){
      sample_data(phyloseq)[i, c("group")] <- "positive control"
    # }else if(x=="H2O"){
      # sample_data(phyloseq)[i, c("group")] <- "water"
    }else{
      sample_data(phyloseq)[i, c("group")] <- "negative control"
    }
  }
        
  if(!is.null(asv.prune)){
      for(a in 1:length(asv.prune)){
        phyloseq <- remove.specific.taxa(phyloseq, asv.prune[a])
      }
    }
  # phyloseq.control <- phyloseq
  
  saveRDS(phyloseq.control, paste0(output.path, "/phyloseq.control.", group.id, ".rds"))
  saveRDS(phyloseq, paste0(output.path, "/phyloseq.", group.id, ".rds"))
  save(list=ls(all.names=TRUE), file=file.path(paste0(output.path, "/envi.", group.id, ".RData")))
  options(warn=defaultW)
}
```

## Sequencing Contaminants

```{r Contaminants}
identify.contaminants <- function(group.id, phyloseq.path, data.set.number){
  phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".pre.processing.rds"))
  phyloseq <- prune_samples(row.names(as.data.frame(sample_data(phyloseq)))[which(as.data.frame(sample_data(phyloseq))
                                                                                  [, "Sample_or_Control"]=="Control Sample")], phyloseq)
  # phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  # if(nrow(data.frame(otu_table(phyloseq)))==10){
  #   phyloseq <- prune_taxa(colSums(data.frame(otu_table(phyloseq))!=0)>=(nrow(data.frame(otu_table(phyloseq)))/2)-1, phyloseq)
  # }else{}
  
  # phyloseq <- prune_taxa(colSums(data.frame(otu_table(phyloseq))!=0)>=(nrow(data.frame(otu_table(phyloseq)))/data.set.number)-1, phyloseq)
  # phyloseq <- prune_taxa(colSums(data.frame(otu_table(phyloseq))!=0)>=((nrow(data.frame(otu_table(phyloseq)))/2)-1)/1, phyloseq)
  # print((nrow(data.frame(otu_table(phyloseq)))/2)-1)
  
  phyloseq <- prune_taxa(colSums(data.frame(otu_table(phyloseq))!=0)>=((nrow(data.frame(otu_table(phyloseq)))/2)-1)/data.set.number, phyloseq)
  phyloseq <- prune_taxa(taxa_sums(phyloseq)>25, phyloseq)
  contam.names <- names(sort(taxa_sums(phyloseq)[taxa_sums(phyloseq)>0], decreasing=TRUE))
  ps.table <- cbind(as.data.frame(tax_table(phyloseq)), t(as.data.frame(otu_table(phyloseq))))
  if(ncol(ps.table)==8){
    ps.table[, "Sum"] <- ps.table[, c(8)]
  }else{
    ps.table[, "Sum"] <- rowSums(ps.table[, c(8:(ncol(ps.table)))])
  }
  
  ps.table[, "ASV"] <- rownames(ps.table)
  ps.table <- ps.table[, c(ncol(ps.table), 1:(ncol(ps.table)-1))]
  # print(ps.table)
  # View(ps.table)
  # ps.table <- ps.table[, -c(1:5)]
  # ps.table %>%
  #   kbl(centering=TRUE, escape=FALSE, align="c") %>%
  #       kable_styling(font_size=15) %>%
  #       kable_classic("striped", full_width=F, html_font="Calibri", position="center") %>%
  #       row_spec(0, bold=TRUE) %>%
  #       save_kable(file=paste(paste0(phyloseq.path, "/table"), group.id, "contaminants", "png", sep="."), zoom=5)
  write_csv(ps.table, paste(paste0(phyloseq.path, "/table"), group.id, "contaminants", "csv", sep="."))
  # print(contam.names)
  print(contam.names)
  return(contam.names)
}

```

```{r}
top.genus <- function(phyloseq, taxa.list, output.path, phyloseq.name){
  phyloseq.reset <- phyloseq
  phyloseq.list <- list()
  for(a in 1:length(taxa.list)){
    phyloseq <- phyloseq.reset
    phyloseq <- trans.sam.counts(phyloseq, "type", TRUE)
    phyloseq <- prune_taxa(rownames(tax_table(phyloseq)[c(which(tax_table(phyloseq)[, "Genus"]==taxa.list[a])), ]), phyloseq)
    phyloseq <- prune_taxa(top.taxa.n(phyloseq, "Species", 3), phyloseq)
    # print(cbind(as.data.frame(tax_table(phyloseq)), t(as.data.frame(otu_table(phyloseq)))))
    phyloseq.list[[a]] <- phyloseq
    if(a > 1){
      phyloseq.list[[1]] <- merge_phyloseq(phyloseq.list[[1]], phyloseq.list[[a]])
    }
  }
  phyloseq <- phyloseq.list[[1]]
  ps.table <- cbind(as.data.frame(tax_table(phyloseq)), t(as.data.frame(otu_table(phyloseq))))
  rownames(ps.table) <- NULL
  colnames(ps.table)[8] <- "Relative Abundance"
  for(b in 1:nrow(ps.table)){
    ps.table[b, 8] <- paste0(round(as.numeric(ps.table[b, 8]), 3))
  }
  
  FileName <- paste0(output.path, "/", phyloseq.name, ".top.5.genus", ".png")
  # kbl.combined <- ps.table[, c(7, 8)] %>%
  #   kbl(centering=TRUE, escape=FALSE, col.names=c("Species", "Relative Abundance"), align="c") %>%
  #   kable_styling(font_size=15) %>%
  #   kable_classic("striped", full_width=F, html_font="Calibri", position="center") %>%
  #   row_spec(0, bold=TRUE) %>%
  #   pack_rows(index=table(ps.table[, "Genus"]), bold=FALSE, label_row_css="border-bottom: 0px solid;") %>%
  #   save_kable(file=FileName, zoom=3)
  # print(ps.table)
}
```

## Phyloseq Processing

```{r Phyloseq Processing}
phyloseq.processing <- function(group.id, initial.file.path, phyloseq.path, output.path, pheno.path=NULL, grouped=FALSE, rareify=FALSE){
  defaultW <- getOption("warn")
  options(warn=-1)
  phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
  phyloseq.control <- readRDS(paste0(phyloseq.path, "/phyloseq.control.", group.id, ".rds"))
  ### Import phenotype file ###
        cat("\n  ", paste("Import phenotype file"))
  if(!is.null(pheno.path)){
    infant.nasal.pheno <- read_csv(pheno.path)
  }
  
  sample_data(phyloseq.control) <- sample_data(phyloseq.control)[c(which(sample_data(phyloseq.control)[, "group"]=="positive control"), 
                                                                   which(sample_data(phyloseq.control)[, "group"]=="negative control"), 
                                                                   which(sample_data(phyloseq.control)[, "group"]=="water")), ]
  
  # print(sum(sample_sums(subset.object(phyloseq, "True Sample"))))
  if(grouped==FALSE){
    phyloseq <- prune_samples(readRDS(paste0(initial.file.path, "/over.min.reads.", group.id, ".rds")), phyloseq)
  }
  
  # print(sum(sample_sums(subset.object(phyloseq, "True Sample"))))
  samdf <- data.frame(sample_data(phyloseq))
  samdf[, "row.names"]<- rownames(samdf)
  samdf <- samdf[which(samdf[, "Sample_or_Control"]=="True Sample"), ]
  samdf[, 2] <- sapply(samdf[, 2], as.numeric)
  if(!is.null(pheno.path)){
    samdf <- merge(samdf, infant.nasal.pheno, by=c("family", "group", "form"), no.dups=FALSE)
  }
  # print(sample_names(phyloseq)[which(sample_names(phyloseq) %in% samdf[, "row.names"]==FALSE)])
  # print(samdf[, c("row.names")])
  rownames(samdf) <- samdf[, c("row.names")]
  samdf <- samdf[, -c(which(colnames(samdf)=="samID"), which(colnames(samdf)=="Sample_or_Control"), which(colnames(samdf)=="row.names"))]
  samdf[, "family"] <- sapply(samdf[, "family"], as.character)
  sample_data(phyloseq) <- samdf
  
  
  sample_data(phyloseq)[, "reads"] <- sample_sums(phyloseq)
  
  saveRDS(phyloseq.control, paste0(output.path, "/phyloseq.control.", group.id, ".rds"))
  saveRDS(phyloseq, paste0(output.path, "/phyloseq.", group.id, ".rds"))

  if(rareify==TRUE){
          cat("\n  ", paste("Min reads for phyloseq"))
          cat("\n    ", paste(min(sample_sums(phyloseq))))
          cat("\n  ", paste("Min reads for phyloseq.control"))
          cat("\n    ", paste(min(sample_sums(phyloseq.control))))
    # phyloseq <- rarefy_even_depth(phyloseq, rngseed=98, sample.size=3000)
    phyloseq <- rarefy_even_depth(phyloseq, rngseed=98, sample.size=2500)
    phyloseq.control <- rarefy_even_depth(phyloseq.control, rngseed=98, sample.size=2500)
    sample_data(phyloseq)[, "reads"] <- sample_sums(phyloseq)
  }
  phyloseq.control <- change.phyloseq.pheno.name(phyloseq.control, NULL, "samID", "Sample", NULL)
  
  # for(a in 1:length(asv.prune)){
  #   phyloseq <- remove.specific.taxa(phyloseq, asv.prune[a])
  # }
  saveRDS(phyloseq, paste0(output.path, "/phyloseq.pruned.", group.id, ".asv.rds"))
  saveRDS(phyloseq.control, paste0(output.path, "/phyloseq.pruned.control.", group.id, ".asv.rds"))
  saveRDS(merge_phyloseq(phyloseq, phyloseq.control), paste0(output.path, "/phyloseq.pruned.both.", group.id, ".asv.rds"))
  # print(sum(sample_sums(phyloseq)))
  phyloseq.reset <- phyloseq
  phyloseq.control.reset <- phyloseq.control
  
  # taxa.glom <- c("Family", "Genus", "Species")
  # for(i in 1:length(taxa.glom)){
  #   phyloseq <- phyloseq.reset
  #   phyloseq.control <- phyloseq.control.reset
  #   
  #   phyloseq.taxa <- tax_glom(phyloseq, taxa.glom[i], NArm=FALSE)
  #   saveRDS(phyloseq.taxa, paste0(output.path, "/phyloseq.pruned.", group.id, ".", tolower(taxa.glom[i]), ".rds"))
  #   
  #   phyloseq.control.taxa <- tax_glom(phyloseq.control, taxa.glom[i], NArm=FALSE)
  #   saveRDS(phyloseq.control.taxa, paste0(output.path, "/phyloseq.pruned.control.", group.id, ".", tolower(taxa.glom[i]), ".rds"))
  #   
  #   phyloseq.both <- merge_phyloseq(phyloseq.taxa, phyloseq.control.taxa)
  #   saveRDS(phyloseq.both, paste0(output.path, "/phyloseq.pruned.both.", group.id, ".", tolower(taxa.glom[i]), ".rds"))
  #   
  #   # if(taxa.glom=="Genus"){
  #   #   short.taxa.name(group.id, "Genus", "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium", output.path)
  #   #   short.taxa.name(group.id, "Genus", "Methylobacterium-Methylorubrum", output.path)
  #   # }else if(taxa.glom=="Species"){
  #   #   short.taxa.name(group.id, "Species", "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium Unclassified Species asv160", output.path)
  #   #   short.taxa.name(group.id, "Species", "populi", output.path)
  #   # }
  # }
  
  save(list=ls(all.names=TRUE), file=file.path(paste0(output.path, "/envi.", group.id, ".RData")))
  options(warn=defaultW)
}
```

## Phenotype Counts

```{r Pheno Counts}
pheno.counts <- function(phyloseq, group.id, phenos, phyloseq.name, output.path){
  for(i in 1:length(phenos)){
    count.by.group <- function(phyloseq, group.a, group.b=NULL){
      pheno.frame <- as.data.frame(sample_data(phyloseq))
      name <- paste0(group.a)
      if(!is.null(group.b)){
        pheno.number <- pheno.frame %>% 
          dplyr::group_by_(group.a, group.b) %>% 
          dplyr::summarise(number=n()) %>%
          as.data.frame(.)
        pheno.number[is.na(pheno.number)] <- "NA"
        colnames(pheno.number)[1] <- "pheno"
        pheno.number[, "pheno.name"] <- phenos[i]
        return(pheno.number)
      } else {
        pheno.number <- pheno.frame %>% 
          dplyr::group_by_(group.a) %>% 
          dplyr::summarise(number=n()) %>%
          as.data.frame(.)
        pheno.number[is.na(pheno.number)] <- "NA"
        colnames(pheno.number)[1] <- "pheno"
        pheno.number[, "pheno.name"] <- phenos[i]
        return(pheno.number)
      }
    }
    assign(paste0("phyloseq.pruned.", group.id, ".", phenos[i]), 
           count.by.group(phyloseq, phenos[i]))
  }
  if(length(phenos)==1){
    combined <- get(paste0("phyloseq.pruned.", group.id, ".", phenos[i]))
  }else{
    for(i in 1:length(phenos)){
      if(i==1){
        combined <- get(paste0("phyloseq.pruned.", group.id, ".", phenos[i]))
        list.all <- get(paste0("phyloseq.pruned.", group.id, ".", phenos[i]))[, c(1, 2)]
      }else{
        combined <- rbind(combined, get(paste0("phyloseq.pruned.", group.id, ".", phenos[i])))
        list.all <- c(list.all, get(paste0("phyloseq.pruned.", group.id, ".", phenos[i]))[, c(1, 2)])
      }
    }
    combined[, "pheno.name"] <- factor(combined[, "pheno.name"], levels=unique(combined[, "pheno.name"]))
  }
  # print(combined)  
  # print(combined[, c(1, 2)])
  FileName <- paste0(output.path, "/", phyloseq.name, ".pheno.counts", ".png")
  # kbl.combined <- combined[, c(1, 2)] %>%
  #   kbl(centering=TRUE, escape=FALSE, col.names=c("Phenotype", "Count"), align="c") %>%
  #   kable_styling(font_size=15) %>%
  #   kable_classic("striped", full_width=F, html_font="Calibri", position="center") %>%
  #   row_spec(0, bold=TRUE) %>%
  #   pack_rows(index=table(combined[, "pheno.name"]), bold=FALSE, label_row_css="border-bottom: 0px solid;") %>%
  #   save_kable(file=FileName, zoom=3)
}
```

# Diversity

## Alpha Diversity Plots of Phenotype Group by Sample Type

```{r Alpha diversity plots of pheno.group by sample type}
richness.alpha <- function(phyloseq, taxa.level, pheno.group, path, phyloseq.name, reads=TRUE){
  path <- paste0(path, "/Output/Diversity/", taxa.level)
  
  type <- identify.type(phyloseq)
  
  cat("\n\n", paste("Richness for", pheno.group))
  sample_data(phyloseq) <- as.data.frame(sample_data(phyloseq))[which(!is.na(as.data.frame(sample_data(phyloseq))[, pheno.group])), ]

  comb.list <- uln(unique(as.data.frame(sample_data(phyloseq))[, pheno.group]))
  comparisons <- as.list(as.data.frame(t(CombPairs(comb.list))))
  size <- 4
  # length(comparisons)
  pheno <- function(pheno){
    if(pheno!="All"){
            cat("\n  ", paste("Subset sample data by", pheno))
      sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("type")]==c(pheno))), ]
    }
          cat("\n    ", paste("Observed, shannon, and richness plot generation"))
            
    # colors <- color.list(5)[c(5, 7)]
    if(length(comparisons)==1){
      colors <- color.list(5)[c(2, 6)]
    }else{
      colors <- color.list(5)[-c(1)]
    }
    
    richness <- estimate_richness(phyloseq, measures=c("Shannon", "Observed"))
    rownames(richness) <- str_remove(rownames(richness), "X")
    rownames(richness) <- rownames(data.frame(sample_data(phyloseq)))
    richness <- merge(data.frame(sample_data(phyloseq)), richness, by='row.names')
    colnames(richness)[which(colnames(richness)=="Observed")] <- paste("Observed", taxa.level)
    # colnames(richness)[which(colnames(richness)=="Observed")] <- "Number of observed species"
    colnames(richness)[which(colnames(richness)=="Shannon")] <- "Shannon index"
    colnames(richness)[which(colnames(richness)=="reads")] <- "Sample Reads"
    
    # print(richness)
    if(pheno.group=="birth_type"){
      richness$birth_type=factor(richness$birth_type, levels=c("Vaginal", "C-Section"))
    }else if(pheno.group=="form"){
      richness$form=factor(richness$form, levels=c("Initial", "Follow up"))
    }
    
    if(reads==TRUE){
      alpha.list <- c(paste("Observed", taxa.level), "Shannon index", "Sample Reads")
    }else{
      alpha.list <- c(paste("Observed", taxa.level), "Shannon index")
      # alpha.list <- c("Number of observed species", "Shannon index")
    }
    
    plot.list <- list()
    richness.reset <- richness
    for(a in 1:length(alpha.list)){
      richness <- richness.reset
      colnames(richness)[which(colnames(richness)==(alpha.list[a]))] <- "alpha"
      stat <- as.formula(paste("alpha", "~", pheno.group))
      stat.test <- richness %>% 
        wilcox_test(stat, comparisons=comparisons) %>%
        add_xy_position(x=pheno.group) %>%
        mutate(myformatted.p=paste0("p=", signif(p, 2)))
      # length(which(richness[, "birth_type"]=="Vaginal"))
      n_fun <- function(x){
        return(data.frame(y=-0.07, label=paste0("n=", length(x))))
      }
      min <- unlist(unname(stat.test[1, "y.position"]))
      for(b in 1:nrow(stat.test)){
        stat.test[b, "y.position"] <- (min+((b-1)*(min/(10-((length(comb.list)-1)*2)))))
      }
      # print(richness)
      plot <- ggplot(richness, aes(x=get(pheno.group), y=alpha)) +
        theme_classic() +
        geom_boxplot(outlier.size=-1) +
        # geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.75, dodge.width=0.75), 
                   # size=2.5, aes(fill=get(pheno.group))) +
        # scale_color_manual(values=colors) +
        # guides(color="none") +
        ylab(alpha.list[a]) +
        ylim(NA, (uln(stat.test[nrow(stat.test), "y.position"])+(min/(14-((length(comb.list)-1)*2)))))+
        stat_pvalue_manual(stat.test, label="myformatted.p", tip.length=0, bracket.size=0.5, vjust=-0.5) +
        stat_summary(fun.data=n_fun, geom="text")
      
      if(reads==FALSE){
        plot <- plot + geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.75, dodge.width=0.75), 
                                  size=2.5, aes(fill=get(pheno.group))) + guides(fill="none")
      }else{
        plot <- plot + geom_text(position=position_jitterdodge(seed=1L, jitter.width=0.75, dodge.width=0.75), 
                                 size=2.5, aes(label=family, fill=get(pheno.group))) + guides(fill="none")
      }
      
      plot <- plot + theme(axis.title.x=element_blank(), axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
                           axis.text.x=element_text(color="black", face="bold"), axis.ticks.x=element_blank(), 
                           panel.border=element_rect(color="black", fill=NA, size=0.75))
      
      plot.list[[a]] <- plot
    }
    if(reads==TRUE){
      plot <- ggarrange(plotlist=plot.list, ncol=3, nrow=1)
    }else{
      plot <- ggarrange(plotlist=plot.list, ncol=2, nrow=1)
    }
    if(str_detect(pheno.group, "_")==FALSE){
      # plot <- annotate_figure(plot, top=text_grob(label=paste(str_to_title(pheno.group)))) +
      plot <- annotate_figure(plot, top=text_grob(label=paste(type))) +
               bgcolor("white") + border(color="white")
    }else{
      plot <- annotate_figure(plot, top=text_grob(label=paste(str_to_title(str_replace(pheno.group, "_", " "))))) +
               bgcolor("white") + border(color="white")
    }
    
      
    if(pheno=="All"){
      if (phyloseq.name=="NULL"){
        fileName <- paste0(path, "/alpha.diversity.", pheno, ".", tolower(pheno.group), ".png")
      }else{
        fileName <- paste0(path, "/alpha.diversity.", phyloseq.name, ".", tolower(taxa.level), ".", 
                           tolower(pheno.group), ".png")
      }
      if(reads==TRUE){
        ggsave(filename=fileName, plot=plot, width=size*3, height=size, scale=0.8)
        # ggsave(filename=fileName, plot=plot, width=18, height=6, scale=0.8)
      }else{
        ggsave(filename=fileName, plot=plot, width=size*2, height=size, scale=0.8)
        # ggsave(filename=fileName, plot=plot, width=12, height=6, scale=0.8)
      }
    }else{
      plot <- annotate_figure(plot, top=text_grob(label=paste(pheno, str_to_title(pheno.group)))) +
        bgcolor("white") + border(color="white")
      return(plot)
    }
  }
  
  if(pheno.group!="type" && pheno.group!="family"){
    if(length(unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"]))))==1){
      # print("yes")
      pheno("All")
    }else{
      phenos <- unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"])))
      # print(phenos)
      pheno("All")
      plot.list <- list()
      for(a in 1:length(phenos)){
        if(pheno.group=="time"||pheno.group=="birth_type"||pheno.group=="sex"||pheno.group=="breastmilk"){
          if(phenos[a]=="BNA"||phenos[a]=="BSA"||phenos[a]=="BST"){
            plot.list[[a]] <-pheno(phenos[a])
          }
        }else{
          plot.list[[a]] <- pheno(phenos[a])
        }
      }
      if(length(phenos)==2){
        all.plot <- ggarrange(plotlist=plot.list, ncol=1, nrow=(2))
        FileName <- paste0(path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, width=((size*3)*1), height=(size*2), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(18*1), height=(6*2), scale=0.8)
      }else if(length(phenos)==3){
        all.plot <- ggarrange(plotlist=plot.list, ncol=1, nrow=(length(phenos)/1))
        FileName <- paste0(path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, width=((size*3)*1), height=(size*(length(phenos)/1)), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(18*1), height=(6*(length(phenos)/1)), scale=0.8)
      }else{
        all.plot <- ggarrange(plotlist=plot.list, ncol=2, nrow=(ceiling(length(phenos)/2)))
        FileName <- paste0(path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, width=((size*3)*2), height=(size*(ceiling(length(phenos)/2))), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(18*2), height=(6*(ceiling(length(phenos)/2))), scale=0.8)
      }
    }
  }
}
```

## Beta diversity plots

```{r Beta diversity plots of pheno.group by sample type, warning=FALSE}
plotBeta <- function(phyloseq, taxa.level, pheno.group, path, phyloseq.name, binaryGroup=TRUE, type.subset=FALSE, grouped.title=NULL){
  path <- paste0(path, "/Output/Diversity/", taxa.level)
        cat("\n\n", paste("Beta diversity for", pheno.group))
  distance="bray"
  num.types <- length(unname(unlist(unique(as.data.frame(sample_data(phyloseq))[, "type"]))))
  sample_data(phyloseq) <- sample_data(phyloseq)[which(!is.na(sample_data(phyloseq)[, pheno.group])), ]
  size <- 4
  pheno <- function(pheno){
    # colors <- color.list(5)[c(2, 6)]
    if(pheno=="all"){
      title <- "All sample types"
    }else{
      title <- pheno
    }
    
    if(pheno!="all"){
            cat("\n  ", paste("Subset sample data by", pheno))
      samdf <- data.frame(sample_data(phyloseq))
      if(pheno %in% samdf[, "type"]){
        sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("type")]==c(pheno))), ]
      }else if(pheno %in% samdf[, "region"]){
        sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("region")]==c(pheno))), ]
      }else if(pheno %in% samdf[, "family"]){
        sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("family")]==c(pheno))), ]
      }else if(pheno %in% samdf[, "group"]){
        sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("group")]==c(pheno))), ]
      }
      # sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("type")]==c(pheno))), ]
    }else{
            cat("\n  ", paste("All samples"))
    }
    
    if(length(unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, pheno.group]))))>2){
      binaryGroup=FALSE
      colors <- color.list(5)[-c(1)]
    }else{
      colors <- color.list(5)[c(2, 6)]
    }
    if(any(is.na(sample_data(phyloseq)[[pheno.group]]))){
      stop("please clear up all the NA values in the category you are interested in.")
    }
    if(length(distance)==0){
      stop("please specify the name of distance metric you want to use.")
    }
    
    phyloseq::sample_data(phyloseq)[[pheno.group]] <- as.factor(phyloseq::sample_data(phyloseq)[[pheno.group]])
    sample_data(phyloseq)$birth_type=factor(sample_data(phyloseq)$birth_type, levels=c("Vaginal", "C-Section"))
    if (binaryGroup){
            cat("\n    ", paste("Adonis"))
      # Function to run adonis test on a physeq object and a variable from metadata
      
      plist <- list()
      for (i in (1:length(distance))){
        res <- doadonis(phyloseq, distance=distance[i], factor=pheno.group)
        ps <- phyloseq::ordinate(phyloseq, method="PCoA", distance=distance[i])
        rela_eig <- percent(ps$values$Relative_eig, accuracy=0.1)
        
              cat("\n    ", paste("Beta diversity plot generation"))
        p <- phyloseq::plot_ordination(phyloseq, ps, color=pheno.group) +
          theme_classic() +
          geom_point(size=2.5) +
          scale_color_manual(values=colors) +
          stat_ellipse(type="norm") +
          xlab(paste0("PC 1 (", rela_eig[1], ")")) + ylab(paste0("PC 2 (", rela_eig[2], ")")) +
          labs(subtitle=paste0("p=", as.character(res[1]))) +
          theme(plot.title=element_text(hjust=0.5), strip.background=element_blank(), 
                strip.text.x=element_blank(), panel.grid.major=element_blank(), panel.grid.minor=element_blank(), 
                legend.position="bottom", 
                # legend.direction="vertical", 
                legend.title=element_blank(), legend.box.spacing=unit(0, 'cm'), legend.margin=margin(0, unit='cm'), 
                plot.margin=unit(c(0.5, 1, 0, 0.5), units="line"), 
                axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
                # legend.text=element_text(face="bold"), 
              panel.border=element_rect(color="black", fill=NA, size=0.75))
        # text=element_text(face="bold", size=10), 
        # p <- annotate_figure(p, top=text_grob(label=paste(pheno, str_to_title(pheno.group)))) +
        
        if(str_detect(pheno.group, "_")==FALSE){
          p <- annotate_figure(p, top=text_grob(label=paste(str_to_title(pheno.group)))) +
                   bgcolor("white") + border(color="white")
        }else{
          p <- annotate_figure(p, top=text_grob(label=paste(str_to_title(str_replace(pheno.group, "_", " "))))) +
                   bgcolor("white") + border(color="white")
        }
        
        plist[[i]] <- p
      }
      
    }else if(!binaryGroup){
      # Function to run adonis test on a physeq object and a variable from metadata
            cat("\n    ", paste("Anosim"))
      doanosim <- function(phyloseq, distance=distance, factor=pheno.group) {
        bdist <- phyloseq::distance(phyloseq, distance)
        col <- as(sample_data(phyloseq), "data.frame")[, pheno.group]

        # Anosim test
        anosim.dist <- vegan::anosim(bdist, col, permutations=999)
        p <- signif(anosim.dist$signif, digits=2)
        r2 <- signif(anosim.dist$statistic, digits=2)
        res <- c(p, r2)
              cat("\n      ", paste("p=", p))
        return(res)
      }
      
      plist <- list()
      for (i in (1:length(distance))){
        res <- doanosim(phyloseq, distance=distance[i], factor=pheno.group)
        ps <- phyloseq::ordinate(phyloseq, method="PCoA", distance=distance[i])
        rela_eig <- percent(ps$values$Relative_eig, accuracy=0.1)
        
              cat("\n    ", paste("Beta diversity plot generation"))
        p <- phyloseq::plot_ordination(phyloseq, ps, color=pheno.group) +
          theme_classic() +
          geom_point(size=2.5) +
          scale_color_manual(values=colors) +
          stat_ellipse(type="norm") +
          xlab(paste0("PC 1 (", rela_eig[1], ")")) + ylab(paste0("PC 2 (", rela_eig[2], ")")) +
          labs(subtitle=paste0("p=", as.character(res[1]))) +
          theme(plot.title=element_text(hjust=0.5), strip.background=element_blank(), 
                strip.text.x=element_blank(), panel.grid.major=element_blank(), panel.grid.minor=element_blank(), 
                legend.position="bottom", 
                # legend.direction="vertical", 
                legend.title=element_blank(), legend.box.spacing=unit(0, 'cm'), legend.margin=margin(0, unit='cm'), 
                plot.margin=unit(c(0.5, 1, 0, 0.5), units="line"), 
                axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
                # legend.text=element_text(face="bold"), 
              panel.border=element_rect(color="black", fill=NA, size=0.75))
        #scale_color_discrete(name="Type")
        ##today
        if(pheno!="All"){
          # p <- annotate_figure(p, top=text_grob(label=paste(pheno, str_to_title(pheno.group)))) +
          if(str_detect(pheno.group, "_")==FALSE){
            p <- annotate_figure(p, top=text_grob(label=paste(str_to_title(pheno.group)))) +
                     bgcolor("white") + border(color="white")
          }else{
            p <- annotate_figure(p, top=text_grob(label=paste(str_to_title(str_replace(pheno.group, "_", " "))))) +
                     bgcolor("white") + border(color="white")
          }
        }
        plist[[i]] <- p
      }
    }
    if (length(plist)==1){
      ggp <- ggarrange(plist[[1]])
    }else if(length(plist)==2){
      ggp <- ggarrange(plist[[1]], plist[[2]], ncols=2, common.legend=FALSE)
    }else if (length(plist)==3){
      ggp <- ggarrange(plist[[1]], plist[[2]], plist[[3]], ncols=2, nrows=2, common.legend=FALSE)
    }
    
    # assign(paste0(pheno, ".p.beta"), annotate_figure(ggp, top=text_grob(paste0(title))), envir=.GlobalEnv)
    # assign(paste0(pheno, ".p.beta"), ggp, envir=parent.frame())
    if(pheno=="all"){
      if (phyloseq.name=="NULL"){
        fileName <- paste0(path, "/beta.diversity.", pheno.group, ".", get(pheno), ".png")
      } else {
        fileName <- paste0(path, "/beta.diversity.", phyloseq.name, ".", tolower(taxa.level), ".", pheno.group, ".png")
      }
      ggsave(ggp, filename=fileName, width=size, height=size, scale=0.8)
      # ggsave(ggp, filename=fileName, width=6, height=6, scale=0.8)
    }else{
      return(ggp)
    }
  }
  if(pheno.group!="type"&&pheno.group!="family"){
    if(length(unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"]))))==1){
      pheno("all")
    }else{
      phenos <- unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"])))
      pheno("all")
      plot.list <- list()
      for(a in 1:length(phenos)){
        if(pheno.group=="time"||pheno.group=="birth_type"||pheno.group=="sex"||pheno.group=="breastmilk"){
          if(phenos[a]=="BNA"||phenos[a]=="BSA"||phenos[a]=="BST"){
            plot.list[[a]] <- pheno(phenos[a])
          }
        }else{
          plot.list[[a]] <- pheno(phenos[a])
        }
      }
      if(length(phenos)==2){
        all.plot <- ggarrange(plotlist=plot.list, nrow=1, ncol=(2))
        FileName <- paste0(path, "/beta.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, height=(size*1), width=(size*2), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(6*1), height=(6*2), scale=0.8)
      }else if(length(phenos)==3){
        all.plot <- ggarrange(plotlist=plot.list, nrow=1, ncol=(3))
        FileName <- paste0(path, "/beta.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, height=(size*1), width=(size*3), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(6*1), height=(6*3), scale=0.8)
      }else{
        all.plot <- ggarrange(plotlist=plot.list, nrow=3, ncol=(ceiling(length(phenos)/3)))
        FileName <- paste0(path, "/beta.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, height=(size*3), width=(size*(ceiling(length(phenos)/3))), scale=0.8)
        # ggsave(filename=FileName, plot=all.plot, width=(6*3), height=(6*(ceiling(length(phenos)/3))), scale=0.8)
      }
    }
  }else if(pheno.group=="type"&&length(uuln(as.data.frame(sample_data(phyloseq))[, "type"]))!=1){
    pheno("all")
    phenos <- c("mother", "baby")
    for(a in 1:length(phenos)){
      plot.list[[a]] <- pheno(phenos[a])
    }
    all.plot <- ggarrange(plotlist=plot.list, nrow=1, ncol=(2))
        FileName <- paste0(path, "/beta.diversity.", phyloseq.name, ".", pheno.group, ".facet.png")
        ggsave(filename=FileName, plot=all.plot, height=(size*1), width=(size*2), scale=0.8)
  }
}
```

# Phyloseq Object Dataframe for Specific Taxa

```{r Phyloseq object dataframe for specific taxa}
phyloseq.data.frame <- function(phyloseq, taxa.level, n=10, specific.taxa=NULL, counts=FALSE, phyloseq.name){

  if(!is.null(specific.taxa)){
    prune <- rownames(tax_table(phyloseq)[c(which(tax_table(phyloseq)[, taxa.level]==specific.taxa)), ])
  } else{
    prune <- names(sort(taxa_sums(phyloseq), decreasing=TRUE))[1:n]
  }

  ps.taxa <- phyloseq
  if(counts==FALSE){
    ps.taxa <- transform_sample_counts(phyloseq, function(y) y/sum(y))
  }
  
  ### Prune for na values ###
  na.value <- row.names(tax_table(phyloseq)[!is.na(tax_table(phyloseq)[, taxa.level]), ])
  phyloseq <- prune_taxa(na.value, phyloseq)
  
  ps.taxa <- prune_taxa(prune, ps.taxa)
  
  ### OTU and Tax dataframe ###
  otu.taxa <- data.frame(otu_table(ps.taxa))
  otu.taxa.t=as.data.frame(t(otu.taxa), check.names=TRUE)
  tax.taxa <- data.frame(tax_table(ps.taxa))
  ps.table <- cbind(tax.taxa, otu.taxa.t)
  

  ### Save phyloseq dataframe ###
  # write.csv(ps.table, file=paste0(output.path, "/", phyloseq.name, ".", taxa.level, ".", specific.taxa, ".table.csv"))
  # assign(paste0(phyloseq.name, ".", taxa.level, ".", specific.taxa, ".table"), ps.table, envir=.GlobalEnv)
  return(ps.table)
}
```

# Species correlation

```{r Species correlation}
species.correlation <- function(phyloseq, taxa.level, n=10, specific.taxa=NULL, pheno.group, pheno.a, pheno.b, counts=FALSE, output.path, phyloseq.name, grouped.title=NULL){
  
  families <- identify.families(phyloseq)
  assign(paste0(phyloseq.name, ".", taxa.level, ".", specific.taxa, ".table"), phyloseq.data.frame(phyloseq, taxa.level, n, specific.taxa, counts, phyloseq.name))
   
  table.prune <-
    t(get(paste0(phyloseq.name, ".", taxa.level, ".", specific.taxa, ".table"))[c(1), -c(1:7)])
  # print(table.prune)
  colnames(table.prune) <- "relative.abundance"
  pheno.table <- as.data.frame(sample_data(phyloseq))[, c("family", pheno.group)]
  merged.family.comparison.linear.plot <- merge(table.prune, pheno.table, by="row.names")[, -c(1)]
  merged.family.comparison.linear.plot.reset <- merged.family.comparison.linear.plot
  for(a in 1:length(pheno.a)){
    merged.family.comparison.linear.plot <- merged.family.comparison.linear.plot.reset
    # print(merged.family.comparison.linear.plot)
    if(pheno.a[a]=="BNA"){
      merged.family.comparison.linear.plot <-
        merged.family.comparison.linear.plot[-c(which(merged.family.comparison.linear.plot[, "type"]=="BSA"), 
                                                which(merged.family.comparison.linear.plot[, "type"]=="BST")), ]
    }else if(pheno.a[a]=="BST"){
      merged.family.comparison.linear.plot <-
        merged.family.comparison.linear.plot[-c(which(merged.family.comparison.linear.plot[, "type"]=="BSA"), 
                                                which(merged.family.comparison.linear.plot[, "type"]=="BNA")), ]
    }else if(pheno.a[a]=="BSA"){
      merged.family.comparison.linear.plot <-
        merged.family.comparison.linear.plot[-c(which(merged.family.comparison.linear.plot[, "type"]=="BNA"), 
                                                which(merged.family.comparison.linear.plot[, "type"]=="BST")), ]
    }

    merged.family.comparison.linear.plot <- as.data.frame(pivot_wider(merged.family.comparison.linear.plot, 
                                                                      names_from=pheno.group, 
                                                                      values_from="relative.abundance"))
    # merged.family.comparison.linear.plot <- as.data.frame(merged.family.comparison.linear.plot)
    merged.family.comparison.linear.plot[, "family"] <- factor(merged.family.comparison.linear.plot[, "family"], 
                                                              levels=families)
    colors <- unlist(mapply(brewer.pal, brewer.pal.info[brewer.pal.info$category=='qual', ]$maxcolors, 
                               rownames(brewer.pal.info[brewer.pal.info$category=='qual', ])))
    colors <- paletteer_d("ggsci::default_igv")
    merged.family.comparison.linear.plot.reset.2 <- merged.family.comparison.linear.plot
    plots.list <- list()
    for(b in 1:length(pheno.b)){
      merged.family.comparison.linear.plot <- merged.family.comparison.linear.plot.reset.2
      merged.family.comparison.linear.plot <- as.data.frame(merged.family.comparison.linear.plot)
      m <- lm(get(pheno.a[a]) ~ get(pheno.b[b]), merged.family.comparison.linear.plot)
      # m <- lm(pheno.a[a] ~ pheno.b[b], merged.family.comparison.linear.plot)
      r2=format(summary(m)$r.squared, digits=2)
      p=format(cor.test(merged.family.comparison.linear.plot[, pheno.a[a]], merged.family.comparison.linear.plot[, pheno.b[b]], 
                        alternative="greater", method="pearson")$p.value, digits=2)
      if(r2>0.1){
        plots.list[[b]] <- ggplot(merged.family.comparison.linear.plot, aes(x=get(pheno.a[a]), y=get(pheno.b[b]), fill(family))) +
                 geom_point(alpha=1, size=3, aes(color=family)) +
                 scale_color_manual(values=colors) +
                 guides(color=guide_legend(ncol=3)) +
                 theme(text=element_text(size=20)) +
                 geom_smooth(method=lm, color="black", se=TRUE) +
                 labs(caption=bquote(atop(r^2==~ .(r2), p==~ .(p))), x=paste(pheno.a[a]), y=pheno.b[b])
      }
    }
    if(length(plots.list)>0){
      all.plot <- ggarrange(plotlist=plots.list, ncol=1, nrow=1) 
      all.plot <- annotate_figure(all.plot, top=text_grob(label=paste(specific.taxa, "relative abundance correlation"), 
                                                            color="black", size=20))
      if(!is.null(grouped.title)){
        all.plot <- annotate_figure(all.plot, top=text_grob(label=grouped.title, color="black", size=20))
      }
      FileName <- paste0(output.path, "/correlation.plot.", phyloseq.name, ".", pheno.a[a], ".", specific.taxa, ".png")
      ggsave(filename=FileName, plot=all.plot, height=(8*1), width=(10*1))
    }
  }
}
```

# Taxa Analysis

## Top table

```{r Top table}
top.table <- function(phyloseq, pheno.group, taxa.level, path=NULL, n=10, save=TRUE, top=NA, reads=FALSE){
  if.asv <- if.asv(taxa.level)
  taxa.level <- if.asv[1]
  asv <- if.asv[2]
  
  phyloseq.name <- deparse(substitute(phyloseq))
  n.tax.col <- which(colnames(as.data.frame(tax_table(phyloseq)))==taxa.level)
  if(n!=0){
    if(is.na(top)){
      top <- top.taxa.n(phyloseq, taxa.level, n=n, pheno.group, reads=reads)
    }
    phyloseq.top <- phyloseq.processing.compare(phyloseq, pheno.group, top, TRUE, reads=reads)
  } else {
    phyloseq.top <- phyloseq
  }
  n.pheno <- nrow(as.data.frame(otu_table(phyloseq.top)))

  otu.top <- as.data.frame(otu_table(phyloseq.top))
  sam.dta <- as.data.frame(sample_data(phyloseq.top))
  otu.top.t=t(otu.top)
  for (i in 1:n.pheno) {
    colnames(otu.top.t)[i][1] <- paste(colnames(otu.top.t)[i][1], "Abundance")
  }
  
  tax.top <- data.frame(tax_table(phyloseq.top))
  tax.top <- subset(tax.top, select=c(1:(n.tax.col)))
  
  phyloseq.top.table <- cbind(tax.top, otu.top.t)

  if(n.pheno==1){
    col.sums.percent <- unname(sum(phyloseq.top.table[, c((n.tax.col+1))]))
  }else{
    col.sums.percent <- unname(colSums(phyloseq.top.table[, c((n.tax.col+1):(n.tax.col+n.pheno))]))
  }
  
  for(i in 1:length(col.sums.percent)){
    col.sums.percent[i] <- round((as.numeric(col.sums.percent[i])*100), 2)
  }
  
  for (i in 1:n.pheno) {
    phyloseq.top.table[, c(n.tax.col+n.pheno+i)] <- list(paste0(round((as.numeric(phyloseq.top.table[, c(n.tax.col+i)])*100), 2), "%"))
    colnames(phyloseq.top.table)[n.tax.col+n.pheno+i][1] <- paste(colnames(phyloseq.top.table)[n.tax.col+i][1], "Percent")
  }
  phyloseq.top.table <- phyloseq.top.table[order(as.numeric(as.character(phyloseq.top.table
                                                                         [, (n.tax.col+1)])), decreasing=TRUE), ]
  if(n!=0){
    return(phyloseq.top.table)
  }
  sum.row <- nrow(phyloseq.top.table)+1
  for (i in 1:n.pheno) {
    phyloseq.top.table[sum.row, c(n.tax.col+2+i)] <- paste0(col.sums.percent[i], "%")
  }
  
  # rownames(phyloseq.top.table) <- NULL
  phyloseq.top.table[is.na(phyloseq.top.table)] <- 0
  
  if(asv==TRUE){
    taxa.level <- "ASV"
  }
  
  if(save==TRUE){
    if (!is.null(path)){
      FileName <- paste0(path, "/", phyloseq.name, ".top.", tolower(taxa.level), ".", 
                         tolower(pheno.group), ".csv")
    } else {
      FileName <- paste0(phyloseq.name, ".top.", tolower(taxa.level), ".", 
                         tolower(pheno.group), ".csv")
    }
    write.csv(phyloseq.top.table, file=FileName)
  }
  
  if(n==0){
    return.list <- list(phyloseq.top.table, col.sums.percent[1], col.sums.percent[2]) 
    return(return.list)
  }
}
```

## Plot Table Preperation
```{r Plot top bar}
plot.table.prep <- function(phyloseq, taxa.level, n=10, pheno.group, group.by.pheno=NULL, top=NA, reads=FALSE){
  samdf <- data.frame(sample_data(phyloseq))
  if (!is.null(group.by.pheno)){
    phyloseq <- subset.object(phyloseq, group.by.pheno)
  }
  
  phyloseq.top.table <- top.table(phyloseq, pheno.group, taxa.level, path, n, FALSE, top, reads=reads)
  
  print(phyloseq.top.table)
  
  if.asv <- if.asv(taxa.level)
  taxa.level <- if.asv[1]
  asv <- if.asv[2]
  
  n.pheno <- sum(!is.na(uuln(as.data.frame(sample_data(phyloseq))[, c(pheno.group)])))
  n.tax.col <- which(colnames(as.data.frame(tax_table(phyloseq)))==taxa.level)
  phyloseq.top.table <- phyloseq.top.table[order(as.numeric(as.character(phyloseq.top.table[, (n.tax.col+1)])), 
                                                 decreasing=TRUE), ]
  if(n.pheno > 0){
    plot.table <- data.frame(phyloseq.top.table, check.names=FALSE)
    # if(taxa.level=="Species"){
    #   for(i in 1:nrow(plot.table)) {
    #     if(grepl("Unclassified", plot.table[i, 7], fixed=TRUE)==FALSE){
    #       plot.table[i, 7] <- paste(plot.table[i, 6], plot.table[i, 7])
    #     } 
    #   }
    # } 
    if(n.pheno!=1){
        # sorted <- plot.table[order(as.numeric(as.character(rowSums(plot.table[, c((n.tax.col+1):(n.tax.col+n.pheno))]))), 
                                 # decreasing=TRUE), n.tax.col]
        sorted <- asv.to.taxa(phyloseq, top.taxa.n(phyloseq, taxa.level, n=n, pheno.group), taxa.level)
    }else{
      sorted <- plot.table[order(as.numeric(as.character(plot.table[, c(n.tax.col+1)])), decreasing=TRUE), n.tax.col]
      # sorted <- asv.to.taxa(phyloseq, top.taxa.n(phyloseq, taxa.level, n=n, pheno.group), taxa.level)
    }
    
    print(plot.table)

    plot.table[is.na(plot.table)] <- 0
    plot.table <- plot.table[, c(1:(n.tax.col+n.pheno))]
    if(n.pheno!=1){
      col.sums <- list(1-(unname(colSums(plot.table[, c((n.tax.col+1):(n.tax.col+n.pheno))]))))
    }else{
      col.sums <- list(1-(unname(sum(plot.table[, c(n.tax.col+1)]))))
    }
    
    print(plot.table)
    
    n.row <- nrow(plot.table)
    plot.table[(n.row+1), (n.tax.col)] <- c("Other")
    plot.table[(n.row+1), c((n.tax.col+1):(n.tax.col+n.pheno))] <- col.sums[[1]]
    for (i in 1:n.pheno) {
      colnames(plot.table)[c(n.tax.col+i)] <- unlist(strsplit(colnames(plot.table)[c(n.tax.col+i)], " ", fixed=TRUE))[1]
    }
    
    print(sorted)
    plot.table[, n.tax.col]=factor(plot.table[, n.tax.col], levels=c("Other", sorted))
    plot.table <- plot.table[, c(1:(n.tax.col+n.pheno))]
    if(n.pheno!=1){
      plot.table <- pivot_longer(as.data.frame(plot.table), (n.tax.col+1):(n.tax.col+n.pheno), 
                               names_to=pheno.group, values_to="Abundance")
    }else{
      plot.table <- pivot_longer(as.data.frame(plot.table), (n.tax.col+1), names_to=pheno.group, values_to="Abundance")
    }
    plot.table <- plot.table[, -c(1:(n.tax.col-1))]
    colnames(plot.table)[1] <- taxa.level
    return(plot.table)
  }
}
    
```

## Plot Top Bar
```{r Plot top bar}
plot.top.bar <- function(phyloseq, taxa.level, n=10, pheno.group, group.by.pheno=NULL, 
                         path=NULL, phyloseq.name, group.title, top=NA, return.frame=FALSE, colors=NA){
  if(is.na(colors)){
    colors <- color.list(n+1)
  } 
  

  if(pheno.group=="family"){
    families <- identify.families(phyloseq)
  }
  
  plot.bar <- data.frame(suppressWarnings(plot.table.prep(phyloseq, taxa.level, n, pheno.group, group.by.pheno, top)))
  
  print(plot.bar)
  
  if(phyloseq.name=="family.phyloseq"){
    if(length(uuln(sample.to.data.frame(family.phyloseq.genus)[,"type"]))==7){
      plot.bar[, "type"] <- factor(plot.bar[, "type"], levels=c("BNA", "BSA", "BST", "MNA", "MSA", "MBM", "MAS"))
    }
  }else if(phyloseq.name=="Control"){
    plot.bar[, "samID"] <- factor(plot.bar[, "samID"], levels=c("02WCNT_I", "09WCNT_I", "14WCNT_I", "08WBNA_I", 
                                                                "06WBNA_I", "44WBNA_I", "35WBNA_I", "02WBNA_I", 
                                                                "39WBNA_I", "34WBNA_I", "21WBNA_I", "31WBNA_I", "43WBNA_I"))
  }
  
  print(plot.bar)
  
  if.asv <- if.asv(taxa.level)
  taxa.level <- if.asv[1]
  asv <- if.asv[2]
  
  #   FileName <- paste(paste0(path, "/table"), phyloseq.name, tolower(taxa.level), "top", n, tolower(pheno.group), sep=".")
  # if(!is.null(group.by.pheno)){
  #   FileName <- paste(FileName, group.by.pheno, sep=".")
  # }
  # FileName <- paste(FileName, "png", sep=".")
  # kbl.tbl <- plot.bar
  # kbl.tbl <- kbl.tbl[order(kbl.tbl[, 2]), ]
  # kibble <- kbl.tbl[, c(1, 3)] %>%
  #   kbl(centering=TRUE, escape=FALSE, col.names=c("Species", "Relative Abundance"), align="c") %>%
  #   kable_styling(font_size=15) %>%
  #   kable_classic("striped", full_width=F, html_font="Calibri", position="center") %>%
  #   row_spec(0, bold=TRUE) 
  # if(length(uln(unique(kbl.tbl[, pheno.group])))>1){
  #   kibble <- kibble %>%
  #     pack_rows(index=table(kbl.tbl[, pheno.group]), bold=FALSE, label_row_css="border-bottom: 0px solid;") 
  # }
  # kibble %>%
  #   save_kable(file=FileName, zoom=3)
  
  # print(data.frame(sample_data(phyloseq))[, "birth_type"])

  # plot.bar.title <- paste(group.title, str_to_title(pheno.group), "Top", n, "Bar Plot for", group.by.pheno)
  plot.bar.title <- paste(group.title, "\nTop", n, "Bar Plot")
  bar.plot <- ggplot(plot.bar, aes(fill=get(taxa.level), y=Abundance, x=get(pheno.group))) +
                theme_classic() +
                geom_bar(position="stack", color="black", stat="identity") +
                # theme(text=element_text(size=30), axis.text.x=element_text(angle=45, vjust=1, hjust=1)) +
                # xlab(str_to_title(pheno.group))+
                # xlab("Infant ID")+
                ylab("Relative Abundance")+
                guides(fill=guide_legend(title=taxa.level, ncol=2)) #+
                # theme(axis.title.x=element_blank(), axis.line.x.bottom=element_blank(), 
                #       axis.text.x=element_text(color="black", vjust=+8), 
                #       axis.ticks.x=element_blank(), axis.line.y.left=element_blank())
  
  if(str_detect(pheno.group, "_")==FALSE){
      bar.plot <- bar.plot + xlab(str_to_title(pheno.group))
  }else{
    bar.plot <- bar.plot + xlab(str_to_title(str_replace(pheno.group, "_", " ")))
  }
  
  if(pheno.group=="family"){
    if(n==20){
      bar.plot <- bar.plot +
        scale_x_discrete(limits=families, expand=c(0.5, 0)) +
        # scale_x_discrete(limits=families, guide=guide_axis(n.dodge=2)) +
        scale_fill_manual(values=colors) +
        # guides(fill=guide_legend(byrow=FALSE)) +
        coord_cartesian(ylim=c(0, 1), expand=0) 
    }else{
      bar.plot <- bar.plot +
        scale_x_discrete(limits=families, expand=c(0.5, 0)) +
        scale_fill_manual(values=colors) +
        coord_cartesian(ylim=c(0, 1), expand=0)
    }
  }else{
    if(n==20){
      bar.plot <- bar.plot +
        scale_fill_manual(values=colors) +
        scale_x_discrete(expand=c(0.5, 0)) +
        # scale_x_discrete(expand=c(0.15, 0.15)) +
        coord_cartesian(ylim=c(0, 1), expand=0)
    }else{
      bar.plot <- bar.plot +
        scale_fill_manual(values=colors) +
        # scale_x_discrete(expand=c(0, 0)) +
        scale_x_discrete(expand=c(0.5, 0)) +
        coord_cartesian(ylim=c(0, 1), expand=0)
    }
  }
  
  if(asv==TRUE){
    taxa.level <- "ASV"
  }
  
  if(is.null(path)){
    return(bar.plot)
    # if(return.frame==FALSE){
    #   return(bar.plot)
    # }else if(return.frame==TRUE){
    #   return(list(plot.bar, bar.plot))
    # }
  }
  
  
  if (!is.null(path)){
    if(n!=0){
      FileName <- paste0(path, "/barplot.", phyloseq.name, ".", tolower(taxa.level), ".top.", n, ".", tolower(pheno.group))
    }else{
      FileName <- paste0(path, "/barplot.", phyloseq.name, ".", tolower(taxa.level), ".all.taxa.", tolower(pheno.group))
    }
  }else {
    paste0("/barplot.", phyloseq.name, ".", tolower(taxa.level), ".top.", n, ".", tolower(pheno.group))
  }
  if(!is.null(group.by.pheno)){
    FileName <- paste0(FileName, ".", group.by.pheno)
  }
  FileName <- paste0(FileName, ".png")
  # width <- 10
  if(group.title=="Family Sequencing Infant Nasal"){
    width <- 10
  }else{
    width <- 8
  }
  if(taxa.level=="Species"||taxa.level=="ASV"){
    width <- width+3
  }
  if(pheno.group=="family"){
    # width <- width+(0.3*length(families))
    width <- width+(0.15*length(families))
  }
  if(pheno.group=="samID"){
    width <- width+(0.3*length(uuln(sample_data(phyloseq)[, pheno.group])))
  }
  if(pheno.group=="type"){
    # width <- width+(0.3*length(uuln(sample_data(phyloseq)[, pheno.group])))
    width <- width-3
  }else{
    width <- width+(0.3*length(uuln(sample_data(phyloseq)[, pheno.group])))
  }

  # print(data.frame(sample_data(phyloseq)) %>% dplyr::group_by_(pheno.group) %>% dplyr::summarise(number=n()))
  # print(data.frame(sample_data(phyloseq)) %>% group_by_(pheno.group) %>% dplyr::count())
  
  # print(table(data.frame(sample_data(phyloseq))[, pheno.group]))
  # cat(paste0("\n  ", table(data.frame(sample_data(phyloseq))[, pheno.group])))
  
  if(pheno.group=="Sample"){
    width <- width+3
    bar.plot <- bar.plot + theme(#axis.title.x=element_blank(), 
                                 axis.line.x.bottom=element_blank(), 
                      axis.text.x=element_text(#angle=45, 
                                               color="black", vjust=3), #, hjust=1.5
                      axis.line.y.left=element_blank(), axis.ticks.length=unit(0, "cm"))
  }else if(pheno.group=="family"){
    # bar.plot <- bar.plot + theme(#axis.title.x=element_blank(), 
    #                              axis.line.x.bottom=element_blank(), 
    #                   axis.text.x=element_text(angle=45, hjust=1, color="black"), #, vjust=3), #hjust=-0.5, 
    #                   axis.ticks.x=element_blank(), axis.line.y.left=element_blank())#, axis.ticks.length=unit(0, "cm"))
    bar.plot <- bar.plot + theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #panel.spacing.x=unit(0, "line"), 
          legend.title.align=0.5, legend.spacing.x=unit(8, "pt"), legend.box.spacing=unit(0, "pt"), 
          panel.grid.major=element_line(linetype="dashed", color="gray95"))
  }else if(pheno.group=="type"){
    width <- width+3
    bar.plot <- bar.plot + theme(#axis.title.x=element_blank(), 
                                 axis.line.x.bottom=element_blank(), 
                      axis.text.x=element_text(#angle=45, 
                                               color="black", vjust=3), #, hjust=1.5
                      axis.line.y.left=element_blank(), axis.ticks.length=unit(0, "cm"))
  }else{
    bar.plot <- bar.plot + theme(#axis.title.x=element_blank(), 
                                 axis.line.x.bottom=element_blank(), 
                      axis.text.x=element_text(color="black", 
                                               angle=45, hjust=1, vjust=1.1
                                               # vjust=3, 
                                               ), #, hjust=1), #vjust=3), #hjust=-0.5, 
                      axis.ticks.x=element_blank(), axis.line.y.left=element_blank(), axis.ticks.length=unit(0, "cm")) #+
      # geom_label(inherit.aes=FALSE, data=data.frame(sample_data(phyloseq)) %>% group_by(get(pheno.group)) %>% dplyr::count(), 
      #       aes(label=paste0("n=", n), x=pheno.group), y=0.9)
      # stat_summary(size=2, fun.data=n_fun, geom="text", data=data.frame(sample_data(phyloseq))[, "birth_type"])
  }
  
  # bar.plot <- bar.plot + theme(axis.title.x=element_blank(), axis.line.x.bottom=element_blank(),
  #                              axis.line.y.left=element_blank(), axis.ticks.length=unit(0, "cm"), 
  #                              plot.background=element_rect(fill="white", color="white"),
  #                              legend.background=element_blank(), panel.background=element_rect(fill="white", color="white"),
  #                              legend.spacing.x=unit(10, "pt"), 
  #                              plot.title=element_blank(), axis.text.x=element_blank(), legend.box.margin=margin(0,10,0,0)) 
  
  bar.plot <- annotate_figure(bar.plot, top=text_grob(label=plot.bar.title)) +
    bgcolor("white") + border(color="white")
  # ggsave(bar.plot, file=FileName, type="cairo", width=width, height=6, scaling=1.5, limitsize=FALSE)
  # ggsave(bar.plot, file=FileName, type="cairo", width=width, height=6, scale=0.75, limitsize=FALSE)
  
  
  ggsave(bar.plot, file=FileName, type="cairo", width=width, height=(1.325 + 2.5), scale=0.75, limitsize=FALSE)
  # ggsave(bar.plot, file=FileName, type="cairo", width=width, height=(1.325 + 2.5 + 4), scale=0.75, limitsize=FALSE)
}
```

## Plot Top Pie

```{r Plot Pie}
plot.top.pie <- function(phyloseq, taxa.level, n=10, pheno.group, group.by.pheno=NULL, path=NULL, phyloseq.name, group.title, colors=NA){
  
  if(is.na(colors)){
    colors <- color.list(n)
  }
  
  plot.bar <- plot.table.prep(phyloseq, taxa.level, n, pheno.group, group.by.pheno, reads=TRUE)
  
  if.asv <- if.asv(taxa.level)
  taxa.level <- if.asv[1]
  asv <- if.asv[2]
  
  # plot.bar <- aggregate(Abundance ~ get(taxa.level), transform(plot.bar, get(taxa.level)), sum)
  # plot.bar <- aggregate(plot.bar[, "Abundance"], plot.bar[, taxa.level], FUN=sum)

  
  # plot.bar[, "Abundance"] <- plot.bar[, "Abundance"]/sum(plot.bar[, "Abundance"])
  
  if(pheno.group=="family"){
    families <- identify.families(phyloseq)
  }
  
  plot.bar2 <- plot.bar %>% 
  mutate(csum=rev(cumsum(rev(Abundance))), 
         pos=Abundance/2 + lead(csum, 1), 
         pos=if_else(is.na(pos), Abundance/2, pos))
  
  # print(plot.bar)
  # if(pheno.group=="family"){
    pie.plot <- ggplot(plot.bar, aes(fill=get(taxa.level), y=Abundance, x=2, label=get(taxa.level))) +
      theme_void() +
      geom_bar(color="black", stat="identity") +
      coord_polar("y", start=0) +
      scale_fill_manual(values=colors) +
      guides(fill=guide_legend(ncol=1)) +
      # xlim(0.5, 2.5) +
      # geom_label_repel(aes(label=get(taxa.level))) +
      # geom_label_repel(data=plot.bar2, aes(y=pos), 
                       # nudge_x=1, show.legend=FALSE) +
      theme(axis.title.x=element_blank(), axis.line.x.bottom=element_blank(), 
            axis.text.y=element_blank(), axis.title.y=element_blank(), axis.line.y.left=element_blank(), 
            legend.title=element_blank(), plot.background=element_rect(color="white", fill="white")) +
      bgcolor("white") + border(color="white")
      
      # guides(fill="none")
      # ggtitle(plot.bar.title)
    # print(pie.plot)
    
    # if(pheno.group=="family"){
    #   if(n==20){
    #     pie.plot <- pie.plot +
    #       scale_x_discrete(limits=families) +
    #       # scale_x_discrete(limits=families, guide=guide_axis(n.dodge=2)) +
    #       scale_fill_manual(values=colors)
    #   }else{
    #     pie.plot <- pie.plot +
    #       scale_x_discrete(limits=families)
    #   }
    # }else{
    #   if(n==20){
    #     pie.plot <- pie.plot +
    #       scale_fill_manual(values=colors)
    #   }
    # }
    
    # if(asv==TRUE){
    #   taxa.level <- "ASV"
    # }

    # if (!is.null(path)){
    #   if(n!=0){
    #     FileName <- paste0(path, "/pie.", phyloseq.name, ".", tolower(taxa.level), ".top.", n, ".", tolower(pheno.group))
    #   }else{
    #     FileName <- paste0(path, "/pie.", phyloseq.name, ".", tolower(taxa.level), ".all.taxa.", tolower(pheno.group))
    #   }
    # }else {
    #   paste0("/pie.", phyloseq.name, ".", tolower(taxa.level), ".top.", n, ".", tolower(pheno.group))
    # }
    # if(!is.null(group.by.pheno)){
    #   FileName <- paste0(FileName, ".", group.by.pheno)
    # }
    # FileName <- paste0(FileName, ".png")
    # if(pheno.group=="family"){
    #   ggsave(pie.plot, file=FileName, type="cairo", width=20, height=10)
    # }else{
    #   ggsave(pie.plot, file=FileName, type="cairo", width=15, height=10)
    # }
  # }
  return(list(pie.plot, plot.bar))
}
```

## Phyloseq call for barplot creation

```{r Phyloseq call for barplot creation}
phyloseq.barplot <- function(phyloseq, taxa.level, n, pheno.group, group.by.pheno, 
                             output.path, phyloseq.name, group.title){
  output.path <- paste0(output.path, "/", taxa.level)
  if(!is.null(group.by.pheno)){
    sample.types <- uuln(as.data.frame(sample_data(phyloseq))[, group.by.pheno])
    # cat(sample.types)
  }else if(is.null(group.by.pheno)){
    sample.types <- NULL
  }
  # if(pheno.group!="Sample"){
  #   if(pheno.group!="type"){
  #     phyloseq <- prune_samples(rownames(as.data.frame(sample_data(phyloseq))[!is.na(as.data.frame(sample_data(phyloseq))[, pheno.group]), ]), phyloseq)
  #     sample.types <- uuln(as.data.frame(sample_data(phyloseq))[, "type"])
  #   }else{
  #     sample.types <- NULL
  #   }
  # }
  if(!is.null(group.by.pheno)){
    for(i in 1:length(sample.types)) {
      # cat("\n", sample.types[i])
      suppressWarnings(plot.top.bar(phyloseq, taxa.level, n, pheno.group, sample.types[i], output.path, phyloseq.name, group.title))
    }
  }else{
    suppressWarnings(plot.top.bar(phyloseq, taxa.level, n, pheno.group, sample.types, output.path, phyloseq.name, group.title))
  }
}
```

# Sample Comparison

## Compare Taxa

```{r}
compare.selected.taxa <- function(phyloseq, taxa.level, pheno.group, phyloseq.name, col.number, row.number, specific.taxa=NULL, top.range=NULL, path, facet=TRUE){
  path <- paste0(path, "/Output/Taxa/", taxa.level)
  comb.list <- uln(unique(as.data.frame(sample_data(phyloseq))[, pheno.group]))
  comparisons <- as.list(as.data.frame(t(CombPairs(comb.list))))
  if(!is.null(specific.taxa)){
    top <- taxa.to.asv(phyloseq, specific.taxa, taxa.level)
    phyloseq <- trans.sam.counts(phyloseq, NULL, FALSE)
    # print(top)
  }else if(!is.null(top.range)){
    top <- top.taxa.n(phyloseq, taxa.level, 0, pheno.group)
    top <- top[top.range]
    phyloseq <- trans.sam.counts(phyloseq, NULL, FALSE)
  }
  for(a in 1:length(top)){
    if(a==1){
      table <- data.frame(sample.id=rownames(sample_data(phyloseq)), pheno.group=uln(sample_data(phyloseq)[, pheno.group]), 
                          asv=otu_table(phyloseq)[, top[a]])
    }else{
      add <- data.frame(sample.id=rownames(sample_data(phyloseq)), pheno.group=uln(sample_data(phyloseq)[, pheno.group]), 
                        asv=otu_table(phyloseq)[, top[a]])
      table <- merge.data.frame(table, add, by=c("sample.id", "pheno.group"))
    }
  }
  # for(a in 3:ncol(table)){
  #   colnames(table)[a] <- asv.to.taxa(phyloseq, colnames(table)[a], taxa.level)
  # }
  
  if(facet==FALSE){
    plot.list <- list()
    for(a in 1:length(top)){
      
      plot <- ggplot(table, aes(x=pheno.group, y=get(top[a])))+
        theme_classic() +
        geom_boxplot(outlier.size=-1) +
        geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.75, dodge.width=0.75), 
                 size=2.5, aes(fill=pheno.group)) +
        stat_compare_means(method="wilcox.test", comparisons=comparisons, aes(label=paste0("p=", ..p.format..), )) +
        ylim(0, max(table[, top[a]]) + (max(table[, top[a]])/12)) + ylab(paste("Relitive Abundance")) +
        theme(legend.position="none", axis.title.x=element_blank(), axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
              axis.text.x=element_text(color="black", face="bold"), axis.ticks.x=element_blank(), 
              panel.border=element_rect(color="black", fill=NA, size=0.75), plot.title=element_text(hjust=0.5, face="bold")) 
      if(taxa.level=="Species"){
        plot <- plot + labs(title=paste(asv.to.taxa(phyloseq, top[a], "Genus"), asv.to.taxa(phyloseq, top[a], "Species")))
      }else if(taxa.level=="Genus"){
        plot <- plot + labs(title=paste(asv.to.taxa(phyloseq, top[a], "Genus")))
      }
      print(plot)
      plot.list[[a]] <- plot
    }
    all.plot <- ggarrange(plotlist=plot.list, nrow=row.number, ncol=ceiling(length(plot.list)/row.number))
    print(all.plot)
    FileName <- paste0(path, "/compare.taxa.", phyloseq.name, ".", taxa.level, ".", pheno.group, ".png")
    ggsave(filename=FileName, plot=all.plot, width=(ceiling(length(plot.list)/row.number)*2), height=(row.number*2), scale=0.8)
  }else{
    names <- c()
    for(a in 1:length(top)){
      names[a] <- asv.to.taxa(phyloseq, top[a], taxa.level)
      if(a==1){
        table <- data.frame(sample.id=rownames(sample_data(phyloseq)), pheno.group=uln(sample_data(phyloseq)[, pheno.group]), 
                            taxa=asv.to.taxa(phyloseq, top[a], taxa.level), asv=otu_table(phyloseq)[, top[a]])
        colnames(table)[4] <- "asv"
      }else{
        add <- data.frame(sample.id=rownames(sample_data(phyloseq)), pheno.group=uln(sample_data(phyloseq)[, pheno.group]), 
                          taxa=asv.to.taxa(phyloseq, top[a], taxa.level), asv=otu_table(phyloseq)[, top[a]])
        colnames(add)[4] <- "asv"
        table <- rbind(table, add)
      }
    }
    table[, "taxa"] <- str_replace_all(table[, "taxa"], "Bacteria Unclassified", "Unclassified")
    names <- str_replace_all(names, "Bacteria Unclassified", "Unclassified")
    
    table <- remove.na(table, "pheno.group")
    print(table)
    comb.list <- uln(unique(table[, "pheno.group"]))
    comparisons <- as.list(as.data.frame(t(CombPairs(comb.list))))
  
    plot <- ggplot(table, aes(x=pheno.group, y=asv))+
      theme_classic() +
      geom_boxplot(outlier.size=-1) +
      geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.75, dodge.width=0.75), 
               size=2.5, aes(fill=pheno.group)) +
      stat_compare_means(method="wilcox.test", comparisons=comparisons, aes(label=paste0("p=", ..p.format..), )) +
      ylim(0, max(table[, "asv"]) + (max(table[, "asv"])/5)) +
      ylab(paste("Relitive Abundance")) +
      facet_wrap(~factor(taxa, levels=names), 
                 nrow=row.number, ncol=col.number) +
      theme(legend.position="none", axis.title.x=element_blank(), axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
            axis.text.x=element_text(color="black", face="bold"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), plot.title=element_text(hjust=0.5, face="bold"))
    print(plot)
  
  
    FileName <- paste0(path, "/compare.taxa.", phyloseq.name, ".", taxa.level, ".", pheno.group, ".png")
    ggsave(filename=FileName, plot=plot, width=(col.number*3), height=(row.number*2), scale=0.8)
  }
}
```

## Transform Sample Counts RA

```{r Transform sample counts}
trans.sam.counts <- function(phyloseq, pheno.group=NULL, merge=FALSE){
  phyloseq.top <- phyloseq
  if(merge==FALSE){
    phyloseq1 <- phyloseq.top
  }else if(length(rownames(as.data.frame(otu_table(phyloseq))))==1){
    phyloseq1 <- phyloseq.top
  }else{
    phyloseq.top <- transform_sample_counts(phyloseq, function(y) y/sum(y))
    otu_table(phyloseq.top) <- replace.nan(otu_table(phyloseq.top), 0)
    phyloseq1 <- merge_samples(phyloseq.top, pheno.group)
  }
  phyloseq.top <- transform_sample_counts(phyloseq1, function(y) y/sum(y))
  return(phyloseq.top)
}
```


## Transform Sample Counts Reads

```{r Transform sample counts}
trans.sam.counts.reads <- function(phyloseq, pheno.group=NULL, merge=FALSE){
  phyloseq.top <- phyloseq
  if(merge==FALSE){
    phyloseq1 <- phyloseq.top
  }else if(length(rownames(as.data.frame(otu_table(phyloseq))))==1){
    phyloseq1 <- phyloseq.top
  }else{
    phyloseq1 <- merge_samples(phyloseq.top, pheno.group)
  }
  phyloseq.top <- transform_sample_counts(phyloseq1, function(y) y/sum(y))
  return(phyloseq.top)
}
```

## Top taxa

```{r Top taxa}
top.taxa.n <- function(phyloseq, taxa.level, n, pheno.group=NULL, reads=FALSE){
  n.tax.col <- which(colnames(as.data.frame(tax_table(phyloseq)))==taxa.level)
  na.value <- row.names(tax_table(phyloseq)[c(!is.na(tax_table(phyloseq)[, c(n.tax.col)])), ])
  phyloseq <- prune_taxa(na.value, phyloseq)
  nan.value <- row.names(tax_table(phyloseq)[c(!is.nan(tax_table(phyloseq)[, c(n.tax.col)])), ])
  phyloseq <- prune_taxa(nan.value, phyloseq)
  otu_table(phyloseq) <- replace.nan(otu_table(phyloseq), 0)
  
  sample_data(phyloseq)[,"merge"] <- "yes"
  
  if(reads==TRUE){
    phyloseq <- trans.sam.counts.reads(phyloseq, "merge", TRUE)
  }else{
    phyloseq <- trans.sam.counts(phyloseq, "merge", TRUE)
  }
  
  # phyloseq <- trans.sam.counts(phyloseq, "merge", TRUE)
  # print(otu.tax(phyloseq))
  
  
  top <- names(sort((taxa_sums(phyloseq)[taxa_sums(phyloseq)!=0]), decreasing=TRUE))
  
  if(n!=0){
    if(length(top)>n){
      top <- top[1:n]
    }
  }
  return(top)
}
```

## Phyloseq processing for comparison

```{r Phyloseq processing for comparison}
phyloseq.processing.compare <- function(phyloseq, pheno.group, top.otu=NULL, merge=FALSE, reads=FALSE){
  if(reads==TRUE){
    phyloseq.top <- trans.sam.counts.reads(phyloseq, pheno.group, merge)
  }else{
    phyloseq.top <- trans.sam.counts(phyloseq, pheno.group, merge)
  }
  
  print(top.otu)
  if(!is.null(top.otu)){
    phyloseq.top <- prune_taxa(top.otu, phyloseq.top)
  }
  tax_table(phyloseq.top)[, colnames(tax_table(phyloseq.top))] <- gsub(tax_table(phyloseq.top)
                                                                       [, colnames(tax_table(phyloseq.top))], 
                                                                       pattern="[a-z]__", replacement="")
  return(phyloseq.top)
}
```

## Phyloseq object creation for comparison

```{r Phyloseq object creation for comparison}
ps.object.creation <- function(phyloseq, taxa.level, n, pheno.group, pheno, family){
  cat("\n  ", "Processing for family", family, pheno)
  
  ### Phyloseq object creation ###
  cat("\n     ", "Phyloseq object creation")
  if(family=="all"){
    phyloseq <- merge_samples(phyloseq, pheno.group)
    otu_table(phyloseq) <- otu_table(phyloseq)[pheno, ]
  }else{
    phyloseq.family.rows <- rownames(sample_data(phyloseq)[sample_data(phyloseq)[, "family"]==family, ])
    sample_data(phyloseq) <- sample_data(phyloseq)[phyloseq.family.rows, ]
    if(any(as.data.frame(sample_data(phyloseq))[, pheno.group]==pheno)==TRUE && !is.na(any(as.data.frame(sample_data(phyloseq))[, pheno.group]==pheno))){
      if(any(as.data.frame(sample_data(phyloseq))==pheno)==TRUE){
        phyloseq.rows <- rownames(sample_data(phyloseq)[sample_data(phyloseq)[, pheno.group]==pheno, ])
        phyloseq <- prune_samples(phyloseq.rows, phyloseq)
        phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
        assign(paste0("phyloseq.", pheno, ".", family), phyloseq, envir=phyloseq.env)
        assign(paste0(pheno, ".", family, ".na"), FALSE, envir=phyloseq.env)
        pheno.na <- "Good"
      }else{
        pheno.na <- NA
        assign(paste0(pheno, ".", family, ".na"), TRUE, envir=phyloseq.env)
      }
    }else{
      pheno.na <- NA
      assign(paste0(pheno, ".", family, ".na"), TRUE, envir=phyloseq.env)
    }
  }
  if(!is.na(pheno.na)){
    ### OTU identification by relative abundance ###
    cat("\n     ", "OTU identification by relative abundance")
    otu.top <- top.taxa.n(phyloseq, taxa.level, n, pheno.group)
    otu.top.names <- as.data.frame(tax_table(phyloseq))[otu.top, c(1:7)]
    otu.top.names[, "ASV"] <- rownames(otu.top.names)
    otu.top.names <- otu.top.names[, c(8, 1:7)]
    assign(paste0("otu.top.names.", pheno, ".", family), otu.top.names, envir=phyloseq.env)
  }
}
```

## OTU phyloseq

```{r OTU phyloseq}
otu.phyloseq <- function(pheno.group, taxa.level, pheno, family, otu.dup.names){
  phyloseq <- get(paste0("phyloseq.", pheno, ".", family), envir=phyloseq.env)
  otu.top.names <- get(paste0("otu.top.names.", pheno, ".", family), envir=phyloseq.env)
  otu.top <- rownames(otu.top.names[is.element(otu.top.names[, "ASV"], otu.dup.names), ])
  phyloseq <- phyloseq.processing.compare(phyloseq, pheno.group, otu.top, TRUE)
  assign(paste0("pruned.phyloseq.", pheno, ".", family), phyloseq, envir=phyloseq.env)
}
```

## Compare top taxa for two samples

```{r Compare top taxa for two samples}
compare.pheno.top.overlap.pheno.processing <- function(phyloseq, taxa.level, n, path=NULL, family, 
                                                       family.compare=NULL, pheno.group, pheno.a, pheno.b, 
                                                       phyloseq.name){
  if(exists(paste0("phyloseq.env"))){
    # rm(list=ls(), envir=phyloseq.env)
  }
  cat("\n\n", paste("Comparison of family", family, pheno.a, "and family", family.compare, pheno.b))
  
  ### Check existance of return objects ###
  if(exists(paste0("phyloseq.", pheno.a, ".", family), envir=phyloseq.env)==FALSE){
    ps.object.creation(phyloseq, taxa.level, n, pheno.group, pheno.a, family)
  }
  if(exists(paste0("phyloseq.", pheno.b, ".", family.compare), envir=phyloseq.env)==FALSE){
    ps.object.creation(phyloseq, taxa.level, n, pheno.group, pheno.b, family.compare)
  }
  
  if(get(paste0(pheno.a, ".", family, ".na"), envir=phyloseq.env)==FALSE && 
     get(paste0(pheno.b, ".", family.compare, ".na"), envir=phyloseq.env)==FALSE){
    if(family!="all"){
      if(n!=0){
        table.title <- paste("<center>Family", family, "Top", n, pheno.a, "and", 
                             pheno.b, "Shared", taxa.level, "</center>")
      }else{
        table.title <- paste("<center>Family", family, "All Taxa", pheno.a, "and", 
                             pheno.b, "Shared", taxa.level, "</center>")
      }
      
    } else {
      table.title <- paste("<center>All Families Top", n, pheno.a, "and", 
                           pheno.b, "Shared", taxa.level, "</center>")
    }
    ### Shared taxa.level names identification ###
    cat("\n  ", "Shared taxa.level names identification")
    otu.top.both.names <- rbind(get(paste0("otu.top.names.", pheno.a, ".", family), envir=phyloseq.env), 
                                           get(paste0("otu.top.names.", pheno.b, ".", family.compare), 
                                           envir=phyloseq.env))
    otu.top.both.names <- otu.top.both.names[duplicated(otu.top.both.names[, "ASV"]), ]

    # print(otu.top.both.names)
    otu.dup.names <- otu.top.both.names[, "ASV"]
    cat("\n     Shared", taxa.level, "\n      ", paste0(otu.dup.names))
    
    
    if(length(otu.dup.names)>0){
      ### OTU identification by shared taxa.level names ###
      cat("\n  ", "OTU identification by shared taxa.level names")
      otu.phyloseq(pheno.group, taxa.level, pheno.a, family, otu.dup.names)
      otu.phyloseq(pheno.group, taxa.level, pheno.b, family.compare, otu.dup.names)
      
      ### Merged phyloseq creation ###
      cat("\n  ", "Merged phyloseq creation")
      phyloseq.dups <- (merge_phyloseq(get(paste0("pruned.phyloseq.", pheno.a, ".", family), envir=phyloseq.env), 
                                       get(paste0("pruned.phyloseq.", pheno.b, ".", family.compare), envir=phyloseq.env)))
      
      top.table.return <- top.table(phyloseq.dups, pheno.group, taxa.level, NULL, n=0, FALSE)
      table.dups <- top.table.return[1]
      percentage.a <- unlist(top.table.return[2])
      cat("\n     ", "Family", family, pheno.a, paste0(percentage.a, "%"))
      percentage.b <- unlist(top.table.return[3])
      cat("\n     ", "Family", family.compare, pheno.b, paste0(percentage.b, "%"))
      
      if(exists(paste0("family.comparison.", family), envir=phyloseq.env)==TRUE){
        assign(paste0("family.comparison.", family), add_row(sample.a.type=pheno.a, sample.a.fam=as.character(family), 
                                                            sample.a.percentage=as.numeric(percentage.a), 
                                                            sample.b.type=pheno.b, 
                                                            sample.b.fam=as.character(family.compare), 
                                                            sample.b.percentage=as.numeric(percentage.b), 
                                                            shared.taxa=length(otu.dup.names),
                                                            table=table.dups, get(paste0("family.comparison.", family), 
                                                                                  envir=phyloseq.env)), 
               envir=phyloseq.env)
      }else{
        assign(paste0("family.comparison.", family), data.frame(sample.a.type="Sample", sample.a.fam="2", 
                                                                sample.a.percentage=79, sample.b.type="Sample", 
                                                                sample.b.fam="8", sample.b.percentage=58, shared.taxa=1,
                                                                nest(table=everything(), tibble(x=c(9, 0), y=c(0, 9)))), 
               envir=phyloseq.env)
        assign(paste0("family.comparison.", family), add_row(sample.a.type=pheno.a, sample.a.fam=as.character(family), 
                                                            sample.a.percentage=as.numeric(percentage.a), 
                                                            sample.b.type=pheno.b, 
                                                            sample.b.fam=as.character(family.compare), 
                                                            sample.b.percentage=as.numeric(percentage.b), 
                                                            shared.taxa=length(otu.dup.names),
                                                            table=table.dups, get(paste0("family.comparison.", family), 
                                                                                  envir=phyloseq.env)), 
               envir=phyloseq.env)
        assign(paste0("family.comparison.", family), get(paste0("family.comparison.", family), 
                                                         envir=phyloseq.env)[-c(1), ], 
               envir=phyloseq.env)
      }
      
      if(is.null(family.compare)){
        assign.name <- paste0(phyloseq.name, ".top.", n, ".", pheno.a, ".", pheno.b, ".fam.", family)
      } else{
        if(n!=0){
          assign.name <- paste0(phyloseq.name, ".top.", n, ".", pheno.a, ".fam.", family, ".", 
                                pheno.b, ".fam.", family.compare)
        }else{
          assign.name <- paste0(phyloseq.name, ".all.taxa.", pheno.a, ".fam.", family, ".", 
                                pheno.b, ".fam.", family.compare)
        }
      }
      
    }else{
      if(exists(paste0("family.comparison.", family), envir=phyloseq.env)==TRUE){
        assign(paste0("family.comparison.", family), add_row(sample.a.type=pheno.a, sample.a.fam=as.character(family), 
                                                            sample.a.percentage=0, sample.b.type=pheno.b, 
                                                            sample.b.fam=as.character(family.compare), 
                                                            sample.b.percentage=0, 
                                                            shared.taxa=0, table=NA, 
                                                            get(paste0("family.comparison.", family), 
                                                                envir=phyloseq.env)), 
               envir=phyloseq.env)
      }else{
        assign(paste0("family.comparison.", family), data.frame(sample.a.type=pheno.a, sample.a.fam=as.character(family), 
                                                               sample.a.percentage=0, sample.b.type=pheno.b, 
                                                               sample.b.fam=as.character(family.compare), 
                                                               sample.b.percentage=0, shared.taxa=0,
                                                               nest(table=everything(), tibble(NA))), 
               envir=phyloseq.env)
      }
    }
  }else{
    cat("\n  ", "Second phenotype does not exist in sample")
  }
}
```

## Compare taxa loop

```{r Compare taxa loop}
phyloseq.taxa <- function(phyloseq, taxa.level, n, pheno.group, sample.set.1, sample.set.2, path, phyloseq.name){
  path <- paste0(path, "/Output/Correlation/Comparisons")
  if(exists(paste0("phyloseq.env"))){
    rm(list=ls(), envir=phyloseq.env)
  }
  assign(paste0("phyloseq.env"), new.env(), envir=.GlobalEnv)
  families <- as.character(unique(unlist(as.data.frame(sample_data(phyloseq))[, "family"])))
  for(i in 1:length(families)){
  # for(i in 34:50){
    if.asv <- if.asv(taxa.level)
    taxa.level <- if.asv[1]
    asv <- if.asv[2]
    for(j in 1:length(families)) {
      for (k in 1:length(sample.set.1)) {
        for(l in 1:length(sample.set.2)) {
          if(families[i]==families[j]&&sample.set.1[k]==sample.set.2[l]){
          }else{
            compare.pheno.top.overlap.pheno.processing(phyloseq=phyloseq, taxa.level=taxa.level, 
                                                       n=n, path=path, family=families[i], family.compare=families[j], 
                                                       pheno.group=pheno.group, pheno.a=sample.set.1[k], 
                                                       pheno.b=sample.set.2[l], phyloseq.name)
          }
        }
      }
    }
    if(asv==TRUE){
      taxa.level <- "ASV"

    }
    if(exists(paste0("family.comparison.", families[i]), envir=phyloseq.env)==TRUE){
      if(n!=0){
        saveRDS(get(paste0("family.comparison.", families[i]), envir=phyloseq.env), 
                paste0(path, "/top.", n, ".", phyloseq.name, ".", tolower(taxa.level), ".comparison.", 
                       sample.set.1, ".family.", families[i], ".rds"))
      }else{
        saveRDS(get(paste0("family.comparison.", families[i]), envir=phyloseq.env), 
                paste0(path, "/all.taxa.", phyloseq.name, ".", tolower(taxa.level), ".comparison.", 
                       sample.set.1, ".family.", families[i], ".rds")) 
      }
      save(phyloseq.env, file=file.path(paste0(path, "/", phyloseq.name, ".", tolower(taxa.level), ".comparison.env.RData")))
    }
  }
}
```


## Compare taxa loop group

```{r Compare taxa loop}
phyloseq.taxa.group <- function(phyloseq, groups, taxa.level, n, pheno.group, sample.set.1, sample.set.2, path, phyloseq.name){
  path <- paste0(path, "/Output/Correlation/Comparisons")
  if(exists(paste0("phyloseq.env"))){
    rm(list=ls(), envir=phyloseq.env)
  }
  assign(paste0("phyloseq.env"), new.env(), envir=.GlobalEnv)
  for(h in 1:length(groups)){
    phyloseq.sub <- subset.object(phyloseq, groups[[h]], "family")
    families <- as.character(unique(unlist(as.data.frame(sample_data(phyloseq.sub))[, "family"])))
    print(families)
    for(i in 1:length(families)){
    # for(i in 34:50){
      if.asv <- if.asv(taxa.level)
      taxa.level <- if.asv[1]
      asv <- if.asv[2]
      for(j in 1:length(families)) {
        for (k in 1:length(sample.set.1)) {
          for(l in 1:length(sample.set.2)) {
            if(families[i]==families[j]&&sample.set.1[k]==sample.set.2[l]){
            }else{
              compare.pheno.top.overlap.pheno.processing(phyloseq=phyloseq.sub, taxa.level=taxa.level, 
                                                         n=n, path=path, family=families[i], family.compare=families[j], 
                                                         pheno.group=pheno.group, pheno.a=sample.set.1[k], 
                                                         pheno.b=sample.set.2[l], phyloseq.name)
            }
          }
        }
      }
      if(asv==TRUE){
        taxa.level <- "ASV"
  
      }
      if(exists(paste0("family.comparison.", families[i]), envir=phyloseq.env)==TRUE){
        if(n!=0){
          saveRDS(get(paste0("family.comparison.", families[i]), envir=phyloseq.env), 
                  paste0(path, "/top.", n, ".", phyloseq.name, ".", tolower(taxa.level), ".comparison.", 
                         sample.set.1, ".family.", families[i], ".rds"))
        }else{
          saveRDS(get(paste0("family.comparison.", families[i]), envir=phyloseq.env), 
                  paste0(path, "/all.taxa.", phyloseq.name, ".", tolower(taxa.level), ".comparison.", 
                         sample.set.1, ".family.", families[i], ".rds")) 
        }
        save(phyloseq.env, file=file.path(paste0(path, "/", phyloseq.name, ".", tolower(taxa.level), ".comparison.env.RData")))
      }
    }
  }
}
```

# Family Comparison

## Table

```{r Family Comparison Table}
family.comparison.table <- function(phyloseq, taxa.level, n, sample.set.1, output.path, phyloseq.name){
  families <- identify.families(phyloseq)
  for (i in 1:length(families)) {
      if(n!=0){
        if(file.exists(paste0(output.path, "/Comparisons/top.", n, ".", phyloseq.name, ".", tolower(taxa.level), ".comparison.", sample.set.1, ".family.", 
                              families[i], ".rds"))){
          assign(paste0("family.comparison.", families[i]), 
                 readRDS(paste0(output.path, "/Comparisons/top.", n, ".", phyloseq.name, ".", tolower(taxa.level), ".comparison.", sample.set.1, ".family.", 
                                families[i], ".rds")))
        }
      }else{
        if(file.exists(paste0(output.path, "/Comparisons/all.taxa.", phyloseq.name, ".", tolower(taxa.level), ".comparison.", sample.set.1, ".family.", 
                              families[i], ".rds"))){
          assign(paste0("family.comparison.", families[i]), 
                 readRDS(paste0(output.path, "/Comparisons/all.taxa.", phyloseq.name, ".", tolower(taxa.level), ".comparison.", sample.set.1, ".family.", 
                                families[i], ".rds")))
        }
      }
    }
    
    for (i in 1:length(families)) {
      if(exists(paste0("family.comparison.", families[i]))){
        assign("merged.family.comparison", get(paste0("family.comparison.", families[i]))[, c(1:8)])
        m=(i+1)
        break()
      }
    }
    ### Dataframe preparations ###
          cat("\n  ", "Dataframe preparations")
    for (i in m:length(families)) {
      if(exists(paste0("family.comparison.", families[i]))){
        merged.family.comparison <- rbind(merged.family.comparison, get(paste0("family.comparison.", families[i]))[, 1:8])
      }
    }
    for(i in 1:nrow(merged.family.comparison)){
      merged.family.comparison[i, "merged.name"] <- paste(merged.family.comparison[i, 1], merged.family.comparison[i, 2], 
                                                         merged.family.comparison[i, 4], merged.family.comparison[i, 5])
      merged.family.comparison[i, "both.fam"] <- paste(merged.family.comparison[i, 2], merged.family.comparison[i, 5])
      if(merged.family.comparison[i, 2]==merged.family.comparison[i, 5]){
        merged.family.comparison[i, "family.type"] <- "Same family"
      } else{
        merged.family.comparison[i, "family.type"] <- "Different family"
      }
    }
    merged.family.comparison <- merged.family.comparison[, c(1:7, 11, 8)]
    merged.family.comparison[, "sample.a.fam"] <- factor(as.numeric(merged.family.comparison[, "sample.a.fam"]), 
                                                  levels=families)
    merged.family.comparison[, "sample.b.fam"] <- factor(as.numeric(merged.family.comparison[, "sample.b.fam"]), 
                                                  levels=families)
    return(merged.family.comparison)
}
```

## Prevalence

```{r Family Comparison Prevalence}
family.comparison.prevalence <- function(phyloseq, dataframe, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name){
  dir.create(paste0(output.path, "/Prevalence"))
  if.asv.lvl <- if.asv(taxa.level)
  taxa.level <- if.asv.lvl[1]
  asv <- if.asv.lvl[2]
  phyloseq <- trans.sam.counts(phyloseq, NULL, FALSE)
  dataframe <- subset.object(dataframe, "Same family")
  
  # Make a list of the ASV that exists in the shared family 
  asv.list <- c()
  for(a in 1:nrow(dataframe)){
    if(dataframe[a, "sample.a.percentage"]!=0){
      asv.list <- c(asv.list, rownames(dataframe[a, "table"][[1]])[1:(nrow(dataframe[a, "table"][[1]])-1)])
    }
  }
  asv.list <- unique(asv.list)
  same.fam.counts <- c(rep(0, length(asv.list)))
  sam.a.abundance <- c(rep(0, length(asv.list)))
  sam.b.abundance <- c(rep(0, length(asv.list)))
  for(a in 1:length(asv.list)){
    for(b in 1:nrow(dataframe)){
      if(dataframe[b, "sample.a.percentage"]!=0){
        if(asv.list[a] %in% rownames(dataframe[b, "table"][[1]])[1:(nrow(dataframe[b, "table"][[1]])-1)]){
          
          # Calculate prevelance in family pairs
          same.fam.counts[a] <- same.fam.counts[a]+1
          sam.a.abundance[a] <- sam.a.abundance[a] + (dataframe[b, "table"][[1]][asv.list[a], which(colnames(dataframe[b, "table"][[1]])==taxa.level)+1])
          sam.b.abundance[a] <- sam.b.abundance[a] + (dataframe[b, "table"][[1]][asv.list[a], which(colnames(dataframe[b, "table"][[1]])==taxa.level)+2])
        }
      }
    }
  }
  
  # sam.a.abundance[a] <- signif(mean(otu_table(subset.phyloseq.1)[, asv.list[a]]), 3)
  # sam.a.abundance[a] <- signif((colSums(otu_table(subset.phyloseq.1))[asv.list[a]])/sam.a.counts[a], 3)
  # sam.b.abundance[a] <- signif(mean(otu_table(subset.phyloseq.2)[, asv.list[a]]), 3)
  # sam.b.abundance[a] <- signif((colSums(otu_table(subset.phyloseq.2))[asv.list[a]])/sam.b.counts[a], 3)
  # Abundances are all non-zero abundances over the prevelance of the taxa
  
  sam.a.counts <- c()
  sam.b.counts <- c()
  
  subset.phyloseq.1 <- subset.object(phyloseq, sample.set.1, pheno.group)
  subset.phyloseq.2 <- subset.object(phyloseq, sample.set.2, pheno.group)
  
  sam.a.abundance.average <- c(rep(0, length(asv.list)))
  sam.b.abundance.average <- c(rep(0, length(asv.list)))
  
  for(a in 1:length(asv.list)){
    sam.a.counts[a] <- colSums(otu_table(subset.phyloseq.1)!=0)[asv.list[a]]
    sam.b.counts[a] <- colSums(otu_table(subset.phyloseq.2)!=0)[asv.list[a]]
    if(same.fam.counts[a]>1){
      sam.a.abundance.average[a] <- (sam.a.abundance[a]/same.fam.counts[a])
      sam.b.abundance.average[a] <- (sam.b.abundance[a]/same.fam.counts[a])
    }
  }
  
  return.dataframe <- data.frame(`ASV`=asv.list, `Prevalence.in.Family.Pairs`=same.fam.counts, 
                                 `Prevalence.in.a`=sam.a.counts, `Abundance.in.a`=sam.a.abundance, `Average.Abundance.in.a`=sam.a.abundance.average, 
                                 `Prevalence.in.b`=sam.b.counts, `Abundance.in.b`=sam.b.abundance, `Average.Abundance.in.b`=sam.b.abundance.average)

  colnames(return.dataframe) <- c("ASV", "Family Pair Prevalence", 
                                  paste(sample.set.1, "Prevalence"), paste(sample.set.1, "Total Abundance"), paste(sample.set.1, "Average Abundance"), 
                                  paste(sample.set.2, "Prevalence"), paste(sample.set.2, "Total Abundance"),  paste(sample.set.2, "Average Abundance"))
  
  asv.table <- as.data.frame(tax_table(phyloseq))
  asv.table[, "ASV"] <- rownames(asv.table)
  asv.table <- asv.table[, -c(1:5)]
  return.dataframe <- merge(asv.table, return.dataframe, by="ASV")
  
  # return.dataframe %>%
  #   kbl(centering=TRUE, escape=FALSE, align="c") %>%
  #       kable_styling(font_size=15) %>%
  #       kable_classic("striped", full_width=F, html_font="Calibri", position="center") %>%
  #       row_spec(0, bold=TRUE) %>%
  #       save_kable(file=paste(paste0(output.path, "/prevalence.table"), phyloseq.name, "png", sep="."), zoom=5)
  
  # print(return.dataframe)
  return.dataframe <- return.dataframe[order(-(return.dataframe[, "Family Pair Prevalence"])), ]
  # print(return.dataframe)
  
  write_csv(return.dataframe, file=paste(paste0(output.path, "/Prevalence/prevalence.table"), phyloseq.name, sample.set.1, sample.set.2, "csv", sep="."))
  return(return.dataframe)
}
```

## Distance Table

```{r Family Comparison Distance Table}
family.comparison.distance.table <- function(phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name){
  families <- identify.families(phyloseq)
  filename <- paste0(output.path, "/distance.table.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".csv")
  if(file.exists(filename)==FALSE){
    phyloseq.sam.1 <- subset.object(phyloseq, sample.set.1, pheno.group)
    phyloseq.sam.2 <- subset.object(phyloseq, sample.set.2, pheno.group)
    for(a in 1:length(families)){
      if(families[a] %in% data.frame(sample_data(phyloseq.sam.1))[, "family"]){
        # print(families[a])
        if(exists(paste("phyloseq", sample.set.1, families[a], sep="."))==FALSE){
          assign(paste("phyloseq", sample.set.1, families[a], sep="."), subset.object(phyloseq.sam.1, families[a], "family"))
        }
        phyloseq.sam.1.fam <- get(paste("phyloseq", sample.set.1, families[a], sep="."))
        for(b in 1:length(families)){
          if(families[b] %in% data.frame(sample_data(phyloseq.sam.2))[, "family"]){
            if(exists(paste("phyloseq", sample.set.2, families[b], sep="."))==FALSE){
              assign(paste("phyloseq", sample.set.2, families[b], sep="."), subset.object(phyloseq.sam.2, families[b], "family"))
            }
            phyloseq.sam.2.fam <- get(paste("phyloseq", sample.set.2, families[b], sep="."))
            merged.phyloseq <- merge_phyloseq(phyloseq.sam.1.fam, phyloseq.sam.2.fam)
            bdist <- phyloseq::distance(merged.phyloseq, "bray")
            if(families[a]==families[b]){
              if(exists("dataframe")==FALSE){
                dataframe <- data.frame(sample.a=sample.set.1, family.a=(families[a]), sample.b=sample.set.2, family.b=families[b], 
                                        family.type="same", distance=bdist[1])
              }else{
                dataframe <- add_row(sample.a=sample.set.1, family.a=families[a], sample.b=sample.set.2, family.b=families[b], 
                                     family.type="same", distance=bdist[1], dataframe)
              }
            }else{
              if(exists("dataframe")==FALSE){
                dataframe <- data.frame(sample.a=sample.set.1, family.a=families[a], sample.b=sample.set.2, family.b=families[b], 
                                        family.type="different", distance=bdist[1])
              }else{
                dataframe <- add_row(sample.a=sample.set.1, family.a=families[a], sample.b=sample.set.2, family.b=families[b], 
                                     family.type="different", distance=bdist[1], dataframe)
              }
            }
          }
        }
      }
    }
    write_csv(dataframe, filename)
  }else{
    dataframe <- data.frame(read_csv(filename))
  }
  return(dataframe)
}
```

## Distance Boxplot

```{r Family Comparison Distance Boxplot}
family.comparison.distance.boxplot <- function(phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name){
  
  
  if(dir.exists( paste0(output.path, "/Distance"))==FALSE){
    dir.create(paste0(output.path, "/Distance"))
  }
  output.path <- paste0(output.path, "/Distance")

  dataframe <- family.comparison.distance.table(phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name)
  
  pwc <- dataframe %>%
      group_by(sample.b) %>%
      t_test(distance ~ family.type) %>%
      adjust_pvalue(method="bonferroni")
  
  sample.b.list <- sample.set.2
  y.value=list()
  for(l in 1:length(sample.b.list)){
    max <- max(subset.object(dataframe, sample.b.list[l], "sample.b")[, "distance"])
    y.value[l] <- max+(max/12)
  }

  pwc[, c("y.position")] <- unlist(y.value)
  level <- subset.object(dataframe, "same", "family.type")[order(subset.object(dataframe, "same")[, "distance"], decreasing=TRUE), "family.a"]
  colors <- color.list.grad(length(level))

  dataframe[, "family.a"] <- factor(dataframe[, "family.a"], levels=unique(level))
  
  
        cat("\n  ", "Plot generation for all families by mother sample")
  plot <- ggplot(dataframe, aes(x=family.type, y=distance)) +
    geom_boxplot(outlier.size=-1) +
    ylim(0, y.value[[1]]+0.075) +
    geom_point(position=position_jitterdodge(seed=1L), alpha=1, size=1.5, aes(color=family.a)) +
    scale_color_manual(values=colors) +
    theme(strip.placement="bottom", strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
          legend.key.size=unit(0.25, "cm"), legend.key=element_blank()) +
    stat_pvalue_manual(pwc, label="p={p}") +
    xlab("Family") + ylab("Bray Distance") + labs(color="Family") +
    facet_grid(cols=vars(sample.b))
  plot.title <- paste(sample.set.1)
  plot <- plot + labs(title=plot.title)
  filename <- paste0(output.path, "/distance.boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                       ".png")
  ggsave(filename, plot, width=6, height=5, scale=0.75)
}
```

## Boxplot

```{r Family Comparison Boxplot}
family.comparison.boxplot <- function(dataframe, phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name, title=NA){
  
  sample.b.list <- sample.set.2
  # two <- color.list(5)[c(2, 6)]
  
  if(length(unique(dataframe[, "sample.b.type"]))==1){
    pwc <- dataframe %>%
      t_test(sample.a.percentage ~ family.type) %>%
      adjust_pvalue(method="bonferroni")%>%
      add_significance(p.col="p.adj")
  }else{
    pwc <- dataframe %>%
      group_by(sample.b.type) %>%
      t_test(sample.a.percentage ~ family.type) %>%
      adjust_pvalue(method="bonferroni")%>%
      add_significance(p.col="p.adj")
  }
  
    sample.b.list <- sample.set.2
    y.value=list()
    for(l in 1:length(sample.b.list)){
      y.value[l] <- max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "sample.a.percentage"])+
        (max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "sample.a.percentage"])/14)
    }
    
    # print(pwc)
    pwc[, c("y.position")] <- unlist(y.value)
    print(pwc)
    # pwc <- pwc[-c(which(pwc[, "p.adj.signif"]=="ns")), ]  # Star p values

    if(nrow(pwc)!=0){
      if(max(pwc[, c("y.position")]) < max(dataframe[, "sample.a.percentage"])){
        ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"]) + 
                 (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"])/30))
      }else{
        ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"]) + 
                 (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"])/9))
      }
    }else{
      ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"]) + 
               (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"])/30))
    }
    

  # sample.b.list <- sample.set.2
  # y.value=list()
  # for(l in 1:length(sample.b.list)){
  #   y.value[l] <- max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "sample.a.percentage"])+5
  # }
  # 
  # pwc[, c("y.position")] <- unlist(y.value)

  level <- subset.object(dataframe, "Same family")[order(subset.object(dataframe, "Same family")[, "sample.a.percentage"], decreasing=TRUE), "sample.a.fam"]

  colors <- rep("#000000", length(level))
  if(length(sample.set.2)==1){
    saveRDS(level, paste(paste0(output.path, "/", tolower(taxa.level)), sample.set.1, sample.set.2, "family.levels", "rds", sep="."))
  }
  
  # pwc[, "p"] <- signif(pwc[, "p"], 2)
  pwc[, "p"] <- formatC(signif(pwc[, "p"], 2), digits=2, format="fg", flag="#")
  
  dataframe[, "sample.a.fam"] <- factor(dataframe[, "sample.a.fam"], levels=unique(level))
  # print(dataframe)
  dataframe[, "family.type"] <- factor(dataframe[, "family.type"], levels=c("Same family", "Different family"))
  
        cat("\n  ", "Plot generation for all families by mother sample")
  plot <- ggplot(dataframe, aes(x=family.type, y=sample.a.percentage)) +
    theme_classic() +
    geom_boxplot(fill="transparent", outlier.size=-1) +
    scale_color_manual(values=colors) +
    geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam)) +
    # ylim(0, max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "sample.a.percentage"])+15) +
    scale_y_continuous(limits=c(0, ylim), 
                       breaks=seq(0, ylim, 20)) +
    # scale_x_discrete(labels=c("Same family"="Same\nfamily", "Different family"="Different\nfamily")) +
    scale_x_discrete(labels=c("Same family"="Own\nMother", "Different family"="Other\nMother")) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(),
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #, angle=45, vjust=1, hjust=1
          panel.border=element_rect(color="black", fill=NA, size=0.75), legend.title=element_blank(), legend.box.spacing=unit(0, "pt"),
          axis.title.x=element_blank(), legend.position="none", strip.placement="bottom", plot.title=element_text(hjust=0.5),
          legend.background=element_blank(), panel.background=element_blank(), plot.background=element_blank(),
          panel.grid.major=element_line(linetype="dashed", color="gray95")) +
    stat_pvalue_manual(pwc, label="p", tip.length=0, bracket.size=0.5) +
    # stat_pvalue_manual(pwc, label="p.adj.signif", tip.length=0, bracket.size=0.5) +  # Star p values
    
    # xlab("Family") + labs(color="Family") +
    # ylab("Shared Relative\nAbundance (%)") + 
    ylab("Percent of infant nasal\nrelative abundance shared") + 
    # facet_grid(cols=vars(sample.b.type))
    facet_wrap("sample.b.type", ncol=4)
  
  # plot <- annotate_figure(plot, top=text_grob(label=paste(sample.set.1, "shared abundance"))) +
  #     bgcolor("#e1e1e1") + border(color="#e1e1e1")
  if(!is.na(title)){
    plot <- annotate_figure(plot, top=text_grob(label=paste(title))) +
        bgcolor("#e1e1e1") + border(color="#e1e1e1")
  }
  
  print(plot)
  
  
  # if(length(sample.set.2)>1){
  #   FileName <- paste0(output.path, "/Boxplot/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, 
  #                      ".top.all.taxa.facet.png")
  #   suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(2.75*length(sample.b.list)), height=4, scale=0.75))
  #   # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  # }else{
  #   if(n!=0){
  #     FileName <- paste0(output.path, "/Boxplot/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
  #                        ".top.", n, ".png")
  #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  #   }else{
  #     FileName <- paste0(output.path, "/Boxplot/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
  #                        ".top.all.taxa.png")
  #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  #   }
  # }
  
  if(length(sample.set.2)>1){
    FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, 
                       ".top.all.taxa.facet.png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(2.75*length(sample.b.list)), height=4, scale=0.75))
    # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  }else{
    if(n!=0){
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.", n, ".png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }else{
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.all.taxa.png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }
  }
  
  return(list(dataframe, level, plot))
  
  
}
```

## Add Taxa

```{r}
family.comparison.add.taxa <- function(dataframe){
  for(a in 1:nrow(dataframe)){
    if(!is.null(dataframe[a, "table"][[1]])){
      dataframe[a, "shared.taxa"] <- nrow(dataframe[a, "table"][[1]])-1
    }else if(is.null(dataframe[a, "table"][[1]])){
      dataframe[a, "shared.taxa"] <- 0
    }
  }
  return(dataframe)
}
```


## Taxa Boxplot

```{r}
family.comparison.taxa.boxplot <- function(dataframe, phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name, title=NA){
  # View(dataframe)
  
  pwc <- dataframe %>%
        group_by(sample.b.type) %>%
        t_test(shared.taxa ~ family.type) %>%
        # adjust_pvalue(method="bonferroni") %>%
        add_significance(p.col="p")
  
    sample.b.list <- sample.set.2
    y.value=list()
    for(l in 1:length(sample.b.list)){
      y.value[l] <- max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "shared.taxa"])+
        (max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "shared.taxa"])/14)
    }
  
    pwc[, c("y.position")] <- unlist(y.value)
    # print(pwc)
    # pwc <- pwc[-c(which(pwc[, "p.signif"]=="ns")), ]  # Star p values
    
    if(nrow(pwc)!=0){
      if(max(pwc[, c("y.position")]) < max(dataframe[, "shared.taxa"])){
        ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"]) + 
                 (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"])/30))
      }else{
        ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"]) + 
                 (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"])/9))
      }
    }else{
      ylim <- (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"]) + 
               (max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"])/30))
    }
  
    level <- subset.object(dataframe, "Same family")[order(subset.object(dataframe, "Same family")[, "shared.taxa"], decreasing=TRUE), "sample.a.fam"]
  
    colors <- rep("#000000", length(level))
    # if(length(sample.set.2)==1){
    #   saveRDS(level, paste(paste0(output.path, "/", tolower(taxa.level)), sample.set.1, sample.set.2, "family.levels", "rds", sep="."))
    # }
    
    # print(pwc)
    pwc[, "p"] <- signif(pwc[, "p"], 2)
    dataframe[, "sample.a.fam"] <- factor(dataframe[, "sample.a.fam"], levels=unique(level))
    dataframe[, "family.type"] <- factor(dataframe[, "family.type"], levels=c("Same family", "Different family"))
    # print(dataframe)
  
  plot <- ggplot(dataframe, aes(x=family.type, y=shared.taxa)) +
    theme_classic() +
    geom_boxplot(fill="transparent", outlier.size=-1) +
    scale_color_manual(values=colors) +
    geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam)) +
    ylim(0, ylim) +
    # scale_y_continuous(limits=c(0, max(subset.object(dataframe, sample.set.2, "sample.b.type")[, "shared.taxa"])+9),
                       # breaks=seq(0, 100, 20)) +
    # scale_x_discrete(labels=c("Different family"="Different\nfamily", "Same family"="Same\nfamily")) +
    scale_x_discrete(labels=c("Same family"="Own\nMother", "Different family"="Other\nMother")) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(),
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #, angle=45, vjust=1, hjust=1
          panel.border=element_rect(color="black", fill=NA, size=0.75), legend.title=element_blank(), legend.box.spacing=unit(0, "pt"),
          axis.title.x=element_blank(), legend.position="none", strip.placement="bottom", plot.title=element_text(hjust=0.5),
          legend.background=element_blank(), panel.background=element_blank(), plot.background=element_blank(),
          panel.grid.major=element_line(linetype="dashed", color="gray95")) +
    stat_pvalue_manual(pwc, label="p", tip.length=0, bracket.size=0.5) +
    # stat_pvalue_manual(pwc, label="p.signif", tip.length=0, bracket.size=0.5) +  # Star p values
    # xlab("Family") + labs(color="Family") +
    ylab("Shared taxa") + 
    facet_wrap("sample.b.type", ncol=4)
  
  # print(plot)
  
  if(!is.na(title)){
    plot <- annotate_figure(plot, top=text_grob(label=paste(title))) +
      bgcolor("#e1e1e1") + border(color="#e1e1e1")
  }
  
  # plot <- annotate_figure(plot, top=text_grob(label=paste(sample.set.1, "shared taxa"))) +
  #     bgcolor("#e1e1e1") + border(color="#e1e1e1")
  
  # if(length(sample.set.2)>1){
  #   FileName <- paste0(output.path, "/Boxplot/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, 
  #                      ".top.all.taxa.facet.png")
  #   suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(2.75*length(sample.b.list)), height=4, scale=0.75))
  #   # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  # }else{
  #   if(n!=0){
  #     FileName <- paste0(output.path, "/Boxplot/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
  #                        ".top.", n, ".png")
  #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  #   }else{
  #     FileName <- paste0(output.path, "/Boxplot/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
  #                        ".top.all.taxa.png")
  #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  #   }
  # }
  
  if(length(sample.set.2)>1){
    FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, 
                       ".top.all.taxa.facet.png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(2.75*length(sample.b.list)), height=4, scale=0.75))
    # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  }else{
    if(n!=0){
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.", n, ".png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }else{
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/boxplot.taxa.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.all.taxa.png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }
  }
  return(list(dataframe, level, plot))
}
```

### Abundance Taxa Linear

```{r}
family.comparison.abundance.taxa <- function(dataframe, phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name){
  sample.b.list <- sample.set.2
  for(a in 1:length(sample.b.list)){
    dataframe.sub <- subset.object(dataframe, sample.b.list[a], "sample.b.type")
    m <- lm(shared.taxa ~ sample.a.percentage, dataframe.sub)
    r2 <- format(summary(m)$r.squared, digits=2)
    p <- format(cor.test(dataframe.sub[, "shared.taxa"], dataframe.sub[, "sample.a.percentage"], 
                      alternative="greater", method="pearson")$p.value, digits=2)
    if(a==1){
      pvalues <- data.frame(sample.b.type=sample.b.list[a], group1="0", group2=(max(dataframe.sub[, "shared.taxa"])/2), r2=r2, p=p, 
                            y.position=(max(dataframe.sub[, "sample.a.percentage"])+
                                          (max(dataframe.sub[, "sample.a.percentage"])/14)))
    }else{
      pvalues <- rbind(pvalues, data.frame(sample.b.type=sample.b.list[a], group1="0", group2=(max(dataframe.sub[, "shared.taxa"])/2), r2=r2, p=p,
                                           y.position=(max(dataframe.sub[, "sample.a.percentage"])+
                                                         (max(dataframe.sub[, "sample.a.percentage"])/14))))
    }
  }
  pvalues[,"p"] <- as.numeric(pvalues[,"p"])
  pvalues <- add_significance(pvalues, p.col="p", output.col="p.signif")
  
  pvalues[,"group1"] <- as.numeric(pvalues[,"group1"])
  pvalues[,"group2"] <- as.numeric(pvalues[,"group2"])
  
  ylim <- max(pvalues[,"y.position"])+(max(pvalues[,"y.position"])/7)
  
  plot <- ggplot(dataframe, aes(x=shared.taxa, y=sample.a.percentage)) +
    theme_classic() +
    # geom_boxplot(fill="transparent", outlier.size=-1) +
    scale_color_manual(values=colors) +
    geom_point() +
    # geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam)) +
    scale_y_continuous(limits=c(0, ylim), breaks=seq(0, ylim, 20)) +
    expand_limits(x=0) +
    scale_x_continuous(breaks=pretty_breaks()) +
    # scale_x_discrete(labels=c("Different family"="Different\nfamily", "Same family"="Same\nfamily")) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(),
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #, angle=45, vjust=1, hjust=1
          panel.border=element_rect(color="black", fill=NA, size=0.75), legend.title=element_blank(), legend.box.spacing=unit(0, "pt"),
          legend.position="none", strip.placement="bottom", plot.title=element_text(hjust=0.5),
          legend.background=element_blank(), panel.background=element_blank(), plot.background=element_blank(),
          panel.grid.major=element_line(linetype="dashed", color="gray95")) +
    # stat_pvalue_manual(pwc, label="p.adj.signif", tip.length=0, bracket.size=0.5) +
    ylab("Shared Relative\nAbundance (%)") + xlab("Shared taxa") +
    stat_pvalue_manual(pvalues, label="p={p}\nr2={r2}", tip.length=0, bracket.size=0.5, remove.bracket=TRUE, size=3) + #size=3.88
    facet_wrap("sample.b.type", ncol=4, scales="free_x")
  print(plot)
  
  plot <- annotate_figure(plot, top=text_grob(label=paste(sample.set.1, "same family"))) +
    bgcolor("#e1e1e1") + border(color="#e1e1e1")
  
  if(length(sample.set.2)>1){
    FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/linear.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, 
                       ".top.all.taxa.facet.png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(2.75*length(sample.b.list)), height=4, scale=0.75))
    # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
  }else{
    if(n!=0){
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/linear.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.", n, ".png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }else{
      FileName <- paste0(output.path, "/Boxplot/", str_split(sample.set.1, " ")[[1]][2], "/linear.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                         ".top.all.taxa.png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(9*length(sample.b.list)), height=9, scale=0.4))
    }
  }
}
```


```{r}

  # geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam)) +
            # phyloseq <- subset.object(phyloseq, sample.set.1, pheno.group)
  # birth_type <- data.frame(sample_data(phyloseq))[, c("family", "birth_type")]
  # colnames(birth_type)[1] <- "sample.a.fam"
  # dataframe <- merge(as.data.frame(dataframe), as.data.frame(birth_type), by="sample.a.fam")
  # phyloseq.frame.sub <- sample.to.data.frame(subset.object(phyloseq, sample.set.1, "type"))[,c("family", "birth_type")]   ###ADDED
  # colnames(phyloseq.frame.sub)[1] <- "sample.a.fam"    ###ADDED
  # print(pwc)
  # sample.b.list <- unique(dataframe[, "sample.b.type"])
  # dataframe[, "sample.a.fam"] <- factor(dataframe[, "sample.a.fam"], levels=families)
  # colors <- color.list.grad(length(level))
  # dataframe <- merge(dataframe, phyloseq.frame.sub, by="sample.a.fam")   ###ADDED
  #   geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=birth_type)) +      ###ADDED
  #   scale_color_manual(values=two) + ###ADDED
    # geom_text(position=position_jitterdodge(seed=1L), alpha=1, size=2, aes(color=sample.a.fam, label=sample.a.fam)) +
    # scale_color_manual(values=colors) +
    # labs(title=paste(sample.set.1, taxa.level)) +
    # ylim(0, 100) +
  #       pvalue <- subset.object(dataframe, "Same family", "family.type") %>%
  #     group_by(sample.b.type) %>%
  #     t_test(sample.a.percentage ~ birth_type) %>%
  #     adjust_pvalue(method="bonferroni")
  # print("Same family")
  # print(pvalue)
  # 
  # pvalue <- subset.object(dataframe, "Different family") %>%
  #     group_by(sample.b.type) %>%
  #     t_test(sample.a.percentage ~ birth_type) %>%
  #     adjust_pvalue(method="bonferroni")
  # print("Different family")
  # print(pvalue)
      
  #         cat("\n  ", "Plot generation for all families by mother sample")
  # plot <- ggplot(dataframe, aes(x=family.type, y=sample.a.percentage)) +
  #   geom_boxplot(outlier.size=-1) +
  #   # ylim(0, y.lim) +
  #   geom_point(position=position_jitterdodge(seed=1L), alpha=1, size=3, aes(color=birth_type)) +#, fill=sample.a.fam)) +
  #   scale_fill_manual(values=two) +
  #   theme(text=element_text(size=20), strip.placement="bottom", 
  #     strip.background=element_rect(fill="white")) +
  #   stat_pvalue_manual(pwc, label="p={p}", size=8) +
  #   xlab("Family") + ylab("Percentage of shared species") + labs(color="Family") +
  #   facet_grid(cols=vars(sample.b.type))
  
  # plot.title <- paste(sample.set.1)
  # if(n!=0){
  #   plot.title <- paste(plot.title, "top", n, "shared")
  # }else{
  #   plot.title <- paste(plot.title, "all taxa shared")
  # }
  # if(taxa.level=="Family"){
  #   # plot.title <- paste(plot.title, "families")
  # }else if(taxa.level=="Genus"){
  #   # plot.title <- paste(plot.title, "genuses")
  # }else if(taxa.level=="Species"){
  #   # plot.title <- paste(plot.title, "species")
  # }else if(taxa.level=="ASV"){
  #   # plot.title <- paste(plot.title, "asvs")
  # }
  # plot <- plot + ggtitle(plot.title)

# for(k in 1:length(sample.set.2)){
    #   merged.family.comparison.subset.2 <- subset.object(merged.family.comparison.subset, sample.set.2[k], "sample.b.type")
    #   
    #                     cat("\n     ", "Boxplot creation")
    #   if(max(merged.family.comparison.subset.2[, 3])+8 >100){
    #     y.lim <- 100
    #   }else{
    #     y.lim <- max(merged.family.comparison.subset.2[, 3])+8
    #   }
    #                     
    #   plot <- ggplot(merged.family.comparison.subset.2, aes(x=sample.a.fam, y=sample.a.percentage)) +
    #     geom_boxplot(outlier.size=-1) + ylim(0, y.lim) +
    #     geom_point(position=position_jitter(0.2), alpha=1, size=2, aes(color=sample.b.fam)) +
    #     scale_color_manual(values=colors) +
    #     # geom_line(aes(group=sample.b.fam, color=sample.b.fam)) +
    #     # scale_color_manual(values=colors) +
    #     theme(text=element_text(size=20)) +
    #     # stat_pvalue_manual(pwc, label="p={p}", size=8) +
    #     xlab(paste0("Family")) + 
    #     ylab(paste0("Percentage of shared species", "\n", sample.set.1[j], " and ", sample.set.2[k])) + 
    #     labs(color="Comparison\nfamily") 
    #     
    #   if(n!=0){
    #     plot <- plot + ggtitle(paste("Shared species for all families top", n))
    #     FileName <- paste(paste0(output.path, "/boxplot"), phyloseq.name, sample.set.1[j], sample.set.2[k], 
    #                        "top", n, "shared.species.by.family.png", sep=".")
    #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=27, height=9))
    #   }else{
    #     plot <- plot + ggtitle(paste("Shared species for all families all taxa"))
    #     FileName <- paste(paste0(output.path, "/boxplot"), phyloseq.name, sample.set.1[j], sample.set.2[k], 
    #                        "top.all.taxa.shared.species.by.family.png", sep=".")
    #     suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=27, height=9))
    #   }
    # }
```


## Beta

```{r Family Comparison Beta}
family.comparison.beta <- function(dataframe, phyloseq, level, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name){
  phyloseq.subset <- subset.object(phyloseq, sample.set.1, pheno.group)

  sample_data(phyloseq.subset)$family <- factor(sample_data(phyloseq.subset)$family, levels=unique(level))
  dataframe <- subset.object(dataframe, "Same family")
  
  two <- color.list(5)[c(2, 6, 1)]
  # cutoff <- c(75, 50, 25, 20, 15, 10, 0)
  cutoff <- c(25)
  for(e in 1:length(cutoff)){
    for(d in 1:nrow(sample_data(phyloseq.subset))){
      row <- which(dataframe[, "sample.a.fam"]==as.character(uln(sample_data(phyloseq.subset)[d, "family"])))
      if(length(row)==1){
        sample.a.percent <- dataframe[row, "sample.a.percentage"]
        # print(sample.a.percent)
        if(sample.a.percent>cutoff[e]){
          sample_data(phyloseq.subset)[d, "pcoa.group"] <- paste0(">", cutoff[e])
        }else if(sample.a.percent<=cutoff[e]){
          sample_data(phyloseq.subset)[d, "pcoa.group"] <- paste0("<", cutoff[e])
        }
      }
    }
    phyloseq.subset <- prune_samples(!is.na(sample_data(phyloseq.subset)$pcoa.group), phyloseq.subset)
    if(length(uuln(sample_data(phyloseq.subset)[, "pcoa.group"]))==2){
            doadonis<- function(phyloseq, distance, factor){
                                  bdist <- phyloseq::distance(phyloseq, distance)
                                  col <- as(sample_data(phyloseq), "data.frame")[, factor]
                                  adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
                                  p <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
                                  r2 <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
                                  
                                        cat("\n      ", paste("p=", p))
                                  res <- c(p, r2)
                                  return(res)
                                  }
      res <- doadonis(phyloseq.subset, distance="bray", factor="pcoa.group")
      ps <- phyloseq::ordinate(phyloseq.subset, method="PCoA", distance="bray")
      # ps <- phyloseq::ordinate(phyloseq, method="NMDS", distance="bray")
      rela_eig <- percent(ps$values$Relative_eig, accuracy=0.1)
      
      cat("\n    ", paste("Beta diversity plot generation"))
      p <- phyloseq::plot_ordination(phyloseq.subset, ps, color="pcoa.group") +
        # theme_classic() +
        geom_point(aes(size=reads)) +
        scale_color_manual(values=two) +
        guides(size=FALSE) +
        stat_ellipse(type="norm") +
        ggtitle(paste(sample.set.1, taxa.level), subtitle=paste0("p=", as.character(res[1]))) +
        xlab(paste0("PC 1 (", rela_eig[1], ")")) + ylab(paste0("PC 2 (", rela_eig[2], ")")) +
        theme(plot.title=element_text(hjust=0.5), strip.background=element_blank(), 
              strip.text.x=element_blank(), panel.grid.major=element_blank(), panel.grid.minor=element_blank(), 
              legend.position="bottom", 
              # legend.direction="vertical", 
              legend.title=element_blank(), legend.box.spacing=unit(0, 'cm'), legend.margin=margin(0, unit='cm'), 
              plot.margin=unit(c(0.5, 1, 0, 0.5), units="line"), 
              axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
              # legend.text=element_text(face="bold"), 
              panel.border=element_rect(color="black", fill=NA, size=0.75))
      FileName <- paste0(output.path, "/Beta/beta.", phyloseq.name, ".", tolower(taxa.level), ".", cutoff[e], ".", sample.set.1, ".", sample.set.2, 
                           ".top.all.taxa.png")
      ggsave(FileName, p, height=5, width=4)
    }
  }
  return(phyloseq.subset)
}
```

## Barplot

```{r Family Comparison Barplot}
family.comparison.barplot <- function(phyloseq, dataframe, level, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name, pheno){
  # phyloseq <- subset.object(phyloseq, sample.set.1, pheno.group)
  dataframe <- subset.object(dataframe, "Same family")
  pheno.frame <- data.frame(sample_data(phyloseq))[, c("family", pheno)]
  colnames(pheno.frame)[1] <- "sample.a.fam"
  sample.b.list <- sample.set.2
  dataframe <- merge(as.data.frame(dataframe), as.data.frame(pheno.frame), by="sample.a.fam")
  # dataframe <- remove.na(dataframe, "time")
  two <- color.list(5)[c(2, 6)]
  
  # print(dataframe)
  
  plot <- ggplot(dataframe, aes(x=reorder(sample.a.fam, -sample.a.percentage), y=sample.a.percentage)) +
    geom_col(aes(fill=get(pheno))) +
    scale_fill_manual(values=two) +
    ggtitle(paste(sample.set.1, pheno))+ xlab("Family") + ylab("Shared species (%)") +
    # facet_wrap(vars(sample.b.type), strip.position="bottom") +
    facet_wrap(vars(sample.b.type)) +
    theme(plot.title=element_text(hjust=0.5), 
    strip.background=element_rect(fill="white"), axis.title.x=element_blank(), axis.text.x=element_text(angle=45, vjust=1, hjust=1), 
    legend.position="bottom", legend.title=element_blank(), legend.box.spacing=unit(0, 'cm'), legend.margin=margin(0, unit='cm')) 

  if(n!=0){
    # plot <- plot + ggtitle(paste(sample.set.1, "top", n, "shared species for all families"))
    FileName <- paste0(output.path, "/Pheno/barplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".", pheno, ".top.", n, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(20*(ceiling(length(sample.b.list)/2))), 
                            height=6*(ceiling(length(sample.b.list)/2)), scale=0.3))
  }else{
    # plot <- plot + ggtitle(paste(sample.set.1, "all taxa shared species for all families"))
    FileName <- paste0(output.path, "/Pheno/barplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".", pheno, ".top.all.taxa.png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(20*(ceiling(length(sample.b.list)/2))), 
                            height=6*(ceiling(length(sample.b.list)/2)), scale=0.3))
  }
  
  
  colnames(dataframe)[which(colnames(dataframe)==pheno)] <- "pheno"
  if(length(unique(dataframe[, "sample.b.type"]))==1){
    pwc <- dataframe %>%
      t_test(sample.a.percentage ~ pheno) %>%
      adjust_pvalue(method="bonferroni")
  }else{
    pwc <- dataframe %>%
      group_by(sample.b.type) %>%
      t_test(sample.a.percentage ~ pheno) %>%
      adjust_pvalue(method="bonferroni")
  }
  
  # sample.b.list <- unique(dataframe[, "sample.b.type"])
  sample.b.list <- sample.set.2
  y.value=list()
  for(l in 1:length(sample.b.list)){
    y.value[l] <- max(subset.object(dataframe, sample.b.list[l], "sample.b.type")[, "sample.a.percentage"])+5
  }
  pwc[, c("y.position")] <- unlist(y.value)
  colors <- rep("#000000", length(level))
  plot <- ggplot(dataframe, aes(x=reorder(pheno, -sample.a.percentage), y=sample.a.percentage)) +
    geom_boxplot(outlier.size=-1) +
    geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam)) +
    scale_color_manual(values=colors) +
    theme(axis.title.x=element_blank(), legend.position="none", strip.placement="bottom", 
      strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5)) +
    stat_pvalue_manual(pwc, label="p={p}") + ylab("Shared species (%)") +
    labs(title=paste(sample.set.1, sample.set.2, pheno))
  print(plot)
  FileName <- paste0(output.path, "/Pheno/boxplot.pheno.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".", pheno, ".top.all.taxa.png")
  suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=18, 
                            height=18, scale=0.2))
}
```

## Linear

```{r Family Comparison Linear}
family.comparison.linear <- function(phyloseq, dataframe, families, level, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name){
  linear <- subset.object(dataframe, "Same family", "family.type")
  linear <- linear[, c("sample.a.fam", "sample.a.percentage")]
  colnames(linear)[2] <- "Same family percentage"
  # print(families)
  for(a in 1:length(families)){
    perc <- subset.object(dataframe, "Different family", "family.type")
    perc <- subset.object(perc, families[a], "sample.a.fam")
    linear[which(linear[, "sample.a.fam"]==families[a]), "Different family percentage"] <- mean(perc[, "sample.a.percentage"])
  }
  # print(linear)
  
  
  # phyloseq <- subset.object(phyloseq, sample.set.1, pheno.group)
  # reads <- data.frame(sample_data(phyloseq))[, c("family", "reads")]
  # colnames(reads)[1] <- "sample.a.fam"
  # linear <- merge(as.data.frame(linear), as.data.frame(reads), by="sample.a.fam")
  
  # print(max(linear[, "Same family percentage"]))
  linear <- remove.na(linear, "Different family percentage")
  colors <- color.list.grad(length(level))
  colors <- rep("#000000", length(level))
  m <- lm(`Same family percentage` ~ `Different family percentage`, linear)
  r2=format(summary(m)$r.squared, digits=2)
  p=format(cor.test(linear[, "Same family percentage"], linear[, "Different family percentage"], 
                    alternative="greater", method="pearson")$p.value, digits=2)
  plot <- ggplot(linear, aes(x=`Same family percentage`, y=`Different family percentage`)) +
    # geom_point(position=position_jitterdodge(seed=1L), alpha=1, aes(color=sample.a.fam, size=reads)) +
    geom_point(position=position_jitterdodge(seed=1L), size=3, alpha=1, aes(color=sample.a.fam)) +
    # guides(size=guide_legend(ncol=2)) +
    scale_color_manual(values=colors) +
    geom_smooth(method=lm, color="black", se=TRUE) +
    # guides(fill="none") +
    labs(title=paste(sample.set.1, sample.set.2, "shared comparison"), caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +#, x=paste(pheno.a[a]), y=pheno.b[b]) +
    theme(legend.position="none", plot.title=element_text(hjust=0.5))
    # theme(legend.box.margin=margin(6, 6, 6, 6))
  
  # plot <- plot + ggtitle(paste(sample.set.1, sample.set.2, "shared comparison"))
  
  if(n!=0){
    # plot <- plot + ggtitle(paste(sample.set.1, "top", n, "shared species for all families"))
    FileName <- paste0(output.path, "/Linear/linear.plot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                       ".top.", n, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=(18*(ceiling(length(sample.b.list)/2))), 
                            height=6*(ceiling(length(sample.b.list)/2)), scale=0.2))
  }else{
    # plot <- plot + ggtitle(paste(sample.set.1, "all taxa shared species for all families"))
    FileName <- paste0(output.path, "/Linear/linear.plot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, 
                       ".top.all.taxa.png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=18, 
                            height=18, scale=0.2))
  }
}
```

## Abundance Bar Plot

```{r Family Comparison Abundance Bar Plot}
family.comparison.abundance.barplot <- function(dataframe, taxa.level, sample.set.1, sample.set.2, n, output.path, phyloseq.name){
  dataframe <- subset.object(dataframe, "Same family")
  for(a in 1:nrow(dataframe)){
    if(dataframe[a, "sample.a.percentage"]!=0){
      abundance.table <- as.data.frame(tibble(dataframe[a, "table"]))[[1]][[1]]
      abundance.table[, "sample.a.fam"] <- dataframe[a, "sample.a.fam"]
      colnames(abundance.table)[8:11] <- c("sample.a.abundance", "sample.b.abundance", "sample.a.abundance.percent", "sample.b.abundance.percent")
      abundance.table <- abundance.table[c(1:(nrow(abundance.table)-1)), ]
      break()
    }else{
      dataframe.zero <- dataframe[a, ]
    }
  }
  
  for(b in a:nrow(dataframe)){
    if(dataframe[b, "sample.a.percentage"]!=0){
      abundance.table.merge <- as.data.frame(tibble(dataframe[b, "table"]))[[1]][[1]]
      abundance.table.merge[, "sample.a.fam"] <- dataframe[b, "sample.a.fam"]
      colnames(abundance.table.merge)[(ncol(abundance.table.merge)-4):(ncol(abundance.table.merge)-1)] <- c("sample.a.abundance", "sample.b.abundance", "sample.a.abundance.percent", "sample.b.abundance.percent")
      abundance.table.merge <- abundance.table.merge[c(1:(nrow(abundance.table.merge)-1)), ]
      abundance.table <- merge.data.frame(abundance.table, abundance.table.merge, all=TRUE)
    }else{
      if(exists("dataframe.zero")==FALSE){
        dataframe.zero <- dataframe[a, ]
      }else{
        dataframe.zero <- rbind(dataframe.zero, dataframe[b, ])
      }
    }
  }
  

  dataframe <- merge.data.frame(dataframe, abundance.table, by="sample.a.fam")
  
  filename.plot <- paste0(output.path, "/Abundance/stacked.barplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", 
                          sample.set.2, ".top.all.taxa.png")
  filename.legend <- paste0(output.path, "/Abundance/legend.stacked.barplot.", phyloseq.name, ".", tolower(taxa.level), ".", sample.set.1, ".", 
                            sample.set.2, ".top.all.taxa.png")
  
  if.asv.lvl <- if.asv(taxa.level)
  taxa.level <- if.asv.lvl[1]
  asv <- if.asv.lvl[2]
  
  if(taxa.level=="Species"){
    for(i in 1:nrow(dataframe)) {
      if(grepl("Unclassified", dataframe[i, "Species"], fixed=TRUE)==FALSE){
        dataframe[i, "Species"] <- paste(dataframe[i, "Genus"], dataframe[i, "Species"])
      }
    }
  }
  
  if(exists("dataframe.zero")==TRUE){
    dataframe <- bind_rows(dataframe, dataframe.zero)
  }
  colors <- color.list((length(uuln(dataframe[, taxa.level])))+1)[c(-1)]
  # dataframe <- replace.na(dataframe, 0)
  cols <- ceiling(length(uuln(dataframe[, taxa.level]))/15)
  # plot <- ggplot(dataframe, aes(x=reorder(sample.a.fam, -sample.a.percentage), y=sample.a.abundance)) +
  #   geom_bar(color="black", size=0.25, stat="identity", aes(fill=get(taxa.level))) +
  #   scale_fill_manual(values=colors) 
  
  write_csv(dataframe, paste0(output.path, "test.csv"))
  
  plot <- ggplot(dataframe, aes(x=reorder(sample.a.fam, -sample.a.percentage), y=sample.a.abundance)) +
    geom_bar(color="black", size=0.25, stat="identity", aes(fill=get(taxa.level))) +
    scale_fill_manual(values=colors)
  
  plot <- plot + ggtitle(paste(sample.set.1, taxa.level))+ xlab("Family") + ylab(paste("Shared", taxa.level, "(%)")) +
    facet_wrap(vars(sample.b.type))
  
  legend <- as_ggplot(get_legend(plot + guides(fill=guide_legend(ncol=cols)) + theme(legend.text=element_text(margin=margin(r=15)), 
                                                                                     legend.background=element_rect(color=NA)))) +
    theme(plot.margin=margin(0, 10, 0, 10), plot.background=element_rect(fill="white"))
  # suppressWarnings(ggsave(legend, file=FileName, type="cairo", width=30, height=3))
  suppressWarnings(ggsave(legend, file=filename.legend, type="cairo", height=5, width=(3.75*cols)))
  
  plot <- plot + guides(fill="none") +
    theme(plot.title=element_text(hjust=0.5), 
    strip.background=element_rect(fill="white"), axis.title.x=element_blank(), axis.text.x=element_text(angle=45, vjust=1, hjust=1)) 
  
  suppressWarnings(ggsave(plot, file=filename.plot, type="cairo", width=20, height=6, scale=0.3))
}
```

## Function Call

```{r Family comparison functions}
family.comparison.processing <- function(phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, n, path, phyloseq.name, pheno=NULL){
  output.path <- paste0(path, "/Output/Correlation")
  cat("\n", "Family comparison plot creation")
          cat("\n  ", "Reading rds phyloseq files to objects")
    phyloseq <- merge_phyloseq(subset.object(phyloseq, sample.set.1), subset.object(phyloseq, sample.set.2))
    families <- identify.families(phyloseq)
    
    merged.family.comparison <- family.comparison.table(phyloseq, taxa.level, n, sample.set.1, output.path, phyloseq.name)
    merged.family.comparison <- subset.object(merged.family.comparison, sample.set.2, "sample.b.type")
    

    family.comparison.prevalence(phyloseq, merged.family.comparison, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name)
    
    family.comparison.distance.boxplot(phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, output.path, phyloseq.name)
  
    for (j in 1:length(sample.set.1)) {
            cat("\n  ", "Subset by", sample.set.1[j])
      merged.family.comparison.subset <- subset.object(merged.family.comparison, sample.set.1[j], "sample.a.type")

      if(length(sample.set.2)>=1){
              cat("\n  ", "T-test for all families by sample types")
        merged.family.comparison.subset <- subset.object(merged.family.comparison.subset, sample.set.1[j], "sample.b.type", TRUE)
        list <- family.comparison.boxplot(merged.family.comparison.subset, phyloseq, taxa.level, pheno.group, sample.set.1, sample.set.2, n, output.path, phyloseq.name)

        merged.family.comparison.subset <- list[[1]]
        level <- list[[2]]

        phyloseq.subset <- family.comparison.beta(merged.family.comparison.subset, phyloseq, level, taxa.level, pheno.group, 
                                                  sample.set.1, sample.set.2, n, output.path, phyloseq.name)

        family.comparison.linear(phyloseq.subset, merged.family.comparison.subset, families, level, taxa.level, pheno.group, 
                                 sample.set.1, sample.set.2, n, output.path, phyloseq.name)
        
        if(!is.null(pheno)){
          family.comparison.barplot(phyloseq.subset, merged.family.comparison.subset, level, taxa.level, pheno.group, 
          sample.set.1, sample.set.2, n, output.path, phyloseq.name, pheno)
        }
        

        family.comparison.abundance.barplot(merged.family.comparison.subset, taxa.level, 
                                            sample.set.1, sample.set.2, n, output.path, phyloseq.name)
      }
  }
}
```

# Percent Shared

## Percent Shared vs Specific Taxa Linear Plot

```{r}
shared.percent.specific.taxa.linear.plot <- function(phyloseq, taxa.level, taxa.list, sample.set.1, sample.set.2, path, id, combine=FALSE, n=NULL){
  path <- paste0(path, "/Output/Correlation")
  if(combine==TRUE){
    table <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, path, "phyloseq.1-37")
    table <- subset.object(table, "Same family")
    colnames(table)[1:2] <- c("type", "family")
    phyloseq <- subset.object(phyloseq, sample.set.1)
    phyloseq<- trans.sam.counts(phyloseq)
    phyloseq.subset <- prune_taxa(taxa.list, phyloseq)
    merge <- merge(data.frame(sample_data(phyloseq.subset)), data.frame(otu_table(phyloseq.subset)),  by="row.names")
    rownames(merge) <- merge[, "Row.names"]
    merge <- merge[, -c(1)]
    
    
    merge[, "asv"] <- rowSums(merge[, c(24:(23+length(intersect(colnames(data.frame(otu_table(phyloseq))), taxa.list))))])
    
    merge <- merge(merge, table, by=c("type", "family"))
    merge <- subset.object(merge, sample.set.2)
    
    m <- lm(asv ~ sample.a.percentage, merge)
    r2=format(summary(m)$r.squared, digits=2)
    p=format(cor.test(merge[, "asv"], merge[, "sample.a.percentage"], 
                      alternative="greater", method="pearson")$p.value, digits=2)
    # print(merge)
    plot <- ggplot(merge, aes(x=asv, y=sample.a.percentage)) +
      theme_classic() +
      geom_point(size=3) +
      geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
      labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
      scale_y_continuous(limits=c(-50, 200)) +
      coord_cartesian(ylim=c(0, 100)) +
      ylab(paste(sample.set.1, "Shared (%)")) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95")) 
    if(length(uln(str_split(id, " ")))>6){
      plot <- plot + xlab(paste0(paste(str_split(id, " ")[[1]][1:(ceiling(length(uln(str_split(id, " ")))/2))], collapse=" "), "\n", paste(str_split(id, " ")[[1]][(ceiling(length(uln(str_split(id, " ")))/2)+1):length(uln(str_split(id, " ")))], collapse=" ")))
    }else{
      plot <- plot + xlab(paste(n, id))
    }
    print(plot)
    if(!is.null(n)){
      FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/", id, "/", n, ".taxa.linear.", tolower(taxa.level), 
                         ".", sample.set.1, ".", sample.set.2, ".combined.png")
    }else{
      FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/", id, "/taxa.linear.", tolower(taxa.level), 
                         ".", sample.set.1, ".", sample.set.2, ".combined.png")
    }
    
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                            height=18, scale=0.2))
  }else if(combine==FALSE){
    table <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, path, "phyloseq.1-37")
    table <- subset.object(table, "Same family")
    colnames(table)[1:2] <- c("type", "family")
    phyloseq <- subset.object(phyloseq, sample.set.1)
    phyloseq<- trans.sam.counts(phyloseq)
    for(a in 1:length(taxa.list)){
      # print(taxa.list[a])
      if(length(intersect(colnames(data.frame(otu_table(phyloseq))), taxa.list[a]))==1){
        phyloseq.subset <- prune_taxa(taxa.list[a], phyloseq)
        merge <- merge(data.frame(sample_data(phyloseq.subset)), data.frame(otu_table(phyloseq.subset)),  by="row.names")
        rownames(merge) <- merge[, "Row.names"]
        merge <- merge[, -c(1)]
        colnames(merge)[24] <- "asv"
        merge <- merge(merge, table, by=c("type", "family"))
        merge <- subset.object(merge, sample.set.2)
        # print(merge)
        
    
        m <- lm(asv ~ sample.a.percentage, merge)
        r2=format(summary(m)$r.squared, digits=2)
        p=format(cor.test(merge[, "asv"], merge[, "sample.a.percentage"], 
                          alternative="greater", method="pearson")$p.value, digits=2)
    
        plot <- ggplot(merge, aes(x=asv, y=sample.a.percentage)) +
          theme_classic() +
          geom_point(size=3) +
          geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
          labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
          scale_y_continuous(limits=c(-50, 200)) +
          coord_cartesian(ylim=c(0, 100)) +
          ylab(paste(sample.set.1, "Shared (%)")) +
          theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
                axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
                panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
                strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
                legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
                panel.grid.major=element_line(linetype="dashed", color="gray95"))
        if(taxa.level=="ASV"){
          if(str_detect(tax_table(phyloseq.subset)[, "Species"], "Unclassified")==TRUE){
            plot <- plot + xlab(paste(tax_table(phyloseq.subset)[, "Species"], "\nAbundance"))
          }else{
            plot <- plot + xlab(paste(tax_table(phyloseq.subset)[, "Genus"], tax_table(phyloseq.subset)[, "Species"], taxa.list[a],  "\nAbundance"))
          }
        }else{
          plot <- plot + xlab(paste(tax_table(phyloseq.subset)[, taxa.level], "\nAbundance"))
        }
        print(plot)
        FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/", id, "/taxa.linear.", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".", taxa.list[a], ".png")
        suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                                height=18, scale=0.2))
      }
    }
  }
}
```

 
## Percent Shared vs Specific Taxa Linear Plot: P value

```{r}
shared.percent.specific.taxa.linear.plot.p.value <- function(phyloseq, taxa.level, sample.set.1, sample.set.2, taxa.sample, path, n=NULL){
  path <- paste0(path, "/Output/Correlation")
  table <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, path, "phyloseq.1-37")
  table <- subset.object(table, "Same family")
  colnames(table)[1:2] <- c("type", "family")
  table <- subset.object(table, sample.set.2)
  # phyloseq <- subset.object(phyloseq, sample.set.1)
  phyloseq<- trans.sam.counts(phyloseq)
  # taxa.list <- top.taxa.n(subset.object(phyloseq, taxa.sample), "Species", n[length(n)])
  # phyloseq.subset <- prune_taxa(taxa.list, phyloseq)
  # merge <- merge(data.frame(sample_data(phyloseq.subset)), data.frame(otu_table(phyloseq.subset)),  by="row.names")
  # rownames(merge) <- merge[, "Row.names"]
  # merge <- merge[, -c(1)]

  n.list <- c()
  p.value.list <- c()
  for(a in 1:length(n)){
    
    taxa.list <- top.taxa.n(subset.object(phyloseq, taxa.sample), "Species", n[a])
    phyloseq.subset <- prune_taxa(taxa.list, phyloseq)
    phyloseq.subset <- subset.object(phyloseq.subset, taxa.sample)
    merge <- merge(data.frame(sample_data(phyloseq.subset)), data.frame(otu_table(phyloseq.subset)),  by="row.names")
    rownames(merge) <- merge[, "Row.names"]
    merge <- merge[, -c(1)]
    
    # print(merge)
    
    # print(merge)
    # print(merge[, c((ncol(merge)-n[a]+1):ncol(merge))])
    merge.sub <- merge
    merge.sub[, "asv"] <- rowSums(merge.sub[, c((ncol(merge.sub)-n[a]+1):ncol(merge.sub))])
    
    # print(table)
    merge.sub <- merge(merge.sub, table, by=c("family"))
    merge.sub <- subset.object(merge.sub, sample.set.2)
    # print(merge.sub)
    
    m <- lm(asv ~ sample.a.percentage, merge.sub)
    r2 <- format(summary(m)$r.squared, digits=2)
    p <- format(cor.test(merge.sub[, "asv"], merge.sub[, "sample.a.percentage"], 
                      alternative="greater", method="pearson")$p.value, digits=2)
    n.list[a] <- n[a]
    p.value.list[a] <- as.numeric(p)
  }
  # print(n.list)
  # print(p.value.list)
  
  p.value.table <- data.frame(taxa.number=n.list, p.value=p.value.list)
  p.value.table[, "p.value"] <- -log10(p.value.table[, "p.value"])
  plot <- ggplot(p.value.table, aes(x=taxa.number, y=p.value)) +
    theme_classic() +
    geom_point(size=3) +
    ylab(bquote(atop(Shared~percent~top~taxa, p~value~(-log[10])))) + xlab(paste(taxa.sample, "number of top", taxa.level, "combined")) +
    labs(title=sample.set.2) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
          legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          panel.grid.major=element_line(linetype="dashed", color="gray95")) 
  
  FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/p.value.plot.", tolower(taxa.level), ".", sample.set.1, ".", sample.set.2, ".top.taxa.of.", taxa.sample, ".png")
        suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=24, 
                                height=18, scale=0.2))
}
```


## Percent Shared vs Reads

```{r}
shared.percent.reads.linear.plot <- function(phyloseq, sample.set.1, sample.set.2, reads.sample, path){
  path <- paste0(path, "/Output/Correlation")
  table <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, path, "phyloseq.1-37")
  table <- subset.object(table, "Same family")
  colnames(table)[2] <- "family"
  phyloseq.subset <- subset.object(phyloseq, reads.sample)
  
  merge <- merge(data.frame(sample_data(phyloseq.subset)), table, by="family")
  merge <- subset.object(merge, sample.set.2)
  
  m <- lm(reads ~ sample.a.percentage, merge)
  r2=format(summary(m)$r.squared, digits=2)
  p=format(cor.test(merge[, "reads"], merge[, "sample.a.percentage"], 
                    alternative="greater", method="pearson")$p.value, digits=2)
  # print(merge)
  plot <- ggplot(merge, aes(x=reads, y=sample.a.percentage)) +
    theme_classic() +
    geom_point(size=3) +
    geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
    labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
    scale_y_continuous(limits=c(-50, 200)) +
    coord_cartesian(ylim=c(0, 100)) +
    ylab(paste(sample.set.1, "Shared (%)")) + xlab(paste(reads.sample, "Reads")) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
          legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          panel.grid.major=element_line(linetype="dashed", color="gray95")) 
  print(plot)
  FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Reads/reads.linear.", ".reads.", sample.set.1, ".", sample.set.2, ".reads.", reads.sample, ".png")
  suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                          height=18, scale=0.2))
}
```

## Percent Shared vs Number of shared asv

```{r}
shared.percent.shared.taxa.linear.plot <- function(phyloseq, sample.set.1, sample.set.2, path, pheno=NULL){
  path <- paste0(path, "/Output/Correlation")
  table <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, path, "phyloseq.1-37")
  table <- subset.object(table, "Same family")
  table <- subset.object(table, sample.set.2)
  colnames(table)[2] <- "family"
  
  for(a in 1:nrow(table)){
    if(table[a, "sample.a.percentage"]!=0){
      table[a, "shared.taxa.num"] <- nrow(table[a, "table"][[1]])-1
    }else{
      table[a, "shared.taxa.num"] <- 0
    }
  }
  table <- subset.object(table, 12, "shared.taxa.num", TRUE)
  table <- subset.object(table, 27, "shared.taxa.num", TRUE)
  table <- subset.object(table, 17, "shared.taxa.num", TRUE)
  table <- subset.object(table, 10, "shared.taxa.num", TRUE)
  table <- subset.object(table, 16, "shared.taxa.num", TRUE)

  phyloseq <- subset.object(phyloseq, sample.set.1)
  table <- merge(data.frame(sample_data(phyloseq)), table, by="family")
  
  
  
  m <- lm(shared.taxa.num ~ sample.a.percentage, table)
  r2=format(summary(m)$r.squared, digits=2)
  p=format(cor.test(table[, "shared.taxa.num"], table[, "sample.a.percentage"], 
                    alternative="greater", method="pearson")$p.value, digits=2)
  
  # print(table)
  
  # table <- subset.object(table, 27, "shared.taxa.num", TRUE)
  # print(table)
  color <- color.list(5)[c(2, 3, 5, 6, 8)]
  
  if(is.null(pheno)){
    
    
    plot <- ggplot(table, aes(x=shared.taxa.num, y=sample.a.percentage)) +
      theme_classic() +
      geom_point(size=3) +
      geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
      labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
      scale_y_continuous(limits=c(-50, 200)) +
      coord_cartesian(ylim=c(0, 100)) +
      ylab(paste(sample.set.1, "Shared (%)")) + xlab("Number of shared ASVs\nin family pairs") +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))
    print(plot)
    FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/shared.taxa.number.linear.", sample.set.1, ".", sample.set.2, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                            height=18, scale=0.2))
    
    
    m <- lm(time.min ~ sample.a.percentage, table)
    r2=format(summary(m)$r.squared, digits=2)
    p=format(cor.test(table[, "time.min"], table[, "sample.a.percentage"], 
                    alternative="greater", method="pearson")$p.value, digits=2)
    
    plot <- ggplot(table, aes(x=time.min, y=sample.a.percentage)) +
      theme_classic() +
      geom_point(size=3) +
      geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
      labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
      scale_y_continuous(limits=c(-50, 200)) +
      coord_cartesian(ylim=c(0, 100)) +
      ylab(paste(sample.set.1, "Shared (%)")) + xlab("Time between birth and\nsample collection (min)") +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))
    print(plot)
    FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/shared.taxa.number.linear.time.", sample.set.1, ".", sample.set.2, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                            height=18, scale=0.2))
    
    m <- lm(birthweight ~ sample.a.percentage, table)
    r2=format(summary(m)$r.squared, digits=2)
    p=format(cor.test(table[, "birthweight"], table[, "sample.a.percentage"], 
                    alternative="greater", method="pearson")$p.value, digits=2)
    
    plot <- ggplot(table, aes(x=birthweight, y=sample.a.percentage)) +
      theme_classic() +
      geom_point(size=3) +
      geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
      labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
      scale_y_continuous(limits=c(-50, 200)) +
      coord_cartesian(ylim=c(0, 100)) +
      ylab(paste(sample.set.1, "Shared (%)")) + xlab("Birthweight") +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))
    print(plot)
    FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/shared.taxa.number.linear.birthweight.", sample.set.1, ".", sample.set.2, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                            height=18, scale=0.2))

    
  }else{
    plot <- ggplot(table, aes(x=shared.taxa.num, y=sample.a.percentage)) +
      theme_classic() +
      geom_point(size=3, aes(color=get(pheno))) +
      scale_color_manual(values=color) +
      guides(color=guide_legend(title=pheno)) +
      geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
      labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
      scale_y_continuous(limits=c(-50, 200)) +
      coord_cartesian(ylim=c(0, 100)) +
      ylab(paste(sample.set.1, "Shared (%)")) + xlab("Number of shared taxa\nin family pairs") +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))
    print(plot)
    FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/shared.taxa.number.linear.", sample.set.1, ".", sample.set.2, ".", pheno, ".png")
    suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=24, 
                            height=18, scale=0.2))
  }
  
  # plot <- ggplot(table, aes(x=shared.taxa.num, y=sample.a.percentage)) +
  #   theme_classic() +
  #   geom_point(size=3, aes(color=pheno)) +
  #   scale_color_manual(values=color) +
  #   geom_smooth(method=lm, color="black", se=TRUE, fill="gray75") +
  #   labs(title=sample.set.2, caption=bquote(atop(r^2==~ .(r2), p==~ .(p)))) +
  #   scale_y_continuous(limits=c(-50, 200)) +
  #   coord_cartesian(ylim=c(0, 100)) +
  #   ylab(paste(sample.set.1, "Shared (%)")) + xlab("Number of shared taxa\nin family pairs") +
  #   theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
  #         axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
  #         panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
  #         strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
  #         #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
  #         panel.grid.major=element_line(linetype="dashed", color="gray95"))
  # print(plot)
  # FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/shared.taxa.number.linear.", sample.set.1, ".", sample.set.2, ".png")
  # suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=24, 
  #                         height=18, scale=0.2))
  
  if(!is.null(pheno)){
    if(length(uuln(table[, pheno]))<6){
      comb.list <- uuln(table[, pheno])
      comparisons <- as.list(as.data.frame(t(CombPairs(comb.list))))
      if(pheno=="sequencing"){
        table[, pheno] <- factor(table[, pheno], levels=c("1-5", "6-10", "11-19", "20-28", "29-37"))
      }
    
      stat <- as.formula(paste("sample.a.percentage", "~", pheno))
          stat.test <- table %>%
            wilcox_test(stat, comparisons=comparisons) %>%
            mutate(myformatted.p=paste0("p=", signif(p, 2)))
    
      color <- rep("#000000", 10)
      plot <- ggplot(table, aes(x=get(pheno), y=sample.a.percentage)) +
        theme_classic() +
        geom_boxplot() +
        geom_point(position=position_jitterdodge(seed=1L), size=3, alpha=1, aes(color=get(pheno))) +
        scale_color_manual(values=color) +
        scale_y_continuous(limits=c(-50, 200)) +
        coord_cartesian(ylim=c(0, 200)) +
        # coord_cartesian(ylim=c(0, 100)) +
        stat_compare_means(method="wilcox.test", comparisons=comparisons, aes(label=paste0("p=", ..p.format..), )) +
        ylab(paste(sample.set.1, "Shared (%)")) + xlab(paste(pheno, "Groups")) + labs(title=sample.set.2) +
        theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
              axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
              panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
              strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
              legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
              panel.grid.major=element_line(linetype="dashed", color="gray95"))
      print(plot)
      FileName <- paste0(path, "/Shared Percent Linear/", sample.set.2, "/Number of Shared Taxa/percent.shared.boxplot.", sample.set.1, ".", sample.set.2, ".", pheno, ".png")
      suppressWarnings(ggsave(plot, file=FileName, type="cairo", width=20, 
                              height=18, scale=0.2))
    }
  }
}
```

## Percent Shared Ranked

```{r}
pecent.shared.ranked <- function(phyloseq, sample.set.1, sample.set.2, path, phyloseq.name){
  three <- color.list(5)[c(2, 3, 5, 6)]
  
  dataframe <- family.comparison.table(phyloseq, "ASV", 0, sample.set.1, paste0(path, "/Output/Correlation"), phyloseq.name)
  dataframe <- subset.object(dataframe, "Same family")
  
  length <- max(sapply(sample.set.2, function(x) nrow(subset.object(dataframe, x))))
  
  plot <- ggplot(dataframe, aes(x=sample.a.fam, y=sample.a.percentage)) +
    theme_classic() +
    geom_boxplot() +
    geom_point(size=2.5, aes(color=sample.b.type)) +
    scale_color_manual(values=three) +
    scale_y_continuous(limits=c(-50, 200)) +
    coord_cartesian(ylim=c(0, 100)) +
    ylab("Shared Relative Abundance (%)") + xlab(sample.set.1) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), legend.title=element_blank(), 
          legend.box.spacing=unit(0, "pt"), 
          #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          panel.grid.major=element_line(linetype="dashed", color="gray95")) 
  print(plot)
  FileName <- paste0(path, "/Output/Correlation/Ranking/boxplot.ranking.asv.", sample.set.1, ".png")
  suppressWarnings(ggsave(plot, file=FileName, type="cairo", 
                          width=(8+(0.8*length)), 
                          height=(0.8*length), scale=0.2))
  
  plot <- ggplot(dataframe, aes(x=sample.a.fam, y=sample.a.percentage)) +
  theme_classic() +
  geom_point(size=2.5, aes(color=sample.b.type)) +
  scale_color_manual(values=three) +
  geom_line(aes(group=sample.b.type, color=sample.b.type)) +
  # geom_line(alpha=0.5, aes(group=sample.b.type)) +
  ylab("Shared Relative Abundance (%)") + xlab(sample.set.1) +
  theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
        axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
        panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
        strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), legend.title=element_blank(), 
        legend.box.spacing=unit(0, "pt"), 
        #legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
        panel.grid.major=element_line(linetype="dashed", color="gray95")) 
  print(plot)
  FileName <- paste0(path, "/Output/Correlation/Ranking/trendline.ranking.asv.", sample.set.1, ".png")
  suppressWarnings(ggsave(plot, file=FileName, type="cairo", 
                          width=(8+(0.8*length)), 
                          height=(0.8*length), scale=0.2))
}
```

# Comparisons

## Same Sample Comparison

```{r}
same.sample.comparison <- function(phyloseq, taxa.level, n, sample.set.1, sample.set.2, path, phyloseq.name){
  path <- paste0(path, "/Output/Correlation")
  families <- identify.families(phyloseq)
  table <- family.comparison.table(phyloseq, taxa.level, n, sample.set.1, path, phyloseq.name)
  linear <- table
  linear <- linear[, c("sample.a.type", "sample.a.fam")]
  linear <- unique(linear)
  for(a in 1:length(families)){
    perc <- table
    perc <- subset.object(perc, families[a], "sample.a.fam")
    linear[which(linear[, "sample.a.fam"]==families[a]), "Average Shared"] <- mean(perc[, "sample.a.percentage"])
  }
  linear.ordered <- linear[order(-(linear[, "Average Shared"])), ]
  linear.ordered[, "sample.a.fam"] <- factor(linear.ordered[, "sample.a.fam"], levels=linear.ordered[, "sample.a.fam"])
  family.order <- linear.ordered[, "sample.a.fam"]
  linear <- merge(linear, table[, c(1:3, 5)])
  colnames(linear) <- c("Type", "Family", "Average Shared", "Percent Shared", "Other Family")
  return(list(linear, family.order))
}
```

## Type Comparison Plot

```{r}
type.comparison.plot <- function(phyloseq, taxa.level, type, path, phyloseq.name, plot.title, color.number){
  list <- same.sample.comparison(phyloseq, taxa.level, 0, type, type, path, phyloseq.name)
  table <- list[[1]]
  colors <- color.list(5)[c(color.number)]
  
  plot <- ggplot(table, aes(x=`Family`, y=`Average Shared`)) +
    theme_classic() +
    geom_point(aes(color=`Type`)) +
    geom_line(aes(color=`Type`, group=`Type`)) +
    scale_color_manual(values=colors) +
    coord_cartesian(ylim=c(0, 100)) +
    scale_x_discrete(limits=c(list[[2]])) +
    geom_point(alpha=0.1, aes(y=`Percent Shared`/1)) +
    scale_y_continuous(name="% Shared", sec.axis=sec_axis(~.*1, name="")) +
    xlab("Family") + ylab("Average percent shared with\nall other samples") + labs(title=plot.title) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          axis.ticks.y.right=element_blank(), axis.title.y.right=element_blank(), axis.text.y.right=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), legend.title=element_blank(), 
          legend.box.spacing=unit(0, "pt"), #legend.position="bottom", legend.box="horizontal", 
          legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          panel.grid.major=element_line(linetype="dashed", color="gray95"))
  print(plot)
  ggsave(paste0(path, "/Output/Correlation/", phyloseq.name, ".", taxa.level, ".type.comparison.png"), plot, width=(1.5+(length(list[[2]])*0.325)), height=6.5, scale=0.5)
}
```

## Type Comparison Plot Combined

```{r}
type.comparison.plot.combined <- function(phyloseq, taxa.level, type, path, phyloseq.name, id, plot.title, color.numbers){
  for(a in 1:length(phyloseq)){
    list <- same.sample.comparison(phyloseq[[a]], taxa.level, 0, type, type, path, phyloseq.name[[a]])
    table <- list[[1]]
    table[, "form"] <- id[[a]]
    if(a==1){
      order <- list[[2]]
      table.all <- table
    }else{
      table.all <- rbind(table.all, table)
    }
  }
  table.all[, "form"] <- factor(table.all[, "form"], levels=id)
  
  colors <- color.list(5)[color.numbers]
  
  plot <- ggplot(table.all, aes(x=`Family`, y=`Average Shared`)) +
    theme_classic() +
    geom_point(aes(color=`form`)) +
    geom_line(aes(color=`form`, group=`form`)) +
    scale_color_manual(values=colors) +
    coord_cartesian(ylim=c(0, 100)) +
    scale_x_discrete(limits=c(order)) +
    # geom_point(alpha=0.1, aes(y=`Percent Shared`/1)) +
    # scale_y_continuous(name="% Shared", sec.axis=sec_axis(~.*1, name="Number of Shared ASVs")) +
    xlab("Family") + ylab("Average percent shared with\nall other samples") + labs(title=plot.title) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          # axis.ticks.y.right=element_blank(), axis.title.y.right=element_blank(), axis.text.y.right=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), legend.title=element_blank(), 
          legend.box.spacing=unit(0, "pt"), legend.position="bottom", legend.box="horizontal", 
          # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          panel.grid.major=element_line(linetype="dashed", color="gray95"))
  print(plot)
  ggsave(paste0(path, "/Output/Correlation/", paste(phyloseq.name, collapse="."), ".", taxa.level, ".type.comparison.png"), plot, width=(1.5+(length(order)*0.325)), height=6.5, scale=0.5)
}
```

## Cross Match Two Lists

```{r Cross Match Two Lists}
cross.match <- function(taxa.level, group.list, family.listings, title, x.title, output.path){
  group.1 <- group.list[1]
  group.2 <- group.list[2]
  family.list.1 <- family.listings[[1]]
  family.list.2 <- family.listings[[2]]
  if(length(family.list.1)>length(family.list.2)){
    nrow <- length(family.list.1)
    for(a in 1:(length(family.list.1)-length(family.list.2))){
      family.list.2 <- c(family.list.2, NA)
    }
  }else if(length(family.list.2)>length(family.list.1)){
    nrow <- length(family.list.2)
    for(a in 1:(length(family.list.2)-length(family.list.1))){
      family.list.1 <- c(family.list.1, NA)
    }
  }else if(length(family.list.1)==length(family.list.2)){
    nrow <- length(family.list.1)
  }
  dataframe <- data.frame(numeric.order=c(1:nrow, 1:nrow), family.list=c(family.list.1, family.list.2), 
                          group=c(rep(group.1, nrow), rep(group.2, nrow)))
  plot <- ggplot(dataframe, aes(x=factor(group, level=c(group.1, group.2)), y=numeric.order)) +
    theme_classic() +
    geom_text(size=3, aes(label=family.list)) +
    geom_line(alpha=0.2, aes(group=family.list)) +
    labs(title=str_wrap(title, 20)) + xlab(str_wrap(x.title, 20)) +
    scale_y_reverse() +
    theme(axis.line.x=element_blank(), axis.line.y=element_blank(), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          axis.title.y=element_blank(), axis.text.y=element_blank(), 
          axis.text.x=element_text(color="black", size=rel(1.2), vjust=5), 
          plot.title=element_text(hjust=0.5), 
          plot.background=element_rect(color="white"), panel.grid=element_line(color="white"))
  
  filename.plot <- paste(paste0(output.path, "/cross.match"), tolower(taxa.level), paste(unlist(str_replace_all(tolower(title), " ", ".")), sep="."), 
                         "png", sep=".")
  suppressWarnings(ggsave(plot, file=filename.plot, type="cairo", width=3, height=10, scale=0.75))
  
}
```

# Simple Beta Diversity

```{r Simple Beta Diversity}
beta <- function(phyloseq, path=NULL, phyloseq.name=NULL, pheno=NULL, connection=FALSE, highlight=NULL, text=FALSE, axes=c(1, 2), title=NA){
  if(!is.null(pheno)){
    sample_data(phyloseq) <- remove.na(sample_data(phyloseq), pheno)
  }
  bdist <- phyloseq::distance(phyloseq, "bray")
  
  
  type <- identify.type(phyloseq)
  
  if(!is.na(title)){
    if(length(str_split(title, " ")[[1]])>3){
      longer <- TRUE
      title <- paste0(paste(str_split(title, " ", n=3)[[1]][1], str_split(title, " ", n=3)[[1]][2]), "\n", str_split(title, " ", n=3)[[1]][3])
    }else{
      longer <- FALSE
    }
  }
  
  if(!is.null(pheno)){
    col <- as(sample_data(phyloseq), "data.frame")[, pheno]
    adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
    p.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
    r2.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
    res.adonis <- c(p.adonis, r2.adonis)
    
    anosim.dist <- vegan::anosim(unname(bdist), col, permutations=999)
    p.anosim <- signif(anosim.dist$signif, digits=2)
    r2.anosim <- signif(anosim.dist$statistic, digits=2)
    res.anosim <- c(p.anosim, r2.anosim)
  }
  
  ps <- phyloseq::ordinate(phyloseq, method="PCoA", distance="bray")
  rela_eig <- percent(ps$values$Relative_eig, accuracy=0.1)
  # View(ps)
  
  if(!is.null(pheno)){

      colors <- color.list(5)[c(-1)]
    # }
    if(!is.null(phyloseq.name)){
      if(phyloseq.name=="family.phyloseq" || phyloseq.name=="family.phyloseq.initial" || phyloseq.name=="family.phyloseq.followup"){
        colors <- color.list(5)[-c(1)]
      # }else if(phyloseq.name=="family.phyloseq.infant" || phyloseq.name=="family.phyloseq.initial.infant" || phyloseq.name=="family.phyloseq.followup.infant"){
      #   colors <- color.list(5)[-c(1)]
      }else if(phyloseq.name=="family.phyloseq.mother" || phyloseq.name=="family.phyloseq.initial.mother" || phyloseq.name=="family.phyloseq.followup.mother"){
        # colors <- color.list(5)[-c(1, 2, 3, 4)]
        colors <- color.list(21)[c(21, 3, 6, 9)]
        # colors <- color.list(5)[-c(1)]
      }else if(phyloseq.name=="type.phyloseq"){
        colors <- color.list(5)[-c(1)]
      }else if(phyloseq.name=="type.phyloseq.infant"){
        colors <- color.list(5)[-c(1)]
      }else if(phyloseq.name=="type.phyloseq.mother"){
        # colors <- color.list(5)[-c(1, 2, 3, 4)]
        colors <- color.list(5)[-c(1)]
      }else if(phyloseq.name=="family.phyloseq.initial.infant" || phyloseq.name=="family.phyloseq.followup.infant"){
        colors <- color.list(21)[c(12, 15, 18)]
      }
    }
    
  }

  
  plot <- plot_ordination(phyloseq, ps, axes=axes) +
    theme_classic()
  plot$layers[[1]]$aes_params$size <- 3
  
  if(text==TRUE){
    plot$layers[[1]]$aes_params$alpha <- 0
    # plot <- plot + geom_text(size=3, aes(label=family))
    plot <- plot + geom_text(size=2, aes(label=family, color=get(pheno)))
    # plot <- plot + geom_point(size=3, aes(color=ifelse(raw.reads>=22000, ">22000", "<22000"))) +
    # scale_color_manual(values=colors)
  }else{
    if(!is.null(pheno)){
      if(is.null(highlight)){
        plot$layers[[1]]$aes_params$alpha <- 0
        plot <- plot + geom_point(size=2.5, aes(color=get(pheno)))
        height <- 0.3 + (0.3 * length(uuln(sample.to.data.frame(phyloseq)[, pheno])))
        if(pheno=="type"){
          width <- 1.75
        }else if(pheno=="seq"){
          width <- 1.25
        }else{
          # height <- 0.3 + (0.3 * 2)
          width <- 1.6
        }
      }else{
        plot$layers[[1]]$aes_params$alpha <- 0
        plot <- plot + geom_point(size=2.5, aes(color=get(pheno), shape=get(highlight)))
      }
    }
  }
  legend <- as_ggplot(get_legend(plot +
                                   # guides(color=guide_legend(nrow=1)) +
                                   scale_color_manual(values=colors) +
                                   theme(#legend.box="horizontal",
                                         legend.title=element_blank(), legend.box.spacing=unit(0, "pt"), legend.spacing=unit(0, "pt"),
                                         legend.text=element_text(margin=margin(r=15)), legend.background=element_rect(fill="#e1e1e1"),
                                         legend.box.just="center", plot.background=element_blank(), legend.box.background=element_rect(fill='transparent')))) +
    theme(plot.margin=margin(0, 10, 0, 10), plot.background=element_blank()) 
  ggsave(filename=paste0(path, "/beta.legend.", phyloseq.name, ".png"), legend, width=width, height=height, dpi=300, scale=0.75)
  # ggsave(filename=paste0(path, "/beta.legend.", phyloseq.name, ".png"), plot, width=6, height=4, scale=0.75)
  
  plot <- plot +
    xlab(paste0("PC ", axes[1], " (", rela_eig[axes[1]], ")")) + ylab(paste0("PC ", axes[2], " (", rela_eig[axes[2]], ")")) +
    theme(axis.line.x=element_blank(), axis.line.y=element_blank(), panel.border=element_rect(color="black", fill=NA, size=0.75),
          panel.background=element_blank(), plot.title=element_text(hjust=0.5), plot.caption=element_text(hjust=0.5), 
          panel.grid.major=element_line(linetype="dashed", color="white"), legend.position="none", #legend.position="right", # #legend.position="bottom",
          legend.title=element_blank(), legend.box.spacing=unit(0, "pt"), legend.spacing=unit(5, "pt"), legend.box="vertical",
          plot.background=element_rect("#e1e1e1"))#, legend.background=element_blank()) #, panel.background=element_rect(fill="gray95") panel.grid.minor=element_line(linetype="dashed", color="white"))
  
  if(!is.null(pheno)){
    if(pheno=="form"){
      plot <- plot +
      scale_color_manual(values=colors, breaks=c("Initial", "Follow up"))
      if(!is.na(title)){
        if(longer==TRUE){
          width <- (0.875 + 3)
          height <- (2.1 + 3)
        }else{
          width <- (0.875 + 3)
          height <- (1.835 + 3)
        }
        # title <- expression(atop(title, scriptstyle("adonis=", as.character(res[1]), "   anosim=", as.character(res.anosim[1]))))
        plot <- plot + labs(title=title, subtitle=paste0("adonis=", as.character(res.adonis[1]), "   anosim=", as.character(res.anosim[1])))
        # plot <- plot + labs(title=title, subtitle=paste0("anosim=", as.character(res.anosim[1])))
      }else{
        # plot <- plot + labs(title=type, subtitle=paste0("adonis=", as.character(res.adonis[1]), "   anosim=", as.character(res.anosim[1])))
        plot <- plot + labs(title=title, subtitle=paste0("anosim=", as.character(res.anosim[1])))
      }
      # width <- 4
      # height <- 4.5
    }else{
      if(length(uuln(sample_data(phyloseq)[, pheno]))>1){
        plot <- plot +
        scale_color_manual(values=colors) +
        # labs(title=title, subtitle=paste0("adonis=", as.character(res.adonis[1]), " anosim=", as.character(res.anosim[1]))) 
        labs(subtitle=paste0("anosim=", as.character(res.anosim[1])))
        # + guides(color=guide_legend(nrow=2, byrow=TRUE))
        # width <- 4
        # height <- 4.55
          width <- (0.875 + 3)
          height <- (1.835 + 3)
      }else{
        plot <- plot +
        scale_color_manual(values=colors) +
        # labs(subtitle=paste0("adonis=", as.character(res.adonis[1]), " anosim=", as.character(res.anosim[1])))
        labs(subtitle=paste0("anosim=", as.character(res.anosim[1])))
        # width <- 4
        # height <- 4.15
      }
    }
  }
  
  if(connection==TRUE){
    plot <- plot + geom_line(alpha=0.2, aes(fill=family))
  }
  
  if(!is.null(path)){
    filename <- paste(paste0(path, "/beta.", phyloseq.name), axes[1], axes[2], sep=".")
    if(!is.null(pheno)){
      filename <- paste(filename, pheno, sep=".")
    }
    
    if(text==TRUE){
      filename <- paste(filename, "text", sep=".")
    }
    
    if(connection==TRUE){
      filename <- paste(filename, "connection", sep=".")
    }
    
    filename <- paste(filename, "png", sep=".")
  }
  
  height <- 0.7 + 3
  if(!is.na(title)){
    height <- height + 0.325
  }
  if(!is.null(pheno)){
    # No legend 
    height <- height + 0.29
  }
  
  if(!is.null(path)&!is.null(pheno)){
    ggsave(filename=filename, plot, width=(width*2), height=(height*2), dpi=600, scale=(0.75/2))
    # ggsave(filename=filename, plot, width=10, height=7, scale=1)
  }else{
    # ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    
    # ggsave(filename=filename, plot, width=4.5, height=4, scale=0.75)
    # ggsave(filename=filename, plot, width=6, height=4, scale=0.75)
    # ggsave(filename=filename, plot, width=6, height=4, scale=1)
  }
  
  p.table <- beta.p.table(phyloseq, pheno, uuln(sample.to.data.frame(phyloseq)[,pheno]))
  # print(p.table)
  
  return(list(legend, plot))
}
```

### Anosim for beta

```{r}
beta.p.table <- function(phyloseq, pheno, comparison.list, distance){
  comb.list <- comparison.list
  comb.list <- remove.na(comb.list)
  comparisons <- as.list(as.data.frame(t(CombPairs(comb.list))))
  
  for(a in 1:length(comparisons)){
    if(distance=="bray"){
      phyloseq.sub <- subset.object(phyloseq, uln(comparisons[a]), pheno)
      bdist <- phyloseq::distance(phyloseq.sub, distance)
      col <- as(sample_data(phyloseq.sub), "data.frame")[, pheno]
      adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
      p.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
      r2.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
      # res.adonis <- c(p.adonis, r2.adonis)
    }else if(distance=="unifrac"||distance=="wunifrac"){
      # set.seed(1L)
      set.seed(8696)
      phyloseq.sub <- subset.object(phyloseq, uln(comparisons[a]), pheno)
      random_tree <- rtree(ntaxa(phyloseq.sub), rooted=TRUE, tip.label=taxa_names(phyloseq.sub))
      phyloseq.sub <- merge_phyloseq(phyloseq.sub, random_tree)
      bdist <- phyloseq::distance(phyloseq.sub, distance)
      # print(bdist)
      col <- as(sample_data(phyloseq.sub), "data.frame")[, pheno]
      adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
      p.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
      r2.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
      # res.adonis <- c(p.adonis, r2.adonis)
    }
    
    
    anosim.dist <- vegan::anosim(unname(bdist), col, permutations=999)
    p.anosim <- signif(anosim.dist$signif, digits=2)
    r2.anosim <- signif(anosim.dist$statistic, digits=2)
    # res.anosim <- c(p.anosim, r2.anosim)
    
    # if(a==1){
    #   p.frame <- data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
    #                         adonis.p=p.adonis, adonis.r2=r2.adonis, 
    #                         anosim.p=p.anosim, anosim.r2=r2.anosim)
    # }else{
    #   p.frame <- rbind(p.frame, data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
    #                                        adonis.p=p.adonis, adonis.r2=r2.adonis, 
    #                                        anosim.p=p.anosim, anosim.r2=r2.anosim))
    # }
    
    if(a==1){
      p.frame <- data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
                            adonis.p=p.adonis, anosim.p=p.anosim)
    }else{
      p.frame <- rbind(p.frame, data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
                                           adonis.p=p.adonis, anosim.p=p.anosim))
    }
  }
  return(p.frame)
}
```

### Anosim for unifrac

```{r}
beta.p.table.unifrac <- function(phyloseq, pheno, comparison.list, distance){
  comparisons <- as.list(as.data.frame(t(CombPairs(comparison.list))))
  
  for(a in 1:length(comparisons)){
    phyloseq.sub <- subset.object(phyloseq, uln(comparisons[a]), pheno)
    random_tree <- rtree(ntaxa(phyloseq.sub), rooted=TRUE, tip.label=taxa_names(phyloseq.sub))
    phyloseq.sub <- merge_phyloseq(phyloseq.sub, random_tree)
    bdist <- phyloseq::distance(phyloseq.sub, distance)
    col <- as(sample_data(phyloseq.sub), "data.frame")[, pheno]
    adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
    p.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
    r2.adonis <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
    # res.adonis <- c(p.adonis, r2.adonis)
    
    anosim.dist <- vegan::anosim(unname(bdist), col, permutations=999)
    p.anosim <- signif(anosim.dist$signif, digits=2)
    r2.anosim <- signif(anosim.dist$statistic, digits=2)
    # res.anosim <- c(p.anosim, r2.anosim)
    
    # if(a==1){
    #   p.frame <- data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
    #                         adonis.p=p.adonis, adonis.r2=r2.adonis, 
    #                         anosim.p=p.anosim, anosim.r2=r2.anosim)
    # }else{
    #   p.frame <- rbind(p.frame, data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
    #                                        adonis.p=p.adonis, adonis.r2=r2.adonis, 
    #                                        anosim.p=p.anosim, anosim.r2=r2.anosim))
    # }
    
    if(a==1){
      p.frame <- data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
                            adonis.p=p.adonis, anosim.p=p.anosim)
    }else{
      p.frame <- rbind(p.frame, data.frame(group1=comparisons[[a]][1], group2=comparisons[[a]][2], 
                                           adonis.p=p.adonis, anosim.p=p.anosim))
    }
  }
  return(p.frame)
}
```


# Simple Alpha Diversity

```{r}
alpha <- function(phyloseq, pheno, path=NULL, phyloseq.name=NULL, text=FALSE, title=NA, shannon=TRUE, observed=TRUE){
  comparisons <- as.list(as.data.frame(t(CombPairs(uuln(sample.to.data.frame(phyloseq)[,pheno])))))
  
  richness <- estimate_richness(phyloseq, measures=c("Shannon", "Observed"))
  rownames(richness) <- str_remove(rownames(richness), "X")
  rownames(richness) <- rownames(data.frame(sample_data(phyloseq)))
  richness <- merge(data.frame(sample_data(phyloseq)), richness, by='row.names')
  if(pheno=="type"){
    # richness[, "type"] <- factor(richness[, "type"], levels=c("BNA", "BSA", "BST", "MNA", "MSA", "MBM", "MAS"))
    richness[, "type"] <- factor(richness[, "type"], levels=c("Infant nasal", "Infant saliva", "Infant stool", 
                                                              "Mother nasal", "Mother saliva", "Mother milk", "Mother skin"))
  }
  
  
  for(a in 1:nrow(richness)){
    if(richness[a, "group"]=="baby"){
      richness[a, "group"] <- "Infant"
    }else if(richness[a, "group"]=="mother"){
      richness[a, "group"] <- "Mother"
    }
  }
  
  stat.obs <- as.formula(paste("Observed", "~", pheno))
  stat.test.obs <- richness %>% 
    wilcox_test(stat.obs, comparisons=comparisons) %>%
    add_xy_position(x=pheno) %>%
    # mutate(myformatted.p=paste0("p=", signif(p, 2)))
    mutate(myformatted.p=paste0(signif(p.adj, 2)))
    # length(which(richness[, "birth_type"]=="Vaginal"))
  
  # stat.test.obs <- stat.test.obs[-c(which(stat.test.obs[, "p.adj.signif"]=="ns")), ] ### Remove the non significant comparisions
  
  if(nrow(stat.test.obs)!=0){
    n_fun <- function(x){
      return(data.frame(y=-0.07, label=paste0("n=", length(x))))
    }
    min <- uln(max(richness[, "Observed"]))
    for(b in 1:nrow(stat.test.obs)){
      stat.test.obs[b, "y.position"] <- (min+((b)*(min/12)))
    }
  }
  
  plot.obs <- ggplot(richness, aes(x=get(pheno), y=Observed)) +
    theme_classic() +
    geom_boxplot(outlier.size=-1, fill="transparent") +
    scale_x_discrete(expand=0.6) +
    # geom_text(aes(label=family)) +
    ylab("Observed") + 
    # stat_summary(fun.data=n_fun, geom="text") +
    # facet_wrap("group", scales="free_x") +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.title.x=element_blank(),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #, angle=45, vjust=1, hjust=1
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", legend.position="none",
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #panel.spacing.x=unit(0, "line"), 
          legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, panel.background=element_blank(),
          panel.grid.major=element_line(linetype="dashed", color="gray95"), plot.background=element_blank())
  
  if(text==FALSE){
    plot.obs <- plot.obs + 
      geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.6, dodge.width=0.6), size=2.5, aes(fill="Observed")) 
  }else if(text==TRUE){
    plot.obs <- plot.obs + 
      geom_text(position=position_jitterdodge(seed=1L, jitter.width=0.6, dodge.width=0.6), size=2.5, aes(fill="Observed", label=family)) 
  }
  
  if(nrow(stat.test.obs)!=0){
    plot.obs <- plot.obs + ylim(NA, (max(stat.test.obs[, "y.position"])+(min/14))) + 
      # stat_pvalue_manual(stat.test.obs, size=5, label="p.adj.signif", tip.length=0, bracket.size=0.5) ### Significance as stars
      stat_pvalue_manual(stat.test.obs, size=3, label="myformatted.p", tip.length=0, bracket.size=0.5) ### Significance as numbers 
  }
  
  if(pheno=="type"){
    plot.obs <- plot.obs + 
      scale_x_discrete(labels=c("Infant nasal"="Infant\nnasal", "Infant saliva"="Infant\nsaliva", "Infant stool"="Infant\nstool", 
                                "Mother nasal"="Mother\nnasal", "Mother saliva"="Mother\nsaliva", "Mother milk"="Mother\nmilk", "Mother skin"="Mother\nskin"))
  }
  # print(plot.obs)
    
  
  
  stat.shan <- as.formula(paste("Shannon", "~", pheno))
  stat.test.shan <- richness %>% 
    wilcox_test(stat.shan, comparisons=comparisons) %>%
    add_xy_position(x=pheno) %>%
    # mutate(myformatted.p=paste0("p=", signif(p, 2)))
    mutate(myformatted.p=paste0(signif(p.adj, 2)))
    # length(which(richness[, "birth_type"]=="Vaginal"))
  # stat.test.shan <- stat.test.shan[-c(which(stat.test.shan[, "p.adj.signif"]=="ns")), ] ### Remove the non significant comparisions
  if(nrow(stat.test.shan)!=0){
    n_fun <- function(x){
      return(data.frame(y=-0.07, label=paste0("n=", length(x))))
    }
    min <- uln(max(richness[, "Shannon"]))
    for(b in 1:nrow(stat.test.shan)){
      stat.test.shan[b, "y.position"] <- (min+((b)*(min/12)))
    }
  }
  print(stat.test.shan)
        
  plot.shan <- ggplot(richness, aes(x=get(pheno), y=Shannon)) +
    theme_classic() +
    geom_boxplot(outlier.size=-1, fill="transparent") +
    scale_x_discrete(expand=0.6) +
    ylab("Shannon diversity") + 
    # stat_summary(fun.data=n_fun, geom="text") +
    # facet_wrap("group", scales="free_x") +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.title.x=element_blank(),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #, angle=45, vjust=1, hjust=1
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", legend.position="none",
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #panel.spacing.x=unit(0, "line"), 
          legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, panel.background=element_blank(),
          panel.grid.major=element_line(linetype="dashed", color="gray95"), plot.background=element_blank())
  
  if(text==FALSE){
    plot.shan <- plot.shan + 
      geom_point(position=position_jitterdodge(seed=1L, jitter.width=0.6, dodge.width=0.6), size=2.5, aes(fill="Shannon")) 
  }else if(text==TRUE){
    plot.shan <- plot.shan + 
      geom_text(position=position_jitterdodge(seed=1L, jitter.width=0.6, dodge.width=0.6), size=2.5, aes(fill="Shannon", label=family)) 
  }
  
  if(nrow(stat.test.shan)!=0){
    plot.shan <- plot.shan + ylim(NA, (max(stat.test.shan[, "y.position"])+(min/14))) + 
      # stat_pvalue_manual(stat.test.shan, size=5, label="p.adj.signif", tip.length=0, bracket.size=0.5) ### Significance as stars
      stat_pvalue_manual(stat.test.shan, size=3, label="myformatted.p", tip.length=0, bracket.size=0.5) ### Significance as numbers 
  }
  
  if(pheno=="type"){
    plot.shan <- plot.shan + 
      scale_x_discrete(labels=c("Infant nasal"="Infant\nnasal", "Infant saliva"="Infant\nsaliva", "Infant stool"="Infant\nstool", 
                                "Mother nasal"="Mother\nnasal", "Mother saliva"="Mother\nsaliva", "Mother milk"="Mother\nmilk", "Mother skin"="Mother\nskin"))
  }
  # print(plot.shan)
  
  
  
  plot <- ggarrange(plotlist=list(plot.obs, plot.shan), ncol=2, nrow=1)
  # plot <- annotate_figure(plot, top=text_grob(label=paste0("Infant Alpha diversity"))) +
  #       bgcolor("#e1e1e1") + border(color="#e1e1e1")
  
  if(is.na(title)){
    plot <- plot +
        bgcolor("#e1e1e1") + border(color="#e1e1e1")
  }else if(!is.na(title)){
    plot <- annotate_figure(plot, top=text_grob(label=paste(title, "alpha diversity"))) +
        bgcolor("#e1e1e1") + border(color="#e1e1e1")
  }
  
  
  print(plot)
  
  width <- 1.025 + (0.75*3)
  height <- 0.465 + 0.62 + 0.34 + 2
  
  if(shannon==TRUE & observed==TRUE){
    width <- width*2
  }
  
  if(!is.null(path)){
    filename <- paste(path, "alpha", sep="/")
    if(!is.null(phyloseq.name)){
      filename <- paste(filename, phyloseq.name, sep=".")
    }
    if(text==TRUE){
      filename <- paste(filename, "text", sep=".")
    }
    filename <- paste(filename, "png", sep=".")
  }
  
  
  if(!is.null(path)){
    ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
  }else{
    return(plot)
  }
}
```

# Distance Comparison

## ID Phyloseq Types

```{r}
id.phyloseq.types <- function(phyloseq){
  if(length(phyloseq)==1){
    type <- identify.type(phyloseq)
    return(type)
  }else{
    for(a in 1:length(phyloseq)){
      type <- uuln(sample_data(phyloseq[[a]])[, "type"])
      if(a==1){
        type.merge <- type
      }else{
        type.merge <- c(type.merge, type)
      }
    }
    return(type.merge)
  }
}
```

## Duped Families

```{r}
duped.families <- function(phyloseq){
  for(a in 1:length(phyloseq)){
    sample.data.table <- data.frame(sample_data(phyloseq[[a]])[, c(1, 3, 5)])
    duped.families <- sample.data.table[which(duplicated(sample.data.table[, 1])==TRUE), 1]
    if(a==1){
      duped.families.merged <- duped.families
    }else{
      duped.families.merged <- c(duped.families.merged, duped.families)
      duped.families.merged <- duped.families.merged[which(duplicated(duped.families.merged)==TRUE)]
    }
  }
  return(duped.families.merged)
}
```

## Bray Table

```{r}
bray.table <- function(phyloseq){
  duped.families <- duped.families(phyloseq)
  for(a in 1:length(phyloseq)){
    sample.data.table <- data.frame(sample_data(subset.object(phyloseq[[a]], "Initial"))[, c(1, 3)])
    sample.data.table <- subset.object(sample.data.table, duped.families, "family")
    
    family <- c()
    distance <- c()
    for(b in 1:length(duped.families)){
      bray <- phyloseq::distance(subset.object(phyloseq[[a]], duped.families[b], "family"), "bray")
      family[b] <- as.numeric(gsub("[a-z]*", "", gsub("[A-Z]*", "", gsub("_", "", colnames(bray)))))
      distance[b] <- bray[1]
    }
    
    bray.table <- merge(sample.data.table, data.frame(family, distance), by="family")
    
    if(a==1){
      bray.table.merge <- bray.table
    }else{
      bray.table.merge <- rbind(bray.table.merge, bray.table)
    }
  }
  return(bray.table.merge)
}
```

## Bray Distance Comparison

```{r}
bray.distance.comparison.plot <- function(phyloseq, path){
  type <- id.phyloseq.types(phyloseq)
  duped.families <- duped.families(phyloseq)
  bray.table <- bray.table(phyloseq)
  print(bray.table)
  write_csv(bray.table, paste0(path, "/test.csv"))
  
  if(length(type)==2){
    pwc <- bray.table %>%
        t_test(distance ~ type, paired=TRUE) %>%
        # t_test(distance ~ type) %>%
        adjust_pvalue(method="bonferroni")
    pwc[, "p"] <- signif(pwc[, "p"], 2)
    
    y.value <- max(bray.table[, "distance"])+(max(bray.table[, "distance"])/20)
    pwc[, c("y.position")] <- y.value
    
    plot <- ggplot(bray.table, aes(x=type, y=distance)) +
      theme_classic() +
      geom_point(size=2.5) +
      geom_line(alpha=0.2, aes(group=family)) +
      ylab("Distance between initial and follow-up\nin PCoA plot") + xlab("Sample Type") +
      stat_pvalue_manual(pwc, label="p={p}", tip.length=0) +
      ylim(0, (max(bray.table[, "distance"])+(max(bray.table[, "distance"])/15))) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
              axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
              panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
              strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
              legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
              panel.grid.major=element_line(linetype="dashed", color="gray95"))#, text=element_text(size=15)) 
    ggsave(paste(paste0(path, "/bray.distance.comparison"), paste(type, collapse="."), "png", sep="."), plot, width=4, height=6.5, scale=0.55)
    print(plot)
  }
  if(length(type)>=2){
    if(length(uuln(bray.table[, "type"]))==2){
      colors <- color.list(5)[c(2, 6)]
    }else if(length(uuln(bray.table[, "type"]))==3){
      colors <- color.list(5)[c(2, 5, 8)]
    }else if(length(uuln(bray.table[, "type"]))==4){
      colors <- color.list(5)[c(2, 5, 6, 8)]
    }else{
      colors <- color.list(5)[c(-1)]
    }
    families <- identify.families(bray.table)
    # bray.table[, "family"] <- as.numeric(bray.table[, "family"])
    plot <- ggplot(bray.table, aes(x=family, y=distance)) +
      theme_classic() +
      geom_point(size=2.5, aes(color=type)) +
      geom_line(alpha=0.2, aes(group=type, color=type)) +
      scale_color_manual(values=colors, breaks=type) +
      ylab("Distance between initial and follow-up\nin PCoA plot") + xlab("Family") + 
      scale_x_discrete(limits=families) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            legend.title=element_blank(), legend.box.spacing=unit(0, "pt"), 
            # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))#, text=element_text(size=15)) 
    ggsave(paste(paste0(path, "/bray.distance.comparison.by.family"), paste(type, collapse="."), "png", sep="."), plot, 
           width=3+(0.3*length(duped.families)), height=6.5, scale=0.55)
  }
  # return(pcoa.point.table)
}
```

## PCoA Point Table

```{r}
pcoa.point.table <- function(phyloseq){
  duped.families <- duped.families(phyloseq)
  for(a in 1:length(phyloseq)){
    sample.data.table <- data.frame(sample_data(phyloseq[[a]])[, c(1, 3, 5)])
    sample.data.table <- subset.object(sample.data.table, duped.families, "family")
    ps <- phyloseq::ordinate(phyloseq[[a]], method="PCoA", distance="bray")
    pcoa.point.table <- ps[["vectors"]][, c(1:2)]
    pcoa.point.table <- merge(sample.data.table, pcoa.point.table, by='row.names')
    pcoa.point.table <- pcoa.point.table[, -c(1)]
    
    for(b in 1:length(duped.families)){
      pcoa.point.table.sub <- subset.object(pcoa.point.table, duped.families[b], "family")
      distance <- sqrt(((pcoa.point.table.sub[1, 4]-pcoa.point.table.sub[2, 4])^2) + ((pcoa.point.table.sub[1, 5]-pcoa.point.table.sub[2, 5])^2))
      pcoa.point.table.sub <- pcoa.point.table.sub[1, c(1:2)]
      pcoa.point.table.sub[, "distance"] <- distance
      if(b==1){
        pcoa.point.table.sub.merge <- pcoa.point.table.sub
      }else{
        pcoa.point.table.sub.merge <- rbind(pcoa.point.table.sub.merge, pcoa.point.table.sub)
      }
    }
    if(a==1){
      pcoa.point.table.merge <- pcoa.point.table.sub.merge
    }else{
      pcoa.point.table.merge <- rbind(pcoa.point.table.merge, pcoa.point.table.sub.merge)
    }
  }
  return(pcoa.point.table.merge)
}
```

## PCoA Distance Comparison

```{r}
pcoa.distance.comparison.plot <- function(phyloseq, path){
  type <- id.phyloseq.types(phyloseq)
  duped.families <- duped.families(phyloseq)
  pcoa.point.table <- pcoa.point.table(phyloseq)
  
  if(length(type)==2){
    pwc <- pcoa.point.table %>%
        t_test(distance ~ type) %>%
        adjust_pvalue(method="bonferroni")
    pwc[, "p"] <- signif(pwc[, "p"], 2)
  
    y.value <- max(pcoa.point.table[, "distance"])+(max(pcoa.point.table[, "distance"])/20)
    pwc[, c("y.position")] <- y.value
    
    plot <- ggplot(pcoa.point.table, aes(x=type, y=distance)) +
      theme_classic() +
      geom_point(size=2.5) +
      geom_line(alpha=0.2, aes(group=family)) +
      ylab("Distance between initial and follow-up\nin PCoA plot") + xlab("Sample Type") +
      stat_pvalue_manual(pwc, label="p={p}", tip.length=0) +
      ylim(0, (max(pcoa.point.table[, "distance"])+(max(pcoa.point.table[, "distance"])/15))) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
              axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), 
              panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
              strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
              legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
              panel.grid.major=element_line(linetype="dashed", color="gray95"))#, text=element_text(size=15)) 
    ggsave(paste(paste0(path, "/pcoa.distance.comparison"), paste(type, collapse="."), "png", sep="."), plot, width=4, height=6.5, scale=0.55)
    print(plot)
  }
  if(length(type)>=2){
    if(length(uuln(pcoa.point.table[, "type"]))==2){
      colors <- color.list(5)[c(2, 6)]
    }else if(length(uuln(pcoa.point.table[, "type"]))==3){
      colors <- color.list(5)[c(2, 5, 8)]
    }else if(length(uuln(pcoa.point.table[, "type"]))==4){
      colors <- color.list(5)[c(2, 5, 6, 8)]
    }else{
      colors <- color.list(5)[c(-1)]
    }
    families <- identify.families(pcoa.point.table)
    # pcoa.point.table[, "family"] <- as.numeric(pcoa.point.table[, "family"])
    plot <- ggplot(pcoa.point.table, aes(x=family, y=distance)) +
      theme_classic() +
      geom_point(size=2.5, aes(color=type)) +
      geom_line(alpha=0.2, aes(group=type, color=type)) +
      scale_color_manual(values=colors, breaks=type) +
      ylab("Distance between initial and follow-up\nin PCoA plot") + xlab("Family") + 
      scale_x_discrete(limits=families) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.ticks.y=element_blank(), 
            axis.text.x=element_text(color="black", angle=45, hjust=1), axis.ticks.x=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), plot.title=element_text(hjust=0.5), 
            legend.title=element_blank(), legend.box.spacing=unit(0, "pt"), 
            # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            panel.grid.major=element_line(linetype="dashed", color="gray95"))#, text=element_text(size=15)) 
    ggsave(paste(paste0(path, "/pcoa.distance.comparison.by.family"), paste(type, collapse="."), "png", sep="."), plot, 
           width=3+(0.3*length(duped.families)), height=6.5, scale=0.55)
  }
  # return(pcoa.point.table)
}
```

# Box Plot Taxa

```{r Box plot taxa}
# specific.type=NULL, 
taxa.box.plot <- function(phyloseq, pheno.group, taxa.level, specific.taxa, path, group.title){
  ps.name <- deparse(substitute(phyloseq))
  lower.pheno.group <- str_to_title(tolower(pheno.group))
  
  families <- identify.families(phyloseq)
  
  phyloseq <- phyloseq.processing.compare(phyloseq, pheno.group, NULL, FALSE)
  
  ps.dta <- as.data.frame(tax_table(phyloseq))
  ps.dta <- ps.dta %>% filter_all(any_vars(. %in% c(specific.taxa)))
  rows.of.taxa <- paste0(rownames(ps.dta)[1])
  
  if (taxa.level=="Species"){
    tax_table(phyloseq)[, 7] <- paste(tax_table(phyloseq)[, 6], tax_table(phyloseq)[, 7])
  }
  
  tax_table(phyloseq)[, colnames(tax_table(phyloseq))] <- gsub(tax_table(phyloseq)[, colnames(tax_table(phyloseq))],   
                                                   pattern="[a-z]__", replacement="")

  
  if (taxa.level=="Species"){
    gg.title <- paste(group.title, tax_table(phyloseq)[rows.of.taxa, 7])
  } else {
    gg.title <- paste(taxa.level, specific.taxa)
  }
  
  samdf <- as.data.frame(sample_data(phyloseq))
  otu.top10 <- as.data.frame(otu_table(phyloseq))
  otu.top10[, pheno.group] <- samdf[c(row.names(otu.top10)), pheno.group]
  for (i in 1:nrow(otu.top10)) {
    otu.top10[i, "sample.name"] <- paste0(samdf[c(row.names(otu.top10)[i]), "family"], samdf[c(row.names(otu.top10)[i]), "type"])
  }
  
  otu.top10[, "type"] <- samdf[c(row.names(otu.top10)), "type"]
  otu.top10[, "family"] <- samdf[c(row.names(otu.top10)), "family"]
  otu.top10[, "family"] <- factor(otu.top10[, "family"], levels=families)
  # print(otu.top10)
  
    stat.test <- otu.top10 %>%
    group_by(type) %>%
    t_test(rows.of.taxa ~ grouping) %>%
    adjust_pvalue(method="bonferroni")
    
    stat.test <- stat.test[which(stat.test[, "p"] < 0.05), ]
      
    stat.test <- stat.test %>% add_xy_position(x=pheno)
    stat.test[, "y.position"] <- c(rep(max(otu.top10[, rows.of.taxa]), nrow(stat.test)))
    plot <- ggplot(otu.top10, aes(x=otu.top10[, pheno.group], y=otu.top10[, rows.of.taxa], group=otu.top10[, pheno.group], 
                                  color=otu.top10[, "family"], label=rownames(otu.top10))) +
      geom_boxplot(outlier.size=-1) +
      geom_point(position=position_jitterdodge(seed=1L), alpha=1, size=2)+
      geom_smooth(method="lm", se=FALSE) +
      guides(fill="none") +
      theme_classic(base_size=25) + ylab('Abundance') + 
      theme(text=element_text(face="bold", size=20), 
            axis.line=element_line(color="black", size=2, linetype="solid"), strip.placement="bottom", 
            strip.background=element_rect(fill="white"), axis.text.x=element_text(size=15)) +
      facet_wrap(vars(type)) +
      stat_pvalue_manual(stat.test, label="p={p}")

  specific.taxa <- str_replace(specific.taxa, " ", ".")
  if (ps.name=="NULL"){
    fileName <- paste0(path, "/boxplot.", tolower(pheno.group), ".", specific.taxa, ".png")
  } else {
      fileName <- paste0(path, "/boxplot.", ps.name, ".", tolower(pheno.group), ".", 
                         specific.taxa, ".png")
  }
  ggsave(fileName, plot, type="cairo", width=28, height=14)
}
```

# Change a Phenotype Name

```{r Change phyloseq phenotype name}
change.phyloseq.pheno.name <- function(phyloseq, group.id, column.name, new.name, output.path){
  ps.sam <- as.data.frame(sample_data(phyloseq))
  n.col.change <- which(colnames(ps.sam)==column.name)
  colnames(ps.sam)[n.col.change] <- new.name
  phyloseq <- phyloseq(otu_table(phyloseq, taxa_are_rows=FALSE), 
                       tax_table(phyloseq), 
                       sample_data(ps.sam))
  return(phyloseq)
  if(!is.null(output.path)){
    saveRDS(phyloseq, paste0(output.path, "/Phyloseq.processing/phyloseq.", group.id, ".rds"))
  }
}
```

# Phyloseq Add Phenotype

```{r Phyloseq add phenotype}
phyloseq.add.pheno <- function(phyloseq, group.id, colname, data, output.path=NULL){
  ps.sam <- as.data.frame(sample_data(phyloseq))
  # ps.sam[, colname] <- data
  names(data) <- colname
  ps.sam <- merge(ps.sam, data, by='row.names', all=TRUE)
  rownames(ps.sam) <- ps.sam[, "Row.names"]
  ps.sam <- ps.sam[, -c(1)]
  phyloseq <- phyloseq(otu_table(phyloseq, taxa_are_rows=FALSE), 
                       tax_table(phyloseq), 
                       sample_data(ps.sam))
  
  if(!is.null(output.path)){
    saveRDS(phyloseq, paste0(output.path, "/Phyloseq.processing/phyloseq.", group.id, ".rds"))
  }else{
    return(phyloseq)
  }
}
```

# Remove Specific Taxa by ASV

```{r Remove specific taxa}
remove.specific.taxa <- function(phyloseq, remove.asv){
  allTaxa <- taxa_names(phyloseq)
  removeTaxa <- allTaxa[!(allTaxa %in% remove.asv)]
  phyloseq <- prune_taxa(removeTaxa, phyloseq)
  return(phyloseq)
}
```

# Shorten Taxa Names

```{r Shorten Taxa Names}
# short.taxa.name <- function(group.id, tax.level, taxa.name, file.path){
#   phyloseq <- readRDS(paste0(file.path, "/phyloseq.pruned.", group.id, ".", tolower(tax.level), ".rds"))
#   tax <- data.frame(tax_table(phyloseq))
#   if(tax.level=="Genus"){
#     tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"] <- strsplit(taxa.name[1], "-")[[1]][1]
#   }else if(tax.level=="Species"){
#     tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"] <- 
#       strsplit(tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"], "-")[[1]][1]
#     if(length(unlist(str_split(taxa.name, " ",  n=2)))!=1){
#       taxa <- unlist(str_split(taxa.name, " ",  n=2))
#       tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], tax.level] <- paste(strsplit(taxa[1], "-")[[1]][1], taxa[2])
#     }
#   }
#   tax_table(phyloseq) <- tax_table(as.matrix(tax))
#   saveRDS(phyloseq, paste0(file.path, "/phyloseq.pruned.", group.id, ".", tolower(tax.level), ".rds"))
# }

short.taxa.name <- function(phyloseq, tax.level, taxa.name){
  tax <- data.frame(tax_table(phyloseq))
  if(tax.level=="Genus"){
    if(!is.na(any(tax==taxa.name))){
      tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"] <- strsplit(taxa.name[1], "-")[[1]][1]
    }
  }else if(tax.level=="Species"){
    if(!is.na(any(tax==taxa.name))){
      tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"] <- 
        strsplit(tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], "Genus"], "-")[[1]][1]
      if(length(unlist(str_split(taxa.name, " ",  n=2)))!=1){
        taxa <- unlist(str_split(taxa.name, " ",  n=2))
        tax[rownames(tax)[which(tax[, tax.level]==taxa.name)], tax.level] <- paste(strsplit(taxa[1], "-")[[1]][1], taxa[2])
      }
    }
  }
  tax_table(phyloseq) <- tax_table(as.matrix(tax))
  return(phyloseq)
}

```

# Phyloseq Functions

## Assign Sample Local

```{r}
assign.sample.local <- function(phyloseq){
  for(a in 1:nrow(sample.to.data.frame(phyloseq))){
    if(sample.to.data.frame(phyloseq)[a, "type"]=="BNA" || sample.to.data.frame(phyloseq)[a, "type"]=="MNA"){
      sample_data(phyloseq)[a, "local"] <- "nasal"
    }else if(sample.to.data.frame(phyloseq)[a, "type"]=="BSA" || sample.to.data.frame(phyloseq)[a, "type"]=="MSA"){
      sample_data(phyloseq)[a, "local"] <- "saliva"
    }else if(sample.to.data.frame(phyloseq)[a, "type"]=="BST"){
      sample_data(phyloseq)[a, "local"] <- "stool"
    }else if(sample.to.data.frame(phyloseq)[a, "type"]=="MBM"){
      sample_data(phyloseq)[a, "local"] <- "milk"
    }else if(sample.to.data.frame(phyloseq)[a, "type"]=="MAS"){
      sample_data(phyloseq)[a, "local"] <- "skin"
    }
  }
  return(phyloseq)
}
```

## Remove Unclassified

```{r}
remove.unclassified <- function(phyloseq, taxa.level){
  phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  tax <- tax.to.data.frame(phyloseq)
  phyloseq <- prune_taxa(rownames(tax)[which(str_detect(tax[, taxa.level], "Unclassified")==FALSE)], phyloseq)
  return(phyloseq)
}
```

## Merge Established Phyloseq

```{r}
merge.established.phyloseq <- function(phyloseq.list){
  for(a in 1:length(phyloseq.list)){
    phyloseq <- phyloseq.list[[a]]
    if(length(rownames(data.frame(refseq(phyloseq))))==length(rownames(tax.to.data.frame(phyloseq)))){
      if(length(rownames(data.frame(refseq(phyloseq))))==length(unique(c(rownames(data.frame(refseq(phyloseq))), rownames(tax.to.data.frame(phyloseq)))))){
        taxa_names(phyloseq) <- uln(data.frame(refseq(phyloseq))[, "refseq.phyloseq."])
      }
    }
    if(a==1){
      phyloseq.merge <- phyloseq
    }else{
      phyloseq.merge <- merge_phyloseq(phyloseq.merge, phyloseq)
    }
  }
  dna <- Biostrings::DNAStringSet(taxa_names(phyloseq.merge))
  names(dna) <- taxa_names(phyloseq.merge)
  phyloseq.merge <- merge_phyloseq(phyloseq.merge, dna)
  taxa_names(phyloseq.merge) <- paste0("asv", seq(ntaxa(phyloseq.merge)))
  return(phyloseq.merge)
}
```

## Combine Phyloseq

```{r Combine Phyloseq}
combine.phyloseq <- function(group.id.list, phyloseq.path.list, new.group.id, phyloseq.path, over.min=FALSE){
  phyloseq.list <- list()
  phyloseq.control.list <- list()
  if(length(group.id.list)!=length(phyloseq.path.list)){
    
  }else{
    for(a in 1:length(group.id.list)){
      group.id <- group.id.list[[a]]
      initial.file.path <- paste0(phyloseq.path.list[a], "/Output/Processing/Initial")
      phyloseq.path <- paste0(phyloseq.path.list[a], "/Output/Processing/Decontam")
      print(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
      phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
      phyloseq.control <- readRDS(paste0(phyloseq.path, "/phyloseq.control.", group.id, ".rds"))
      
      sample_data(phyloseq)[, "grouping"] <- group.id
      # sample_data(phyloseq.control)[, "grouping"] <- group.id
      
      if(over.min==TRUE){
        over.min.reads <- readRDS(paste0(initial.file.path, "/over.min.reads.", group.id, ".rds"))
        phyloseq <- prune_samples(over.min.reads, phyloseq)
        for(b in 1:length(sample_names(phyloseq))){
          sample_names(phyloseq)[b] <- paste0(sample_names(phyloseq)[b], unname(unlist(as.data.frame(sample_data(phyloseq))[b, "grouping"])))
        }
      }
      phyloseq.list[[a]] <- phyloseq
      phyloseq.control.list[[a]] <- phyloseq.control
      if(a > 1){
        phyloseq.list[[1]] <- merge_phyloseq(phyloseq.list[[1]], phyloseq.list[[a]])
        phyloseq.control.list[[1]] <- merge_phyloseq(phyloseq.control.list[[1]], phyloseq.control.list[[a]])
        if(over.min==TRUE){
          over.min.reads.merged <- c(over.min.reads.merged, over.min.reads)
        }
      }else{
        if(over.min==TRUE){
          over.min.reads.merged <- over.min.reads
        }
      }
    }
    phyloseq <- phyloseq.list[[1]]
    # phyloseq.control <- phyloseq.control.list[[1]]
    saveRDS(phyloseq, paste0(phyloseq.path, "/Output/Processing/Decontam/phyloseq.", new.group.id, ".rds"))
    # saveRDS(phyloseq.control, paste0(phyloseq.path, "/Output/Processing/Decontam/phyloseq.control.", new.group.id, ".rds"))
    if(over.min==TRUE){
      saveRDS(over.min.reads.merged, paste0(phyloseq.path, "/Output/Processing/Initial/over.min.reads.", new.group.id, ".rds"))
    }
  }
}
```

## Simple Group Phyloseq

```{r}
simple.group.phyloseq <- function(phyloseq.list, col.name, identifiers){
  merge.list <- c()
  for(a in 1:length(phyloseq.list)){
    sample_data(phyloseq.list[[a]])[, col.name] <- identifiers[[a]]
    for(b in 1:length(sample_names(phyloseq.list[[a]]))){
      sample_names(phyloseq.list[[a]])[b] <- paste0(sample_names(phyloseq.list[[a]])[b], "_", 
                                                    unname(unlist(as.data.frame(sample_data(phyloseq.list[[a]]))[b, col.name])))
    }
    # merge.list[[a]] <- phyloseq.list[[a]]
    # if(a!=1){
    #   merge.list[[1]] <- merge_phyloseq(merge.list[[1]], merge.list[[a]])
    # }
  }
  # return(merge.list[[1]])
  return(phyloseq.list)
}
```

## Family Function

```{r}
family.fun <- function(id){
  # all.files <- sort(list.files(paste0("/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-46/", id, "/Sequence"), full.names=TRUE))
  # new <- c()
  # for(b in 1:length(all.files)){
  #   new[b] <- paste(sapply(str_split(all.files[b], "_",  n=3), `[`, 1), sapply(str_split(all.files[b], "_",  n=3), `[`, 2), 
  #                   "I", sapply(str_split(all.files[b], "_",  n=3), `[`, 3), sep="_")
  # }
  # file.rename(all.files, new)
  
  # initial.dada.processing(2, id, 100, get(paste0("path.", id)))
  # phyloseq.creation(id, get(paste0("path.", id)))
  my.function(paste0("/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/", id), id, id, TRUE, FALSE, "NA")
}
```

## Shared Between Samples

```{r}
shared.between.samples <- function(phyloseq.list, col.name, identifiers){
  phyloseq <- simple.group.phyloseq(phyloseq.list, col.name, identifiers)
  phyloseq <- merge_samples(phyloseq, col.name)
  phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  print(phyloseq)
  cat("total combined", sum(sample_sums(phyloseq)), "\n  ")
  phyloseq <- prune_taxa(colSums(otu_table(phyloseq)!=0)>1, phyloseq)
  phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  overlap <- c()
  for(a in 1:ncol(otu_table(phyloseq))){
    overlap[a] <- min(otu_table(phyloseq)[, a])
  }
  # which((colSums(otu_table(phyloseq)!=0)<1)==TRUE)
  print(phyloseq)
  cat("all shared", sum(sample_sums(phyloseq)), "\n  ")
  cat("overlap", sum(overlap), "\n  ")
}
```

## Combine Tax and OTU tables

```{r}
combine.tax.otu <- function(phyloseq, counts=TRUE){
  if(counts==FALSE){
    phyloseq <- trans.sam.counts(phyloseq)
  }
  dataframe <- merge(tax_table(phyloseq), t(otu_table(phyloseq)), by="row.names")
  rownames(dataframe) <- dataframe[, "Row.names"]
  dataframe <- dataframe[, -c(1)]
  return(dataframe)
}
```

```{r}
# combine.phyloseq <- function(group.id.list, phyloseq.path.list, new.group.id, phyloseq.path, over.min=FALSE){
#   phyloseq.list <- list()
#   phyloseq.control.list <- list()
#   if(length(group.id.list)!=length(phyloseq.path.list)){
#     
#   }else{
#     for(a in 1:length(group.id.list)){
#       group.id <- group.id.list[[a]]
#       initial.file.path <- paste0(phyloseq.path.list[a], "/Output/Processing/Initial/")
#       phyloseq.path <- paste0(phyloseq.path.list[a], "/Output/Processing/Decontam/")
#       print(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
#       phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
#       # phyloseq.control <- readRDS(paste0(phyloseq.path, "/phyloseq.control.", group.id, ".rds"))
#       
#       sample_data(phyloseq)[, "grouping"] <- group.id
#       # sample_data(phyloseq.control)[, "grouping"] <- group.id
#       
#       if(over.min==TRUE){
#         over.min.reads <- readRDS(paste0(initial.file.path, "/over.min.reads.", group.id, ".rds"))
#         phyloseq <- prune_samples(over.min.reads, phyloseq)
#         for(b in 1:length(sample_names(phyloseq))){
#           sample_names(phyloseq)[b] <- paste0(sample_names(phyloseq)[b], unname(unlist(as.data.frame(sample_data(phyloseq))[b, "grouping"])))
#         }
#       }
#       phyloseq.list[[a]] <- phyloseq
#       # phyloseq.control.list[[a]] <- phyloseq.control
#       if(a > 1){
#         phyloseq.list[[1]] <- merge_phyloseq(phyloseq.list[[1]], phyloseq.list[[a]])
#         # phyloseq.control.list[[1]] <- merge_phyloseq(phyloseq.control.list[[1]], phyloseq.control.list[[a]])
#         if(over.min==TRUE){
#           over.min.reads.merged <- c(over.min.reads.merged, over.min.reads)
#         }
#       }else{
#         if(over.min==TRUE){
#           over.min.reads.merged <- over.min.reads
#         }
#       }
#     }
#     phyloseq <- phyloseq.list[[1]]
#     # phyloseq.control <- phyloseq.control.list[[1]]
#     saveRDS(phyloseq, paste0(phyloseq.path, "/Output/Processing/Decontam/phyloseq.", new.group.id, ".rds"))
#     # saveRDS(phyloseq.control, paste0(phyloseq.path, "/Output/Processing/Decontam/phyloseq.control.", new.group.id, ".rds"))
#     if(over.min==TRUE){
#       saveRDS(over.min.reads.merged, paste0(phyloseq.path, "/Output/Processing/Initial/over.min.reads.", new.group.id, ".rds"))
#     }
#   }
# }
```

## FEAST Table

```{r}
FEAST.table <- function(phyloseq, path){
  families <- identify.families(phyloseq)

  for(a in 1:length(families)){
    phyloseq.sub <- subset.object(phyloseq, families[a], "family")
    meta <- sample.to.data.frame(phyloseq.sub)
    for(b in 1:nrow(meta)){
      if(meta[b, "group"]=="baby"){
        meta[b, "SourceSink"] <- "Sink"
      }else if(meta[b, "group"]=="mother" || meta[b, "group"]=="negative control"){
        meta[b, "SourceSink"] <- "Source"
      }
    }
  
    colnames(meta)[1] <- "Env"
    meta[, "SampleID"] <- uln(rownames(meta))
    meta <- meta[, c("SampleID", "Env", "SourceSink")]
    rownames(meta) <- c(1:nrow(meta))
    
    for(b in 1:nrow(meta)){
      if(meta[b, "SourceSink"]=="Sink"){
        meta[b, "id"] <- b
      }
    }
    meta[, "SampleID"] <- substring(gsub('[0-9]*', "", sapply(meta[, "SampleID"], function(x) str_split(x, "_")[[1]][1])), 2)
    
    write.table(meta, paste0(path, "/meta.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
    meta <- Load_metadata(paste0(path, "/meta.txt"))
    otu <- as.data.frame(t(as(otu_table(phyloseq.sub), "matrix")))
    colnames(otu) <- substring(gsub('[0-9]*', "", sapply(colnames(otu), function(x) str_split(x, "_")[[1]][1])), 2)
    ncol <- ncol(otu)
    otu[, "#OTU ID"] <- rownames(otu)
    otu <- as(otu[, c(ncol+1, 1:ncol)], "matrix")
    write.table(otu, paste0(path, "/otus.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
    otus <- Load_CountMatrix(CountMatrix_path=paste0(path, "/otus.txt"))
    
    FEAST_output <- FEAST(C=otus, metadata=meta, different_sources_flag=0, dir_path=paste0(path, "/"), outfile="family")
    
    table <- read.table(paste0(path, "/family_source_contributions_matrix.txt"))
    colnames(table) <- sapply(colnames(table), function(x) str_split(x, "_")[[1]][1])
    rownames(table) <- sapply(rownames(table), function(x) str_split(x, "_")[[1]][1])
    
    # if(exists("all.table")==FALSE){
    if(a==1){
      all.table <- data.frame(matrix(ncol=4, nrow=0))
      colnames(all.table) <- c("family", "Sink", "Source", "percentage")
      z <- 1
    }
    
    for(b in 1:ncol(table)){
      for(c in 1:nrow(table)){
        all.table[z, ] <- c(families[a], rownames(table)[c], colnames(table)[b], table[c, b])
        z <- z+1
      }
    }
  }
  
  all.table[, "percentage"] <- as.numeric(all.table[, "percentage"])
  # if(all(uuln(all.table[, "Source"])==c("MAS", "MBM", "MNA", "MSA", "Unknown"))==TRUE){
  #   all.table[, "Source"] <- factor(uln(all.table[, "Source"]), levels=c("Unknown", "MNA", "MSA", "MBM", "MAS"))
  # }else if(all(uuln(all.table[, "Source"])==c("MNA", "MSA", "MBM", "MAS", "Unknown"))==TRUE){
  #   all.table[, "Source"] <- factor(uln(all.table[, "Source"]), levels=c("Unknown", "MNA", "MSA", "MBM", "MAS"))
  # }else if(all(uuln(all.table[, "Source"])==c("MAS", "MBM", "MNA", "MSA", "CNT", "Unknown"))==TRUE){
  #   all.table[, "Source"] <- factor(uln(all.table[, "Source"]), levels=c("Unknown", "CNT", "MNA", "MSA", "MBM", "MAS"))
  # }else if(all(uuln(all.table[, "Source"])==c("CNT", "Unknown"))==TRUE){
  #   all.table[, "Source"] <- factor(uln(all.table[, "Source"]), levels=c("Unknown", "CNT"))
  # }
  all.table[, "family"] <- factor(all.table[, "family"], levels=families)
  return(all.table)
}
```

## FEAST Table Pheno

```{r}
FEAST.table.pheno <- function(phyloseq, table){
  phyloseq.sub <- subset.object(phyloseq, unique(table[, "Sink"]))
  sample.table <- sample.to.data.frame(phyloseq.sub)
  colnames(sample.table)[c(3)] <- "Sink"
  table.merge <- merge(table, sample.table, by=c("family", "Sink"))
  return(table.merge)
}
```

## FEAST Plot Average for Baby Samples

```{r}
package.average.plot <- function(table, id, path=NA, phyloseq.name=NA){
  # colors <- color.list(5)[-c(1)]

  average.all.table <- aggregate(percentage ~ Sink + Source, table[, -c(1)], mean)
  average.all.table <- subset.object(average.all.table, "Unknown", "Source", remove=TRUE)
  sums <- aggregate(percentage ~ Sink, average.all.table, sum)
  for(a in 1:nrow(sums)){
    average.all.table <- rbind(average.all.table, data.frame(Sink=sums[a, "Sink"], Source="Unknown", percentage=(1-sums[a, "percentage"])))
  }
  
  
  ## Added for Dr. Wang
  # colors <- color.list(5)[c(1, 2, 3, 5, 6)]
  colors <- color.list(31)[c(35, 3, 9, 14)]
  # colors <- color.list(21)[c(1, 21, 3, 6, 9)]
  average.all.table[, "Source"] <- as.character(average.all.table[, "Source"])
  print(average.all.table)
  for(a in 1:nrow(average.all.table)){
    if(average.all.table[a, "Sink"]=="BNA"){
      average.all.table[a, "Sink"] <- "Infant nasal"
    }else if(average.all.table[a, "Sink"]=="BSA"){
      average.all.table[a, "Sink"] <- "Infant saliva"
    }else if(average.all.table[a, "Sink"]=="BST"){
      average.all.table[a, "Sink"] <- "Infant stool"
    }
    
    if(average.all.table[a, "Source"]=="MNA"){
      average.all.table[a, "Source"] <- "Mother nasal"
    }else if(average.all.table[a, "Source"]=="MSA"){
      average.all.table[a, "Source"] <- "Mother saliva"
    }else if(average.all.table[a, "Source"]=="MBM"){
      average.all.table[a, "Source"] <- "Mother milk"
    }else if(average.all.table[a, "Source"]=="MAS"){
      average.all.table[a, "Source"] <- "Mother skin"
    }
  }
  print(average.all.table)
  average.all.table[, "Source"] <- factor(uln(average.all.table[, "Source"]), levels=c("Unknown", "Mother nasal", "Mother saliva", "Mother milk", "Mother skin"))
  width <- (2.57 + (length(uuln(table[, "Sink"]))*0.5))
  
  average.all.table
  
  
  
  # plot <- ggplot(average.all.table, aes(x=Sink, y=percentage)) +
  #   theme_classic() +
  #   geom_bar(position="stack", color="black", stat="identity", width=0.8, aes(fill=Source)) +
  #   scale_fill_manual(values=colors) +
  #   coord_cartesian(ylim=c(0, 1), expand=0) + 
  #   ylab("Percent contribution") + xlab("Sink") + #labs(title=title) + #ylab("Percent contribution\nfrom source") + #labs(title=plot.title) +
  #   theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.title.x=element_blank(),
  #         axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #angle=45, hjust=1
  #         # axis.ticks.y.right=element_blank(), axis.title.y.right=element_blank(), axis.text.y.right=element_blank(), 
  #         panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
  #         strip.background=element_rect(fill="white"), #plot.title=element_text(hjust=0.5), #legend.title=element_blank(), 
  #         legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, #legend.position="bottom", legend.box="horizontal", 
  #         # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
  #         panel.grid.major=element_line(linetype="dashed", color="gray95"))#, 
  #         # plot.background=element_blank(), legend.background=element_blank(), panel.background=element_blank())
  # print(plot)
  
  plot <- ggplot(average.all.table, aes(x=Sink, y=percentage)) +
    theme_classic() +
    geom_bar(position="stack", color="black", stat="identity", width=0.8, aes(fill=Source)) +
    scale_fill_manual(values=colors) +
    # coord_cartesian(ylim=c(0, 1), expand=0) +
    scale_y_continuous(limits=c(0,1), expand=c(0.005, 0.005)) +
    # scale_x_discrete(expand=c(0.15, 0.15), labels=c("Infant nasal"="Infant\nnasal", "Infant saliva"="Infant\nsaliva", "Infant stool"="Infant\nstool")) +
    scale_x_discrete(labels=c("Infant nasal"="Infant\nnasal", "Infant saliva"="Infant\nsaliva", "Infant stool"="Infant\nstool"), expand=c(0.14, 0.14)) +
    ylab("Percent contribution") + xlab("Sink") + #labs(title=title) + #ylab("Percent contribution\nfrom source") + #labs(title=plot.title) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.title.x=element_blank(),
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #angle=45, hjust=1
          # axis.ticks.y.right=element_blank(), axis.title.y.right=element_blank(), axis.text.y.right=element_blank(), 
           strip.placement="bottom", #panel.border=element_rect(color="black", fill=NA, size=0.75),
          strip.background=element_rect(fill="white"), #plot.title=element_text(hjust=0.5), #legend.title=element_blank(), 
          legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, #legend.position="bottom", legend.box="horizontal", 
          # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
          # panel.grid.major=element_line(linetype="dashed", color="gray95"))#, 
          plot.background=element_blank(), legend.background=element_blank(), panel.background=element_blank())
  
  
  # width <- (2.25 + (length(uuln(table[, "Sink"]))*0.5))
  # height <- 1.325 + 2.5
  height <- 1.325 + 2.5
  
  
  # if(id=="Type Sequencing Follow up"){
  #   plot <- annotate_figure(plot, top=text_grob(label=paste0(paste0(str_split(id, " ", n=3)[[1]][1], " ", str_split(id, " ", n=3)[[1]][2], "\n", str_split(id, " ", n=3)[[1]][3]), ";\nAverage of all families"))) +
  #     bgcolor("white") + border(color="white")
  # }else{
  #   plot <- annotate_figure(plot, top=text_grob(label=paste0(id, ";\nAverage of all families"))) +
  #     bgcolor("white") + border(color="white")
  # }
  
  # if(id=="Type Sequencing Follow up"){
  #   plot <- annotate_figure(plot, top=text_grob(label=paste0(paste0(str_split(id, " ", n=3)[[1]][1], " ", str_split(id, " ", n=3)[[1]][2], "\n", str_split(id, " ", n=3)[[1]][3]), ";\nAverage of all families"))) +
  #     bgcolor("#e1e1e1") + border(color="#e1e1e1")
  # }else{
  #   plot <- annotate_figure(plot, top=text_grob(label=paste0(id, ";\nAverage of all families"))) +
  #     bgcolor("#e1e1e1") + border(color="#e1e1e1")
  # }
  
  width <- width+1.75 
  if(!is.na(path)){
    if(!is.na(phyloseq.name)){
      filename <- paste(paste0(path, "/", id, "/FEAST.plot"), phyloseq.name, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }else{
      filename <- paste(paste0(path, "/", id, "/FEAST.plot"), id, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }
  }
  print(plot)
}
```

## FEAST Plot Average Phenotype

```{r}
FEAST.average.pheno.plot <- function(table, id, pheno, path=NA, phyloseq.name=NA){
  if(pheno=="birth_type"){
    table[, pheno] <- factor(uln(table[, pheno]), levels=c("Vaginal", "C-Section"))
  }
  
  table.average <- aggregate(percentage ~ Sink + Source + get(pheno), table[, c("family", "Sink", "Source", "percentage", pheno)], mean)
  print(table.average)
  table.average[, "Source"] <- as.character(table.average[, "Source"])
  for(a in 1:nrow(table.average)){
    if(table.average[a, "Sink"]=="BNA"){
      table.average[a, "Sink"] <- "Infant nasal"
    }else if(table.average[a, "Sink"]=="BSA"){
      table.average[a, "Sink"] <- "Infant saliva"
    }else if(table.average[a, "Sink"]=="BST"){
      table.average[a, "Sink"] <- "Infant stool"
    }
    
    if(table.average[a, "Source"]=="MNA"){
      table.average[a, "Source"] <- "Mother nasal"
    }else if(table.average[a, "Source"]=="MSA"){
      table.average[a, "Source"] <- "Mother saliva"
    }else if(table.average[a, "Source"]=="MBM"){
      table.average[a, "Source"] <- "Mother milk"
    }else if(table.average[a, "Source"]=="MAS"){
      table.average[a, "Source"] <- "Mother skin"
    }
  }
  table.average[, "Source"] <- factor(uln(table.average[, "Source"]), levels=c("Unknown", "Mother nasal", "Mother saliva", "Mother milk", "Mother skin"))
  # table.average[, "Source"] <- factor(uln(table.average[, "Source"]), levels=c("Unknown", "MNA", "MSA", "MBM", "MAS"))
  colors <- color.list(5)[c(1, 2, 3, 5, 6)]
  colnames(table.average)[3] <- pheno
  plot <- ggplot(table.average, aes(x=Sink, y=percentage)) +
    theme_classic() +
    geom_bar(position="stack", color="black", stat="identity", width=0.8, aes(fill=Source)) +
    scale_fill_manual(values=colors) +
    coord_cartesian(ylim=c(0, 1), expand=0) + 
    xlab("Sink") + ylab("Percent contribution") + #labs(title=title) + #ylab("Percent contribution\nfrom source") + #labs(title=plot.title) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black"), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #panel.spacing.x=unit(0, "line"), 
          legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, 
          # panel.grid.major=element_line(linetype="dashed", color="gray95"), 
          plot.background=element_blank(), legend.background=element_blank(), panel.background=element_blank())

  # if(pheno=="birth_type"){
  #   plot <- plot + facet_wrap(~factor(get(pheno), levels=c("Vaginal", "C-Section")), nrow=1, ncol=2)
  # }else{
  #   plot <- plot + facet_wrap(get(pheno), nrow=1, ncol=2)
  # }
  
  plot <- plot + facet_wrap(pheno, nrow=1, ncol=2)
  
  print(plot)
  # print((0.325 + (length(uuln(table[, "Sink"]))*0.5)))
  # width <- 7
  width <- (2.25 + ((length(uuln(table[, "Sink"]))*0.5)*length(uuln(table[, pheno]))) + (0.15*(length(uuln(table[, pheno]))-1))) +0.5
  height <- 1.325 + 0.35 + 2.5
  # height <- (0.325 + (length(uuln(table[, pheno]))*0.5))
  
  # plot <- annotate_figure(plot, top=text_grob(label=paste0(id, "; Average of\nfamilies grouped by ", str_replace(pheno, "_", " ")))) +
  #   bgcolor("#e1e1e1") + border(color="#e1e1e1")
  plot <- annotate_figure(plot, top=text_grob(label=paste0("Average of\nfamilies grouped by ", str_replace(pheno, "_", " ")))) +
    bgcolor("#e1e1e1") + border(color="#e1e1e1")
  
  if(!is.na(path)){
    if(!is.na(phyloseq.name)){
      filename <- paste(paste0(path, "/", id, "/FEAST.pheno.plot"), phyloseq.name, pheno, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }else{
      filename <- paste(paste0(path, "/", id, "/FEAST.pheno.plot"), id, pheno, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }
  }
  
  
  
  plot <- ggplot(table.average, aes(x=get(pheno), y=percentage)) +
    theme_classic() +
    geom_bar(position="stack", color="black", stat="identity", width=0.8, aes(fill=Source)) +
    scale_fill_manual(values=colors) +
    coord_cartesian(ylim=c(0, 1), expand=0) + 
    xlab(paste(str_to_title(str_replace(pheno, "_", " ")))) + ylab("Percent contribution") + #labs(title=title) + #ylab("Percent contribution\nfrom source") + #labs(title=plot.title) +
    theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), 
          axis.text.x=element_text(color="black", angle=45, vjust=1, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), 
          panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
          strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #panel.spacing.x=unit(0, "line"), 
          legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, 
          # panel.grid.major=element_line(linetype="dashed", color="gray95"),
          plot.background=element_blank(), legend.background=element_blank(), panel.background=element_blank())

  # if(pheno=="birth_type"){
  #   plot <- plot + facet_wrap(~factor(get(pheno), levels=c("Vaginal", "C-Section")), nrow=1, ncol=2)
  # }else{
  #   plot <- plot + facet_wrap(get(pheno), nrow=1, ncol=2)
  # }
  
  plot <- plot + facet_wrap("Sink", nrow=1, ncol=(length(uuln(table[, "Sink"]))))
  
  print(plot)
  # print((0.325 + (length(uuln(table[, "Sink"]))*0.5)))
  # width <- 7
  width <- (2.25 + ((length(uuln(table[, pheno]))*0.5)*length(uuln(table[, "Sink"]))) + (0.15*(length(uuln(table[, "Sink"]))-1))) +0.5
  height <- 1.325 + 0.35 + 2.5
  # height <- (0.325 + (length(uuln(table[, pheno]))*0.5))
  
  # plot <- annotate_figure(plot, top=text_grob(label=paste0(id, "; Average of\nfamilies grouped by ", str_replace(pheno, "_", " ")))) +
  #   bgcolor("#e1e1e1") + border(color="#e1e1e1")
  plot <- annotate_figure(plot, top=text_grob(label=paste0("Average of\nfamilies grouped by ", str_replace(pheno, "_", " ")))) +
    bgcolor("#e1e1e1") + border(color="#e1e1e1")
  
  if(!is.na(path)){
    if(!is.na(phyloseq.name)){
      filename <- paste(paste0(path, "/", id, "/FEAST.reverse.pheno.plot"), phyloseq.name, pheno, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }else{
      filename <- paste(paste0(path, "/", id, "/FEAST.reverse.pheno.plot"), id, pheno, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }
  }
}
```

## FEAST Family Plot 

```{r}
FEAST.family.plot <- function(table, id, path=NA, phyloseq.name=NA){
  # colors <- color.list(5)[-c(1)]
  
  ## Added for Dr. Wang
  colors <- color.list(5)[c(1, 2, 3, 5, 6)]
  colors <- color.list(31)[c(1, 35, 3, 9, 14)]
  table[, "Source"] <- as.character(table[, "Source"])
  print(table)
  for(a in 1:nrow(table)){
    if(table[a, "Sink"]=="BNA"){
      table[a, "Sink"] <- "Infant nasal"
    }else if(table[a, "Sink"]=="BSA"){
      table[a, "Sink"] <- "Infant saliva"
    }else if(table[a, "Sink"]=="BST"){
      table[a, "Sink"] <- "Infant stool"
    }
    
    if(table[a, "Source"]=="MNA"){
      table[a, "Source"] <- "Mother nasal"
    }else if(table[a, "Source"]=="MSA"){
      table[a, "Source"] <- "Mother saliva"
    }else if(table[a, "Source"]=="MBM"){
      table[a, "Source"] <- "Mother milk"
    }else if(table[a, "Source"]=="MAS"){
      table[a, "Source"] <- "Mother skin"
    }
  }
  print(table)
  table[, "Source"] <- factor(uln(table[, "Source"]), levels=c("Unknown", "Mother nasal", "Mother saliva", "Mother milk", "Mother skin"))
  width <- (2.57 + (length(uuln(table[, "family"]))*0.25))
  # table[, "family"] <- factor(uln(table[, "family"]), levels=order(uln(table[, "family"])))
  
  plot <- ggplot(table, aes(x=family, y=percentage)) +
      theme_classic() +
      geom_bar(position="stack", color="black", stat="identity", width=0.8, aes(fill=Source)) +
      scale_fill_manual(values=colors) +
      coord_cartesian(ylim=c(0, 1), expand=0) + 
      ylab("Percent contribution") + xlab("Sink") + #labs(title=title) + #ylab("Percent contribution\nfrom source") + #labs(title=plot.title) +
      theme(axis.line.x.bottom=element_blank(), axis.line.y.left=element_blank(), axis.title.x=element_blank(),
            axis.text.x=element_text(color="black", angle=45, vjust=1, hjust=1), axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), #angle=45, hjust=1
            # axis.ticks.y.right=element_blank(), axis.title.y.right=element_blank(), axis.text.y.right=element_blank(), 
            panel.border=element_rect(color="black", fill=NA, size=0.75), strip.placement="bottom", 
            strip.background=element_rect(linetype="dashed", fill=NA, size=0.5), #plot.title=element_text(hjust=0.5), #legend.title=element_blank(), 
            legend.box.spacing=unit(1, "pt"), legend.title.align=0.5, #legend.position="bottom", legend.box="horizontal", 
            # legend.key.size=unit(0.25, "cm"), legend.key=element_blank(), legend.position="none", 
            # panel.grid.major=element_line(linetype="dashed", color="gray95"), 
            plot.background=element_blank(), legend.background=element_blank(), panel.background=element_blank())
  plot <- plot + facet_wrap("Sink", nrow=length(uuln(table[, "Sink"])), ncol=1)
  plot <- annotate_figure(plot, top=text_grob(label=paste0(id, ";\nDirect family comparison"))) +
                       bgcolor("white") + border(color="white")
  
  
  # plot <- annotate_figure(plot, top=text_grob(label=paste0("Direct family comparison"))) +
  #                      bgcolor("#white") + border(color="#white")
  
  print(plot)
  # width <- (2.25 + (length(uuln(table[, "family"]))*0.25))
  height <- 1.325 + ((1 + 0.35)*length(uuln(table[, "Sink"]))) + (0.15*(length(uuln(table[, "Sink"]))-1))
  
  if(!is.na(path)){
    if(!is.na(phyloseq.name)){
      filename <- paste(paste0(path, "/", id, "/FEAST.family.plot"), phyloseq.name, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }else{
      filename <- paste(paste0(path, "/", id, "/FEAST.family.plot"), id, "png", sep=".")
      # filename <- paste(filename, , sep=".")
      ggsave(filename=filename, plot, width=width, height=height, scale=0.75)
    }
  }
  return(plot)
}
```

## Phyloseq Summary

```{r}
phyloseq.summary <- function(phyloseq, sequencing.id){
  phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  cat("\n\n", paste(sequencing.id))
  cat("\n   ", paste("All"))
  cat("\n      ", paste("Number of Taxa", ntaxa(phyloseq)))
  cat("\n      ", paste("Reads", sum(taxa_sums(phyloseq))))
  phyloseq <- prune_taxa(rownames(tax_table(phyloseq))[which(str_detect(tax_table(phyloseq)[, "Species"], "Unclassified Phylum")==FALSE)], phyloseq)
  phyloseq <- prune_taxa(taxa_sums(phyloseq)!=0, phyloseq)
  cat("\n   ", paste("No Unidentified Phylum"))
  cat("\n      ", paste("Number of Taxa", ntaxa(phyloseq)))
  cat("\n      ", paste("Reads", sum(taxa_sums(phyloseq))))
}
```

## Phyloseq Summary Dataframe

```{r}
phyloseq.summary.data.frame <- function(){
  tax.frame <- tax.to.data.frame(phyloseq)
  otu.frame <- t(otu.to.data.frame(phyloseq))
  tax.otu.frame <- merge(tax.frame, otu.frame, by="row.names")
  colnames(tax.otu.frame)[1] <- "ASV"
  refseq <- refseq(phyloseq)
  for(a in 1:nrow(tax.otu.frame)){
    tax.otu.frame[a,"Seq"] <- data.frame(refseq)[which(rownames(data.frame(refseq))==tax.otu.frame[a, "ASV"]), "refseq"]
  }
  tax.otu.frame <- tax.otu.frame[, c(ncol(tax.otu.frame), 1:(ncol(tax.otu.frame)-1))]
}
```

## Add seq groups
```{r}
seq.group <- function(phyloseq, family.group){
  # family.group <- c("1-5", "6-10", "11-19", "20-37", "38-46", "47-50")
  # family.group <- c("1-5", "6-10", "11-19", "20-28", "29-37", "38-46", "47-50")
  # family.group <- c("2-19", "38-46", "47-50")
  frame <- sample.to.data.frame(phyloseq)
  for(a in 1:nrow(frame)){
    for(b in 1:length(family.group)){
      # print(as.numeric(str_split(family.group[b], "-")[[1]][1]))
      if(as.numeric(str_split(family.group[b], "-")[[1]][1]) <= as.numeric(frame[a, "family"]) &
         as.numeric(str_split(family.group[b], "-")[[1]][2]) >= as.numeric(frame[a, "family"])){
        # print(paste(frame[a, "family"], family.group[b]))
        sample_data(phyloseq)[a, "seq"] <- family.group[b]
      }
    }
  }
  return(phyloseq)
}
```

## Subset phyloseq by specific taxa

```{r}
subset.phyloseq.by.taxa <- function(phyloseq, taxa.level, specific.taxa){
  # list of taxa at taxa level 
  taxa.list <- tax.to.data.frame(phyloseq)[, taxa.level]
  
  if(length(specific.taxa)==1){
    # frame of specific taxa
    taxa.frame <- tax.to.data.frame(phyloseq)[which(taxa.list==specific.taxa),]
  }else{
    for(a in 1:length(specific.taxa)){
      if(a!=1){
        taxa.frame <- rbind(taxa.frame, tax.to.data.frame(phyloseq)[which(taxa.list==specific.taxa[a]),])
      }else{
        taxa.frame <- tax.to.data.frame(phyloseq)[which(taxa.list==specific.taxa[a]),]
      }
    }
  }
  
  
  # row names of desired taxa
  taxa.asv <- rownames(taxa.frame)
  
  # Subset taxa by specific taxa
  phyloseq <- prune_taxa(taxa.asv, phyloseq)
  
  return(phyloseq)
}
```


# Various Functions

## Unlist Unname

```{r Unlist Unname}
uln <- function(string){
  return(unname(unlist(string)))
}
```

## Unique Unlist Unname

```{r Unlist Unname}
uuln <- function(string){
  return(unique(unname(unlist(string))))
}
```

## Taxa ASV Level

```{r Taxa ASV Level}
if.asv <- function(taxa.level){
  if(taxa.level=="ASV"){
    asv <- TRUE
    taxa.level <- "Species"
  }else{
    asv <- FALSE
  }
  return(c(taxa.level, asv))
}
```

## Add Families to Phyloseq

```{r}
add.families.phyloseq <- function(phyloseq){
  sample.table <- sample.to.data.frame(phyloseq)
  sample.table[, "family"] <- as.numeric(gsub('[a-z]*', "", gsub('[A-Z]*', "", sapply(row.names(sample.table), function(x) strsplit(x, "_")[[1]][1]))))
  sample_data(phyloseq) <- sample.table
  return(phyloseq)
}
```

## Add Families to Phyloseq

```{r}
add.families.dataframe <- function(dataframe){
  dataframe[, "family"] <- as.numeric(gsub('[a-z]*', "", gsub('[A-Z]*', "", sapply(row.names(dataframe), function(x) strsplit(x, "_")[[1]][1]))))
  return(dataframe)
}
```

## Identify Families

```{r Identify Families}
identify.families <- function(object){
  if(class(object)=="phyloseq"){
    return(as.character(sort(as.numeric(as.character(unique(unlist(as.data.frame(sample_data(object))[, "family"])))))))
  }else if(class(object)=="data.frame"){
    return(as.character(sort(as.numeric(as.character(unique(unlist(object[, "family"])))))))
  }
}
```

## Identify Type

```{r}
identify.type <- function(phyloseq){
  return(uuln(sample_data(phyloseq)[, "type"]))
}
```

## Row name

```{r}
row.name <- function(object, cell, column=NA){
  if(class(object)=="phyloseq"){
    samdf <- data.frame(sample_data(object))
    if(is.na(column)){
      for(a in 1:ncol(samdf)){
        if(cell %in% samdf[, a]){
          row.name <- rownames(samdf)[c(which(samdf[, a]==c(cell)))]
        }
      }
    }else{
      row.name <- rownames(samdf)[c(which(samdf[, column]==c(cell)))]
    }
  }else if(class(object)=="data.frame"){
    if(is.na(column)){
      for(a in 1:ncol(object)){
        if(cell %in% object[, a]){
          row.name <- rownames(object)[c(which(object[, a]==c(cell)))]
        }
      }
    }else{
      row.name <- rownames(object)[c(which(object[, column]==c(cell)))]
    }
  }
  return(row.name)
}
```

## Do Adonis

```{r}
doadonis<- function(phyloseq, distance=distance, factor) {
        bdist <- phyloseq::distance(phyloseq, distance)
        col <- as(sample_data(phyloseq), "data.frame")[, factor]
        # Adonis test
        adonis.bdist <- vegan::adonis(unname(bdist) ~ col)
        p <- signif(as.data.frame(adonis.bdist[1])$aov.tab.Pr..F.[1], digits=2)
        r2 <- signif(as.data.frame(adonis.bdist[1])$aov.tab.R2[1], digits=2)
        
              cat("\n      ", paste("p=", p))
        res <- c(p, r2)
        return(res)
      }
```

## Subset Object *************

```{r Subset Object}
subset.object <- function(object, pheno, col.name=NULL, remove=FALSE){
        # cat("\n  ", paste("Subset sample data by", pheno))
  if(class(object)=="phyloseq"){
      if(length(pheno)==1){
      samdf <- data.frame(sample_data(object))
      if(is.null(col.name)){
        if(remove==FALSE){
          for(a in 1:ncol(samdf)){
            if(pheno %in% samdf[, a]){
              sample_data(object) <- sample_data(object)[c(which(sample_data(object)[, a]==c(pheno))), ]
            }
          }
        }else if(remove==TRUE){
          for(a in 1:ncol(samdf)){
            if(pheno %in% samdf[, a]){
              sample_data(object) <- sample_data(object)[c(which(sample_data(object)[, a]!=c(pheno))), ]
            }
          }
        }
      }else{
        if(remove==FALSE){
          sample_data(object) <- sample_data(object)[c(which(sample_data(object)[, col.name]==c(pheno))), ]
        }else if(remove==TRUE){
          sample_data(object) <- sample_data(object)[c(which(sample_data(object)[, col.name]!=c(pheno))), ]
        }
      }
      if(length(which(taxa_sums(object)!=0))!=0){
        object <- prune_taxa(taxa_sums(object)!=0, object)
      }
    }else{
      phyloseq.list <- list()
      object.sub <- object
      for(a in 1:length(pheno)){
        samdf <- data.frame(sample_data(object))
        if(is.null(col.name)){
          if(remove==FALSE){
            object.sub <- object
            for(b in 1:ncol(samdf)){
              if(pheno[a] %in% samdf[, b]){
                sample_data(object.sub) <- sample_data(object)[c(which(sample_data(object)[, b]==c(pheno[a]))), ]
              }
            }
          }else if(remove==TRUE){
            for(a in 1:ncol(samdf)){
              if(pheno[a] %in% samdf[, b]){
                sample_data(object.sub) <- sample_data(object)[c(which(sample_data(object)[, b]!=c(pheno[a]))), ]
              }
            }
          }
        }else{
          if(remove==FALSE){
            if(pheno[a] %in% samdf[, col.name]){
              if(exists("check.for.exist")==FALSE){
                check.for.exist <- a
              }
              object.sub <- object
              sample_data(object.sub) <- sample_data(object)[c(which(sample_data(object)[, col.name]==c(pheno[a]))), ]
            }
          }else if(remove==TRUE){
            sample_data(object.sub) <- sample_data(object)[c(which(sample_data(object)[, col.name]!=c(pheno[a]))), ]
          }
        }
        if(remove==FALSE){
          if(length(which(taxa_sums(object.sub)!=0))!=0){
            phyloseq.list[[a]] <- prune_taxa(taxa_sums(object.sub)!=0, object.sub)
          }else{
            phyloseq.list[[a]] <- object.sub
          }
          if(exists("check.for.exist")==TRUE){
              if(a!=check.for.exist){
              phyloseq.list[[check.for.exist]] <- merge_phyloseq(phyloseq.list[[check.for.exist]], phyloseq.list[[a]])
            }
          }
        }
      }
      if(remove==FALSE){
        if(exists("check.for.exist")==TRUE){
          object <- phyloseq.list[[check.for.exist]]
          if(length(which(taxa_sums(object)!=0))!=0){
            object <- prune_taxa(taxa_sums(object)!=0, object)
          }
        }
      }else if(remove==TRUE){
        object <- object.sub
        if(length(which(taxa_sums(object)!=0))!=0){
          object <- prune_taxa(taxa_sums(object)!=0, object)
        }
      }
    }
  }else if(class(object)=="data.frame"){
    object.og <- object
    if(is.null(col.name)){
      if(remove==FALSE){
        if(length(pheno)>1){
          for(a in 1:length(pheno)){
            for(b in 1:ncol(object)){
              if(pheno %in% object[, b]){
                merge.object <- object.og[c(which(object.og[, b]==c(pheno[a]))), ]
              }
            }
            if(a==1){
              object <- merge.object
            }else{
              object <- rbind(object, merge.object)
            }
          }
        }else{
          for(a in 1:ncol(object)){
            if(pheno %in% object[, a]){
              object <- object[c(which(object[, a]==c(pheno))), ]
            }
          }
        }
      }else if(remove==TRUE){
        for(a in 1:ncol(object)){
          if(pheno %in% object[, a]){
            object <- object[c(which(object[, a]!=c(pheno))), ]
          }
        }
      }
    }else{
      if(remove==FALSE){
        if(length(pheno)>1){
          for(a in 1:length(pheno)){
            merge.object <- object.og[c(which(object.og[, col.name]==c(pheno[a]))), ]
            if(a==1){
              object <- merge.object
            }else{
              object <- rbind(object, merge.object)
            }
          }
        }else{
          object <- object[c(which(object[, col.name]==c(pheno))), ]
        }
      }else if(remove==TRUE){
        if(length(pheno)>1){
          for(a in 1:length(pheno)){
            object <- object[c(which(object[, col.name]!=c(pheno[a]))), ]
          }
        }else{
          object <- object[c(which(object[, col.name]!=c(pheno))), ]
        }
      }
    }
  }
  
  return(object)
}
```

## Replace NAN

```{r Replace NAN}
replace.nan <- function(dataframe, value){
  dataframe[is.nan(dataframe)] <- value
  return(dataframe)
}
```

## Remove NAN

```{r}
remove.nan <- function(object, column=NA){
  if(class(object)=="data.frame"){
    if(!is.nan(column)){
      if(length(which(is.nan(object[, column])==FALSE))!=nrow(object)){
        object <- object[-c(which(is.nan(object[, column])==TRUE)), ]
      }
    }else{
      for(a in 1:length(ncol(object))){
          if(length(which(is.nan(object[, a])==FALSE))!=nrow(object)){
          object <- object[-c(which(is.nan(object[, a])==TRUE)), ]
        }
      }
    }
  }else if(class(object)=="list"){
    for(a in 1:length(list)){
      if(length(which(is.nan(object)==FALSE))!=length(object)){
        object <- object[-c(which(is.nan(object)==TRUE))]
      }
    }
  }
  return(object)
}
```

## Replace NA

```{r Replace NA}
replace.na <- function(dataframe, value){
  dataframe[is.na(dataframe)] <- value
  return(dataframe)
}
```

## Remove NA

```{r}
remove.na <- function(object, column=NA){
  if(class(object)=="data.frame"){
    if(!is.na(column)){
      if(length(which(is.na(object[, column])==FALSE))!=nrow(object)){
        object <- object[-c(which(is.na(object[, column])==TRUE)), ]
      }
    }else{
      for(a in 1:length(ncol(object))){
          if(length(which(is.na(object[, a])==FALSE))!=nrow(object)){
          object <- object[-c(which(is.na(object[, a])==TRUE)), ]
        }
      }
    }
  }else if(class(object)=="list" || class(object)=="character"){
    if(length(which(is.na(object)==FALSE))!=length(object)){
      object <- object[-c(which(is.na(object)==TRUE))]
    }
  }else if(class(object)=="phyloseq"){
    object.frame <- sample.to.data.frame(object)
    non.na <- row.names(object.frame[c(which(is.na(object.frame[, column])==FALSE)), ])
    object <- prune_samples(non.na, object)
  }
  return(object)
}
```


## Remove duplicated

```{r}
remove.duped <- function(list){
  return(list[-c(which(duplicated(list)))])
}
```


## Replace 0

```{r Replace 0}
replace.0 <- function(dataframe, value){
  dataframe[dataframe==0] <- value
  return(dataframe)
}
```

## Remove 0

```{r}
remove.0 <- function(data.frame, column){
  if(length(which(data.frame[, column]==0))!=nrow(data.frame)){
    data.frame <- data.frame[-c(which(data.frame[, column]==0)), ]
  }
  return(data.frame)
}
```

## Remove taxa under 

```{r}
remove.taxa.under <- function(phyloseq, value){
  phyloseq <- prune_taxa(taxa_sums(phyloseq)>value, phyloseq)
  return(phyloseq)
}
```


## Replace Specific

```{r Replace 0}
replace.specific <- function(list, value, replacement){
  list[list==value] <- replacement
  return(list)
}
```

## ASV to Taxa

```{r}
asv.to.taxa <- function(phyloseq, asv, taxa.level){
  if(length(asv)>1){
    taxa <- c()
    for(a in 1:length(asv)){
      taxa[a] <- uln(data.frame(tax_table(phyloseq))[which(rownames(data.frame(tax_table(phyloseq)))==asv[a]), taxa.level])
    }
  }else{
    taxa <- uln(data.frame(tax_table(phyloseq))[which(rownames(data.frame(tax_table(phyloseq)))==asv), taxa.level])
  }
  return(taxa)
}
```

## Taxa to ASV

```{r}
taxa.to.asv <- function(phyloseq, taxa, taxa.level){
  if(length(taxa)>1){
    asv <- c()
    for(a in 1:length(taxa)){
      asv[a] <- rownames(data.frame(tax_table(phyloseq)))[which(data.frame(tax_table(phyloseq))[, taxa.level]==taxa)]
    }
  }else{
    asv <- rownames(data.frame(tax_table(phyloseq)))[which(data.frame(tax_table(phyloseq))[, taxa.level]==taxa)]
  }
  return(asv)
}
```

## Tax Table to Data Frame

```{r}
tax.to.data.frame <- function(phyloseq){
  return(data.frame(tax_table(phyloseq)))
}
```

## OTU Table to Data Frame

```{r}
otu.to.data.frame <- function(phyloseq){
  return(data.frame(otu_table(phyloseq)))
}
```

## Sample Data to Data Frame

```{r}
sample.to.data.frame <- function(phyloseq){
  return(data.frame(sample_data(phyloseq)))
}
```

## Combine Tax and OTU

```{r}
otu.tax <- function(phyloseq){
  tax.table <- tax.to.data.frame(phyloseq)
  otu.table <- otu.to.data.frame(phyloseq)
  combined.table <- merge(tax.table, t(otu.table), by="row.names")
  rownames(combined.table) <- combined.table[,"Row.names"]
  combined.table <- combined.table[,-c(1)]
  return(combined.table)
}
```

## Combine Pheno, Tax, and OTU

```{r}
pheno.otu.tax <- function(phyloseq){
  pheno.table <- sample.to.data.frame(phyloseq)
  pheno.table[,"sample"] <- row.names(pheno.table)
  # print(pheno.table)
  tax.table <- tax.to.data.frame(phyloseq)
  tax.table[,"asv"] <- rownames(tax.table)
  otu.table <- otu.to.data.frame(phyloseq)
  combined.table <- merge(tax.table, t(otu.table), by="row.names")
  rownames(combined.table) <- combined.table[,"Row.names"]
  combined.table <- combined.table[,-c(1)]
  combined.table <- as.data.frame(pivot_longer(combined.table, c(9:ncol(combined.table)), names_to="sample", values_to="value"))
  # row.names(combined.table) <- combined.table[,"sample"]
  # print(pheno.table)
  combined.table <- merge(pheno.table, combined.table, by="sample")
  colnames(combined.table)[1] <- "samples"
  return(combined.table)
}
```

## Form reassignment 

```{r}
form.reassign <- function(data.frame){
  for(a in 1:nrow(data.frame)){
    print(a)
    if(data.frame[a, "form"]=="I"){
      data.frame[a, "form"] <- "Initial"
    }else if(data.frame[a, "form"]=="F1"||data.frame[a, "form"]=="F"){
      data.frame[a, "form"] <- "Follow up 1"
    }else if(data.frame[a, "form"]=="F2"){
      data.frame[a, "form"] <- "Follow up 2"
    }else if(data.frame[a, "form"]=="F3"){
      data.frame[a, "form"] <- "Follow up 3"
    }else if(data.frame[a, "form"]=="F4"){
      data.frame[a, "form"] <- "Follow up 4"
    }
  }
  return(data.frame)
}
```

## Sample type rename

```{r}
sample.type.rename <- function(phyloseq){
  for(a in 1:nrow(sample_data(phyloseq))){
    if(sample_data(phyloseq)[a, "type"]=="BNA"){
      sample_data(phyloseq)[a, "type"] <- "Infant nasal"
    }else if(sample_data(phyloseq)[a, "type"]=="BSA"){
      sample_data(phyloseq)[a, "type"] <- "Infant saliva"
    }else if(sample_data(phyloseq)[a, "type"]=="BST"){
      sample_data(phyloseq)[a, "type"] <- "Infant stool"
    }else if(sample_data(phyloseq)[a, "type"]=="MNA"){
      sample_data(phyloseq)[a, "type"] <- "Mother nasal"
    }else if(sample_data(phyloseq)[a, "type"]=="MSA"){
      sample_data(phyloseq)[a, "type"] <- "Mother saliva"
    }else if(sample_data(phyloseq)[a, "type"]=="MBM"){
      sample_data(phyloseq)[a, "type"] <- "Mother milk"
    }else if(sample_data(phyloseq)[a, "type"]=="MAS"){
      sample_data(phyloseq)[a, "type"] <- "Mother skin"
    }
  }
  return(phyloseq)
}
```

## Round to 5
```{r}
round.to.5 <- function(number){
  return((ceiling(number/5)*5))
}
```


# Sigfill
```{r}
sigfill <- function(values, sigfigs=2, as.p=FALSE){
  values <- uln(values)
  out <- c()
  if(!as.p){
    for(a in 1:length(values)){
      if(values[a]==1){
        out[a] <- paste0("1.", paste0(rep(0, sigfigs), collapse=""))
      }else if((values[a] < as.numeric(paste0("0.", paste0(rep(0, (sigfigs-1)), collapse=""), "1")))){
        out[a] <- paste0("<0.", paste0(rep(0, (sigfigs-1)), collapse=""), "1")
      }else{
        sig.dec <- (which(strsplit(as.character(values[a]), "")[[1]]==".")-1)+sigfigs
        out[a] <- formatC(round(values[a], digits=sigfigs), digits=sig.dec, format="fg", flag="#")
        out[a] <- strtrim(out[a], sig.dec + c(1, 2)[grepl("-", out[a], fixed=TRUE) + 1])
      }
    }
  }else{
    for(a in 1:length(values)){
      if(values[a]==1){
        out[a] <- paste0("1.", paste0(rep(0, sigfigs), collapse=""))
      }else if((values[a] < as.numeric(paste0("0.", paste0(rep(0, (sigfigs-1)), collapse=""), "1")))){
        out[a] <- paste0("p<0.", paste0(rep(0, (sigfigs-1)), collapse=""), "1")
      }else{
        sig.dec <- (which(strsplit(as.character(values[a]), "")[[1]]==".")-1)+sigfigs
        out[a] <- formatC(round(values[a], digits=sigfigs), digits=sig.dec, format="fg", flag="#")
        out[a] <- paste0("p=", strtrim(out[a], sig.dec + c(1, 2)[grepl("-", out[a], fixed=TRUE) + 1]))
      }
    }
  }
  return(out)
  # print(sig.dec)
  # out <- gsub("\\.$", "",
  #             formatC(round(values, digits=sigfigs),
  #                     digits=which(strsplit(as.character(values), "")[[1]]==".")+1, format="fg", flag="#"))
  # out[grepl(".", out, fixed=TRUE)] <- strtrim(out[grepl(".", out, fixed=TRUE)],
  #                                               sig.dec + c(1, 2)[grepl("-", out, fixed=TRUE) + 1])
  # return(out)
}
# sigfill(c(1000.453, 0.009))
```


# Trans Samples in Dataframe

```{r}
sample.type.rename.data.frame <- function(dataframe, column=NA){
  if(is.na(column)){
    for(a in 1:nrow(dataframe)){
      if(dataframe[a, "Sink"]=="BNA"||dataframe[a, "Sink"]=="0BNA"||dataframe[a, "Sink"]=="WBNA"){
        dataframe[a, "Sink"] <- "Infant nasal"
      }else if(dataframe[a, "Sink"]=="BSA"||dataframe[a, "Sink"]=="0BSA"||dataframe[a, "Sink"]=="WBSA"){
        dataframe[a, "Sink"] <- "Infant saliva"
      }else if(dataframe[a, "Sink"]=="BST"||dataframe[a, "Sink"]=="0BST"||dataframe[a, "Sink"]=="WBST"){
        dataframe[a, "Sink"] <- "Infant stool"
      }
      if(dataframe[a, "Source"]=="MNA"||dataframe[a, "Source"]=="0MNA"||dataframe[a, "Source"]=="WMNA"){
        dataframe[a, "Source"] <- "Mother nasal"
      }else if(dataframe[a, "Source"]=="MSA"||dataframe[a, "Source"]=="0MSA"||dataframe[a, "Source"]=="WMSA"){
        dataframe[a, "Source"] <- "Mother saliva"
      }else if(dataframe[a, "Source"]=="MBM"||dataframe[a, "Source"]=="0MBM"||dataframe[a, "Source"]=="WMBM"){
        dataframe[a, "Source"] <- "Mother milk"
      }else if(dataframe[a, "Source"]=="MAS"||dataframe[a, "Source"]=="0MAS"||dataframe[a, "Source"]=="WMAS"){
        dataframe[a, "Source"] <- "Mother skin"
      }
    }
  }else{
    for(a in 1:nrow(dataframe)){
      if(dataframe[a, column]=="BNA"||dataframe[a, column]=="0BNA"||dataframe[a, column]=="WBNA"){
        dataframe[a, column] <- "Infant nasal"
      }else if(dataframe[a, column]=="BSA"||dataframe[a, column]=="0BSA"||dataframe[a, column]=="WBSA"){
        dataframe[a, column] <- "Infant saliva"
      }else if(dataframe[a, column]=="BST"||dataframe[a, column]=="0BST"||dataframe[a, column]=="WBST"){
        dataframe[a, column] <- "Infant stool"
      }
      if(dataframe[a, column]=="MNA"||dataframe[a, column]=="0MNA"||dataframe[a, column]=="WMNA"){
        dataframe[a, column] <- "Mother nasal"
      }else if(dataframe[a, column]=="MSA"||dataframe[a, column]=="0MSA"||dataframe[a, column]=="WMSA"){
        dataframe[a, column] <- "Mother saliva"
      }else if(dataframe[a, column]=="MBM"||dataframe[a, column]=="0MBM"||dataframe[a, column]=="WMBM"){
        dataframe[a, column] <- "Mother milk"
      }else if(dataframe[a, column]=="MAS"||dataframe[a, column]=="0MAS"||dataframe[a, column]=="WMAS"){
        dataframe[a, column] <- "Mother skin"
      }
    }
  }
  
  return(dataframe)
}
```


## Data Summary

```{r}
data.summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean=mean(x[[col]], na.rm=TRUE),
      sd=sd(x[[col]], na.rm=TRUE))
  }
  data_sum <- ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean"=varname))
 return(data_sum)
}
```


## Add number of taxa

```{r}
add.taxa.number <- function(phyloseq){
  sample_data(phyloseq)[, "samID"] <- sample_names(phyloseq)
  sample.frame <- sample.to.data.frame(phyloseq)
  for(a in 1:nrow(sample.to.data.frame(phyloseq))){
    phyloseq.sub <- subset.object(phyloseq, sample.frame[a,"samID"], "samID")
    sample.frame.sub <- sample.to.data.frame(phyloseq.sub)
    sample_data(phyloseq)[rownames(sample.frame)[a], "taxa.number"] <- sum(otu.to.data.frame(phyloseq.sub) != 0)
  }
  sample_data(phyloseq) <- sample_data(phyloseq)[, -c(which(colnames(sample_data(phyloseq))=="samID"))]
  return(phyloseq)
}
```

## Which Col Name

```{r}
which.col.name <- function(dataframe, col.name){
  return(which(colnames(dataframe)==col.name))
}
```

## Round to

```{r}
round.to <- function(value, to, top.bottom=NA){
  if(is.na(top.bottom)){
    return(round(value/to)*to)
  }else if(top.bottom=="top"){
    return(ceiling(value/to)*to)
  }else if(top.bottom=="bottom"){
    return(floor(value/to)*to)
  }
}
```

## Find break limit

```{r}
find.break.limit <- function(value){
  if(value<=500000){
    test <- TRUE
    a <- 1
    while(test){
      if(value<=(500000/(a)) && value>=(200000/(a))){
        test <- FALSE
        return((100000/(a)))
      }else if(value<=(200000/(a)) && value>=(100000/(a))){
        test <- FALSE
        return((50000/(a)))
      }else if(value<=(100000/(a)) && value>(50000/(a))){
        test <- FALSE
        return((25000/(a)))
      }
      a <- (a*10)
    }
  }
}
```

```{r}
### Reassign NA to unclassified ###
na.to.unclassified <- function(phyloseq){
        cat("\n  ", paste("Reassign NA to unclassified"))
  na.table <- as.data.frame(tax_table(phyloseq))
  for(i in 1:nrow(tax_table(phyloseq))){
    for(j in 2:7){
      if(is.na(tax_table(phyloseq)[i, j])){
        k=j
        for(k in k:7){
          tax_table(phyloseq)[i, k] <- paste(tax_table(phyloseq)[i, j-1], "Unclassified", 
                                              colnames(tax_table(phyloseq))[j], rownames(tax_table(phyloseq))[i])
        }
      }
    }
  }
  tax_table(phyloseq) <- tax_table(phyloseq)[!is.na(tax_table(phyloseq)[, 1]), ]
  return(phyloseq)
}
```

## Reword unclassified
```{r}
reword.unclassified <- function(phyloseq){
  phyloseq.tax <- phyloseq
  tax <- tax_table(phyloseq.tax)
  for(a in 1:nrow(tax)){
    if(str_detect(tax[a, "Species"], "Unclassified")==TRUE){
      tax[a, "Species"] <- paste("Unclassified", str_split(tax[a, "Species"], " Unclassified ")[[1]][1])
    }
    if(str_detect(tax[a, "Genus"], "Unclassified")==TRUE){
      if(str_detect(tax[a, "Genus"], "Bacteria")==TRUE){
        tax[a, "Genus"] <- paste("Unclassified", str_split(str_split(tax[a, "Genus"], " Unclassified ")[[1]][2], " ")[[1]][2])
      }else{
        tax[a, "Genus"] <- paste0("Unclassified\n", str_split(tax[a, "Genus"], " Unclassified ")[[1]][1], " ",
                                 str_split(str_split(tax[a, "Genus"], " Unclassified ")[[1]][2], " ")[[1]][2])
      }
    }
    # if(str_detect(tax[a, "Phylum"], "Unclassified")==TRUE){
    #   if(str_detect(tax[a, "Phylum"], "Bacteria")==TRUE){
    #     tax[a, "Phylum"] <- paste("Unclassified", str_split(str_split(tax[a, "Phylum"], " Unclassified ")[[1]][2], " ")[[1]][2])
    #   }else{
    #     tax[a, "Phylum"] <- paste0("Unclassified\n", str_split(tax[a, "Phylum"], " Unclassified ")[[1]][1], " ",
    #                              str_split(str_split(tax[a, "Phylum"], " Unclassified ")[[1]][2], " ")[[1]][2])
    #   }
    # }
  }
  tax_table(phyloseq.tax) <- tax_table(tax)
  return(phyloseq.tax)
}
```


# Bioinformatics WIP

## Prep Adonis

```{r}
prep_bray <- function(phyloseq) {
  
  bray <- phyloseq %>%
    otu_table() %>%
    # t %>%
    vegdist(., method="bray")
  return(bray)
}
run_adonis <- function(bc, meta_adonis, var, perm=1000) {
  
  adonis_res <- adonis2(as.formula(glue("bc ~ {var}")), permutations=perm, data=meta_adonis)
  return(adonis_res)
}
```

## Infant Effect size Niche

```{r Infant Effect size Niche}
WIP.adonis.type <- function(phyloseq, type){
  type.list <- uuln(sample.to.data.frame(phyloseq)[,"type"])[which(uuln(sample.to.data.frame(phyloseq)[,"type"])!=type)]
  for(a in 1:length(type.list)){
    phyloseq.sub <- subset.object(phyloseq, c(type, type.list[a]), "type")
  
    adonis <- run_adonis(prep_bray(phyloseq.sub), sample.to.data.frame(phyloseq.sub), "type")
    if(a==1){
      data.frame <- data.frame(group1=type, group2=type.list[a], Df=adonis[1, "Df"], 
                               SumOfSqs=adonis[1, "SumOfSqs"], R2=adonis[1, "R2"], 
                               F=adonis[1, "F"], `Pr(>F)`=signif(adonis[1, "Pr(>F)"], 2))
    }else{
      data.frame <- rbind(data.frame, data.frame(group1=type, group2=type.list[a], Df=adonis[1, "Df"], 
                                                 SumOfSqs=adonis[1, "SumOfSqs"], R2=adonis[1, "R2"], 
                                                 F=adonis[1, "F"], `Pr(>F)`=signif(adonis[1, "Pr(>F)"], 2)))
    }
  }
  print(data.frame)
  # print(p.adjust(data.frame[,"Pr..F."], method="BH"))
  # data.frame[,"Pr..F."] <- p.adjust(data.frame[,"Pr..F."], method="BH")
  data.frame[,"Pr..F."] <- adjust_pvalue(data.frame, p.col="Pr..F.", method="hochberg")[,"Pr..F..adj"]
  data.frame[, "group2"] <- factor(data.frame[, "group2"], levels=c("Infant nasal", "Infant saliva", "Infant stool", "Mother nasal", "Mother saliva", "Mother milk", "Mother skin"))
  return(data.frame)
}
```

# Color Functions 

## Color List

```{r Color List}
color.list <- function(n, other=FALSE){
  # colors <- c("#900000", "#B34A00", "#C9AB00", "#2A6116", "#001BA5", "#4F02A0", "#6902A0")
  # colors <- c("#900000", "#B34A00", "#C9AB00", "#2A6116", "#001BA5", "#4F02A0", "#8302A0")
  # colors <- c("#900000", "#B34A00", "#C9AB00", "#2A6116", "#001BA5", "#4F02A0", "#9602A0")
  colors <- c("#900000", "#B34A00", "#C9AB00", "#2A6116", "#001BA5", "#4F02A0", "#A00281")
  if(other==TRUE){
    colors <- c("#900000", "#B34A00", "#C9AB00", "#2A6116", "#165061", "#001BA5", "#4F02A0", "#A00281", "#382417")
  }
  
  color.list <- "#FFFFFF00"
  for(a in 1:length(colors)){
    my_col_alpha_all <- character()
    if(n<=5){
      offset=(0+0.5)
      my_col_alpha_all <- adjustcolor(colors[a], offset=c(offset, offset, offset, offset))
    }else if(n==20){
      for(b in 1:3) {
        offset=(0+((b-1)*0.25))
        my_col_alpha_all[b] <- adjustcolor(colors[a], offset=c(offset, offset, offset, offset))
      }
    }else{
      # for(b in 1:(ceiling(n/20)*3)){
      for(b in 1:(ceiling(n/7))){
        offset=(0+((b-1)*(0.25/ceiling(n/25))))
        my_col_alpha_all[b] <- adjustcolor(colors[a], offset=c(offset, offset, offset, offset))
      }
    }
    # show_col(my_col_alpha_all)
    color.list <- c(color.list, my_col_alpha_all)
  }
  # set.seed(872436)
  # x_rand <- sample(color.list)
  # show_col(color.list)
  return(color.list)
}
```

## Color List Gradient

```{r Color List Gradient}
color.list.grad <- function(n, color.list){
  color <- c(color.list)
  color.grad <- c()
  if(length(color.list)==1){
    for(a in 1:(n)){
      offset=(0+((a-1)/((n)+2)))
      color.grad[a] <- adjustcolor(color.list[1], offset=c(offset, offset, offset, offset))
    }
  }else if(length(color.grad)==1){
    for(a in 1:(n/2)){
      offset=(0+((a-1)/((n/2)+2)))
      color.grad[a] <- adjustcolor(color.list[1], offset=c(offset, offset, offset, offset))
    }
    for(b in (n/2):(n)){
      offset=(0+(((b-(n/2))-1)/((n/2)+2)))
      color.grad[(n/2)+(n-b)] <- adjustcolor(color.list[2], offset=c(offset, offset, offset, offset))
    }
  }
  
  # show_col(color.grad)
  return(color.grad)
}

show_col(color.list.grad(8,color.list(20)[2]))
# color.list.grad(46)
```

# Source Tracker

## Source Tracker Files

```{r}
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang $HOME:$HOME $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8081:8787' bsub -M 12GB -R 'select[port8081=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang $HOME:$HOME $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8083:8787' bsub -M 12GB -R 'select[port8083=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang $HOME:$HOME $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8086:8787' bsub -M 12GB -R 'select[port8086=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang $HOME:$HOME $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8089:8787' bsub -M 12GB -R 'select[port8089=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# qiime dev refresh-cache

# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang  $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8081:8787' bsub -M 12GB -R 'select[port8081=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang  $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8083:8787' bsub -M 12GB -R 'select[port8083=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang  $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8086:8787' bsub -M 12GB -R 'select[port8086=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# LSF_DOCKER_VOLUMES="/storage1/fs1/leyao.wang/Active:/storage1/fs1/leyao.wang/Active /scratch1/fs1/leyao.wang:/scratch1/fs1/leyao.wang  $HOME:/var/lib/rstudio-server" PATH="/storage1/fs1/leyao.wang/Active:/opt/conda/envs/qiime2-2020.6/bin:$PATH" LSF_DOCKER_PORTS='8089:8787' bsub -M 12GB -R 'select[port8089=1]' -R "rusage[mem=12GB]" -G compute-leyao.wang -Is -q general-interactive -a 'docker(daniel0325/qiime2:0.0.0.2)' /bin/bash 
# qiime dev refresh-cache

source.tracker.files <- function(phyloseq, path){
  # phyloseq <- readRDS("/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Processing/Phyloseq/phyloseq.pruned.1-50.asv.rds")
  if(dir.exists(paste0(path))==FALSE){
    dir.create(paste0(path))
  }
  
  families <- identify.families(phyloseq)
  
  for(a in 1:length(families)){
    
    if(dir.exists(paste0(path, "/", families[a]))==FALSE){
      dir.create(paste0(path, "/", families[a]))
    }

    phyloseq.sub <- subset.object(phyloseq, families[a], "family")
    rare.sink <- (min(sample_sums(subset.object(phyloseq.sub, c("baby"), "group"))))
    rare.source <- (min(sample_sums(subset.object(phyloseq.sub, c("mother"), "group"))))
    # rare <- (min(sample_sums(subset.object(phyloseq.sub, c("baby", "mother"), "group"))))
    
    # Taxa txt
    tax <- as(tax_table(phyloseq.sub),"matrix")
    tax_cols <- colnames(tax)
    tax <- as.data.frame(tax)
    tax$taxonomy <- do.call(paste, c(tax[tax_cols], sep=";"))
    for(co in tax_cols) tax[co] <- NULL
    colnames(tax) <- "Taxon"
    tax[,"Feature ID"] <- rownames(tax)
    tax <- tax[, c(2, 1)]
    tax[,2] <- sapply(tax[,2], function(x) str_replace_all(x, ";", "_"))
    write.table(tax, paste0(path, "/", families[a], "/tax.txt"), quote=FALSE, col.names=TRUE, row.names=FALSE, sep="\t")
    
    # OTU biom and txt
    otu <- as.data.frame(t(as(otu_table(phyloseq.sub),"matrix")))
    ncol <- ncol(otu)
    otu[,"#OTU ID"] <- rownames(otu)
    otu <- as(otu[,c(ncol+1, 1:ncol)],"matrix")
    write.table(otu,paste0(path, "/", families[a], "/otu.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
    
    otu <- as.data.frame(t(as(otu_table(phyloseq.sub),"matrix")))
    #if taxa_are_rows=TRUE
    otu_biom <- make_biom(data=otu)
    write_biom(otu_biom, paste0(path, "/", families[a], "/otu_biom.biom"))
    
    
    # or from the phyloseq object, 't' to transform if taxa_are_rows=FALSE, no 't' if taxa_are_rows=TRUE
    write.table(t(otu_table(phyloseq.sub)), paste0(path, "/", families[a], "/seqtab.txt"),sep="\t", row.names=TRUE, col.names=NA, quote=FALSE)
    
    # Export metadata (if you have a properly formatted metadata file that you imported in your phyloseq pipeline, you can skip this step and just use that text file directly in QIIME 2)
    print(sample_data(phyloseq.sub))
    write.table(sample_data(phyloseq.sub), paste0(path, "/", families[a], "/sample-metadata.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
    
    ## The Env is samID
    map <- sample_data(phyloseq.sub)
    for(b in 1:nrow(map)){
      if(map[b, "type"]=="Mother nasal"||map[b, "type"]=="Mother saliva"||map[b, "type"]=="Mother milk"||map[b, "type"]=="Mother skin"){
        map[b, "SourceSink"] <- "source"
      }else{
        map[b, "SourceSink"] <- "sink"
      }
    }
    # map <- subset.object(data.frame(map), "POS", "type", TRUE)
    map <- data.frame(map[,c("type", "SourceSink")])
    colnames(map)[1] <- "Env"
    map[,"#Sample ID"] <- rownames(map)
    map[,"Env"] <- map[,"#Sample ID"]
    map <- map[,c(3, 1, 2)]
    write.table(map, paste0(path, "/", families[a], "/map.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
    
    # cat("\n", "\n", "\n", paste("# Source Tracker code for family", families[a]))
    cat("\n", "\n", paste("# Source Tracker code for family", families[a]))
    # cat("\n", "\n", paste("qiime tools import --type", paste0("'", "FeatureData[Taxonomy]", "'"), "\\"))
    cat("\n", paste("qiime tools import --type", paste0("'", "FeatureData[Taxonomy]", "'"), "\\"))
    cat("\n", paste("--input-path", paste0(path, "/", families[a], "/tax.txt"), "\\"))
    cat("\n", paste("--input-format", "HeaderlessTSVTaxonomyFormat", "\\"))
    cat("\n", paste("--output-path", paste0(path, "/", families[a], "/taxonomy.qza")))
    
    # cat("\n", "\n", paste("qiime tools import --type", paste0("'", "FeatureTable[Frequency]", "'"), "\\"))
    # cat("\n", paste("--input-path", paste0(path, "/", form, "/otu_biom.biom"), "\\"))
    # cat("\n", paste("--output-path", paste0(path, "/", form, "/otu.qza")))
    
    # cat("\n", "\n", paste("qiime tools import", "\\"))
    cat("\n", paste("qiime tools import", "\\"))
    cat("\n", paste("--input-path", paste0(path, "/", families[a], "/otu_biom.biom"), "\\"))
    cat("\n", paste("--output-path", paste0(path, "/", families[a], "/otu.qza"), "\\"))
    cat("\n", paste("--type", paste0("'", "FeatureTable[Frequency]", "'"), "\\"))
    cat("\n", paste("--input-format", "BIOMV100Format"))
    
    # cat("\n", "\n", paste("qiime sourcetracker2 gibbs", "\\"))
    cat("\n", paste("qiime sourcetracker2 gibbs", "\\"))
    cat("\n", paste("--i-feature-table", paste0(path, "/", families[a], "/otu.qza"), "\\"))
    cat("\n", paste("--m-sample-metadata-file", paste0(path, "/", families[a], "/map.txt"), "\\"))
    cat("\n", paste("--p-source-sink-column", "SourceSink", "\\"))
    cat("\n", paste("--p-no-loo", "\\"))
    cat("\n", paste("--p-source-column-value", "source", "\\"))
    cat("\n", paste("--p-sink-column-value", "sink", "\\"))
    cat("\n", paste("--p-source-category-column", "Env", "\\"))
    cat("\n", paste("--p-sink-rarefaction-depth", rare.sink, "\\"))
    cat("\n", paste("--p-source-rarefaction-depth", rare.source, "\\"))
    cat("\n", paste("--output-dir", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo")))
    
    # cat("\n", "\n", paste("qiime sourcetracker2 barplot", "\\"))
    cat("\n", paste("qiime sourcetracker2 barplot", "\\"))
    cat("\n", paste("--i-proportions", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/mixing_proportions.qza"), "\\"))
    cat("\n", paste("--m-sample-metadata-file", paste0(path, "/", families[a], "/map.txt"), "\\"))
    cat("\n", paste("--o-visualization", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/proportions-barplot.qzv")))
    
    # cat("\n", "\n", paste("qiime metadata tabulate", "\\"))
    cat("\n", paste("qiime metadata tabulate", "\\"))
    cat("\n", paste("--m-input-file", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments_map.qza"), "\\"))
    cat("\n", paste("--o-visualization", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments_map.qzv")))
    
    # cat("\n", "\n", paste("qiime tools export", "\\"))
    cat("\n", paste("qiime tools export", "\\"))
    cat("\n", paste("--input-path", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments_map.qzv"), "\\"))
    cat("\n", paste("--output-path", paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments_map")))
  }
}
```

## Source Tracker Frame

```{r}
source.tracker.frame <- function(phyloseq, path){
  if(dir.exists(paste0(path))==FALSE){
    dir.create(paste0(path))
  }
  
  families <- identify.families(subset.object(phyloseq, "baby", "group"))
  for(a in 1:length(families)){
    tax <- data.frame(read_qza(paste0(path, "/", families[a], "/taxonomy.qza"))[["data"]][-c(1),])
    rownames(tax) <- tax[,c("Feature.ID")]
    tax[, "Kingdom"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][1])
    tax[, "Phylum"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][2])
    tax[, "Class"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][3])
    tax[, "Order"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][4])
    tax[, "Family"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][5])
    tax[, "Genus"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][6])
    tax[, "Species"] <- sapply(tax[, "Taxon"], function(x) strsplit(x, "_")[[1]][7])
    tax <- as.matrix(tax[, -c(2)])
    
    per_sink_assignments <- t(as.matrix(read_qza(paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments.qza"))[["data"]]))
    
    meta.data <- data.frame(read_tsv(paste0(path, "/", families[a], "/qiime2results/sourcetracker-noloo/per_sink_assignments_map/metadata.tsv")))
    rownames(meta.data) <- meta.data[,c("sampleid")]
    meta.data <- meta.data[-c(1),-c(1)]
    meta.data[,"family"] <- as.character(as.numeric(gsub('[A-Z]*', "", uln(sapply(uln(meta.data[,"Sink"]), function(x)strsplit(x, "_")[[1]][1])))))
    meta.data[,"sink.type"] <- as.character(sapply(gsub('[1-9]*', "", uln(sapply(uln(meta.data[,"Sink"]), function(x)strsplit(x, "_")[[1]][1]))), 
                                                   function(x)str_split(x, "", n=2)[[1]][2]))
    meta.data[,"source.type"] <- as.character(sapply(gsub('[1-9]*', "", uln(sapply(uln(meta.data[,"Source"]), function(x)strsplit(x, "_")[[1]][1]))), 
                                                     function(x)str_split(x, "", n=2)[[1]][2]))
    meta.data[,"source.type"] <- sapply(uln(meta.data[,"source.type"]), function(x)str_replace(x, "nknown", "Unknown"))
    meta.data[,"form"] <- uln(sapply(uln(meta.data[,"Sink"]), function(x)strsplit(x, "_")[[1]][2]))
    
    phyloseq.source.tracker <- phyloseq(otu_table(per_sink_assignments, taxa_are_rows=FALSE),
                                        tax_table(tax),
                                        sample_data(meta.data))
    
    sample.names <- c()
    for(b in 1:nrow(sample.to.data.frame(phyloseq.source.tracker))){
      sample.names[b] <- paste(sample.to.data.frame(phyloseq.source.tracker)[b, "family"], 
                               sample.to.data.frame(phyloseq.source.tracker)[b, "sink.type"], 
                               sample.to.data.frame(phyloseq.source.tracker)[b, "source.type"], sep=".")
    }
    sample_names(phyloseq.source.tracker) <- sample.names
    
    if(exists("ST.phyloseq.merge")==FALSE){
      ST.phyloseq.merge <- phyloseq.source.tracker
    }else{
      ST.phyloseq.merge <- merge_phyloseq(ST.phyloseq.merge, phyloseq.source.tracker)
    }
    
    sink.samples <- uuln(sample.to.data.frame(phyloseq.source.tracker)[,"sink.type"])
    source.samples <- uuln(sample.to.data.frame(phyloseq.source.tracker)[,"source.type"])
    for(b in 1:length(sink.samples)){
      phyloseq.source.tracker.sub <- subset.object(phyloseq.source.tracker, sink.samples[b], "sink.type")
      total.reads <- sum(sample_sums(phyloseq.source.tracker.sub))
      for(c in 1:length(source.samples)){
        # phyloseq.source.tracker.sub.2 <- subset.object(phyloseq.source.tracker.sub, source.samples[c], "source.type")
        percent <- uln(sample_sums(phyloseq.source.tracker.sub)[paste(families[a], sink.samples[b], source.samples[c], sep=".")])/total.reads
        if(exists("ST.frame.function.percent.frame")==FALSE){
          ST.frame.function.percent.frame <- data.frame(family=families[a], Sink=sink.samples[b], Source=source.samples[c], percentage=percent)
        }else{
          ST.frame.function.percent.frame <- rbind(ST.frame.function.percent.frame, data.frame(family=families[a], Sink=sink.samples[b], Source=source.samples[c], percentage=percent))
        }
      }
    }
    
    # assign(paste(families[a], "all.count" sep="."), sample_sums
    
    
    # if(a==1){
    #   phyloseq.source.tracker.merge <- phyloseq.source.tracker
    # }else{
    #   phyloseq.source.tracker.merge <- merge_phyloseq(phyloseq.source.tracker.merge, phyloseq.source.tracker)
    # }
    # if(a==1){
    #   tax.merge <- tax
    # }else{
    #   tax.merge <- rbind(tax.merge, tax)
    # }
  }
  for(a in 1:nrow(ST.frame.function.percent.frame)){
    if(str_detect(ST.frame.function.percent.frame[a, "Sink"], "0")==TRUE){
      ST.frame.function.percent.frame[a, "Sink"] <- str_remove(ST.frame.function.percent.frame[a, "Sink"], "0")
    }
    if(str_detect(ST.frame.function.percent.frame[a, "Source"], "0")==TRUE){
      ST.frame.function.percent.frame[a, "Source"] <- str_remove(ST.frame.function.percent.frame[a, "Source"], "0")
    }
  }
  for(a in 1:nrow(ST.frame.function.percent.frame)){
    if(str_detect(ST.frame.function.percent.frame[a, "Sink"], "W")==TRUE){
      ST.frame.function.percent.frame[a, "Sink"] <- str_remove(ST.frame.function.percent.frame[a, "Sink"], "W")
    }
    if(str_detect(ST.frame.function.percent.frame[a, "Source"], "W")==TRUE){
      ST.frame.function.percent.frame[a, "Source"] <- str_remove(ST.frame.function.percent.frame[a, "Source"], "W")
    }
  }

  for(a in 1:nrow(sample_data(ST.phyloseq.merge))){
    if(str_detect(sample_data(ST.phyloseq.merge)[a, "sink.type"], "0")==TRUE){
      sample_data(ST.phyloseq.merge)[a, "sink.type"] <- str_remove(sample_data(ST.phyloseq.merge)[a, "sink.type"], "0")
    }
    if(str_detect(sample_data(ST.phyloseq.merge)[a, "source.type"], "0")==TRUE){
      sample_data(ST.phyloseq.merge)[a, "source.type"] <- str_remove(sample_data(ST.phyloseq.merge)[a, "source.type"], "0")
    }
  }
  for(a in 1:nrow(sample_data(ST.phyloseq.merge))){
    if(str_detect(sample_data(ST.phyloseq.merge)[a, "sink.type"], "W")==TRUE){
      sample_data(ST.phyloseq.merge)[a, "sink.type"] <- str_remove(sample_data(ST.phyloseq.merge)[a, "sink.type"], "W")
    }
    if(str_detect(sample_data(ST.phyloseq.merge)[a, "source.type"], "W")==TRUE){
      sample_data(ST.phyloseq.merge)[a, "source.type"] <- str_remove(sample_data(ST.phyloseq.merge)[a, "source.type"], "W")
    }
  }
  ST.phyloseq.merge <- transform_sample_counts(ST.phyloseq.merge, function(y) y/10)
  
  sample.names <- c()
  for(b in 1:nrow(sample.to.data.frame(ST.phyloseq.merge))){
    sample.names[b] <- paste(sample.to.data.frame(ST.phyloseq.merge)[b, "family"], 
                             sample.to.data.frame(ST.phyloseq.merge)[b, "sink.type"], 
                             sample.to.data.frame(ST.phyloseq.merge)[b, "source.type"], sep=".")
  }
  sample_names(ST.phyloseq.merge) <- sample.names
  
  tax_table(ST.phyloseq.merge) <- tax_table(ST.phyloseq.merge)[, -c(1)]
  
  return(list(ST.phyloseq.merge, ST.frame.function.percent.frame))
}
```



```{r}
april.percent.shared.calls <- function(dataframe, phyloseq, infant.sample, infant.sample.name, phyloseq.name){
  family.comparison.boxplot(dataframe, phyloseq, "ASV", "type", infant.sample.name, c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Random/Correlation", phyloseq.name)
  family.comparison.taxa.boxplot(dataframe, phyloseq, "ASV", "type", infant.sample.name, c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Random/Correlation", phyloseq.name)
  family.comparison.abundance.taxa(subset.object(dataframe, "Same family", "family.type"), phyloseq, "ASV", "type", infant.sample.name, c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Random/Output/Correlation", phyloseq.name)
  
  # family.list <- uuln(sample.to.data.frame(phyloseq)[which(sample.to.data.frame(phyloseq)[,"birth_type"]=="Vaginal"), "family"])
  # phyloseq.vaginal <- subset.object(phyloseq, family.list, "family")
  # dataframe.sub <- subset.object(dataframe, family.list, "sample.a.fam")
  # dataframe.sub <- subset.object(dataframe.sub, family.list, "sample.b.fam")
  # family.comparison.boxplot(dataframe.sub, phyloseq.vaginal, "ASV", "type", paste(infant.sample.name, "vaginally born"), c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Correlation", "phyloseq.vaginal")
  # family.comparison.taxa.boxplot(dataframe.sub, phyloseq.vaginal, "ASV", "type", paste(infant.sample.name, "vaginally born"), c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Correlation", "phyloseq.vaginal")
  # 
  # family.list <- uuln(sample.to.data.frame(phyloseq)[which(sample.to.data.frame(phyloseq)[,"birth_type"]=="C-Section"), "family"])
  # phyloseq.c.section <- subset.object(phyloseq, family.list, "family")
  # dataframe.sub <- subset.object(dataframe, family.list, "sample.a.fam")
  # dataframe.sub <- subset.object(dataframe.sub, family.list, "sample.b.fam")
  # family.comparison.boxplot(dataframe.sub, phyloseq.c.section, "ASV", "type", paste(infant.sample.name, "c-section"), c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Correlation", "phyloseq.c.section")
  # family.comparison.taxa.boxplot(dataframe.sub, phyloseq.c.section, "ASV", "type", paste(infant.sample.name, "c-section"), c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin"), 0, "/storage1/fs1/leyao.wang/Active/Users/bquinn/2023/1-50/Output/Correlation", "phyloseq.c.section")
}
```


devlib <- "/storage1/fs1/leyao.wang/Active/R_libraries/Camp Pontanezen.Camp Pontanezen"
BiocManager::install("qiime2R", lib=devlib, force=TRUE, dependencies=TRUE)
install.packages("stR", lib=devlib, force=TRUE, dependencies=TRUE)
install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis", lib=devlib)

C99E00
C9AB00
4E0088
8102A0
4502A0
    for(a in 1:length(group.id.list)){
      group.id <- group.id.list[a]
      phyloseq.path <- phyloseq.path.list[a]
      phyloseq <- readRDS(paste0(phyloseq.path, "/phyloseq.", group.id, ".rds"))
      
      if(a==1){
        phyloseq.list <- phyloseq
      }else{
        phyloseq.list <- c(phyloseq.list, phyloseq)
      }
    }

Origional
colors <- c("#EB8581", "#EB0900", "#A30500", "#FFBC8C", "#FF6A00", "#B34A00", "#FFF7C9", "#EDCA00", "#B09600", "#9ED9B6", 
            "#059C41", "#024F21", "#85BEED", "#445FE3", "#2B3C8F", "#C5A2DE", "#8300E0", "#410070", "#D4CDAB", "#A19C82", 
            "#615E4E")
Reversed
colors <- c("#615E4E", "#A19C82", "#D4CDAB", "#410070", "#8300E0", "#C5A2DE", "#2B3C8F", "#445FE3", "#85BEED", "#024F21", 
            "#059C41", "#9ED9B6", "#B09600", "#EDCA00", "#FFF7C9", "#B34A00", "#FF6A00", "#FFBC8C", "#A30500", "#EB0900", 
            "#EB8581")
White Origional
colors <- c("#FFFFFF00", "#EB8581", "#EB0900", "#A30500", "#FFBC8C", "#FF6A00", "#B34A00", "#FFF7C9", "#EDCA00", "#B09600", "#9ED9B6", 
            "#059C41", "#024F21", "#85BEED", "#445FE3", "#2B3C8F", "#C5A2DE", "#8300E0", "#410070", "#D4CDAB", "#A19C82", 
            "#615E4E")
            
            
<!-- richness.alpha <- function(phyloseq, taxa.level, pheno.group, output.path, phyloseq.name){ -->
<!--   cat("\n\n", paste("Richness for", pheno.group)) -->
<!--   sample_data(phyloseq) <- as.data.frame(sample_data(phyloseq))[which(!is.na(as.data.frame(sample_data(phyloseq))[, pheno.group])), ] -->

<!--   comb.list <- unname(unlist(unique(as.data.frame(sample_data(phyloseq))[, pheno.group]))) -->
<!--   comparisons <- as.list(as.data.frame(t(CombPairs(comb.list)))) -->

<!--   if(pheno.group=="birth_type"){ -->
<!--     sample_data(phyloseq)$birth_type=factor(sample_data(phyloseq)$birth_type, levels=c("Vaginal", "C-Section")) -->
<!--   } -->

<!--   colors <- color.list(5)[c(5, 7)] -->

<!--   pheno <- function(pheno){ -->
<!--     if(pheno!="All"){ -->
<!--             cat("\n  ", paste("Subset sample data by", pheno)) -->
<!--       sample_data(phyloseq) <- sample_data(phyloseq)[c(which(sample_data(phyloseq)[, c("type")]==c(pheno))), ] -->
<!--             cat("\n    ", paste("Observed and shannon plot generation")) -->
<!--       richness.plot <- plot_richness(phyloseq, x=pheno.group, measures=c("Shannon", "Observed")) + -->
<!--                          # ggtitle(label=pheno) + -->
<!--                          geom_boxplot(outlier.size=-1) + -->
<!--                          # geom_point(position=position_jitterdodge(seed=1L), aes(fill=family)) + -->
<!--                          guides(fill="none") + -->
<!--                          geom_text(position=position_jitterdodge(seed=1L), aes(fill=family, label=family), ) + -->
<!--                          # color="#5D73E3" -->
<!--                          theme(text=element_text(size=25), axis.title.x=element_blank(),  -->
<!--                                axis.text.x=element_text(angle=45, vjust=1, hjust=1)) + -->
<!--                          stat_compare_means(method="wilcox.test", comparisons=comparisons,  -->
<!--                                             aes(label=paste0("p=", ..p.format..), ), size=5) -->
<!--       richness.plot$layers <- richness.plot$layers[-1] -->
<!--       sam.data <- as.data.frame(sample_data(phyloseq)) -->
<!--             cat("\n    ", paste("Reads boxplot plot generation")) -->
<!--       reads.boxplot <- ggplot(sam.data, aes(x=get(pheno.group), y=reads)) + -->
<!--                          geom_boxplot(outlier.size=-1) +  -->
<!--                          # geom_point(position=position_jitterdodge(seed=1L), aes(fill=family)) + -->
<!--                          guides(fill="none") + -->
<!--                          ylab("Reads") + -->
<!--                          geom_text(position=position_jitterdodge(seed=1L), aes(fill=family, label=family)) + -->
<!--                          theme(text=element_text(size=25), axis.title.x=element_blank(), axis.text.x=element_text(angle=45, vjust=1, hjust=1)) + -->
<!--                          stat_compare_means(method="wilcox.test", comparisons=comparisons,  -->
<!--                                             aes(label=paste0("p=", ..p.format..), ), size=5) -->
<!--       plot <- ggarrange(richness.plot, reads.boxplot, ncol=2, nrow=1) -->
<!--       plot <- annotate_figure(plot, top=text_grob(label=paste(pheno, taxa.level, str_to_title(pheno.group)), color="black", size=25)) + -->
<!--         bgcolor("white") -->

<!--       # assign(paste0(pheno, ".plot"), plot, envir=parent.frame()) -->
<!--       return(plot) -->
<!--     }else{ -->
<!--             cat("\n  ", paste("All sample types")) -->
<!--             cat("\n    ", paste("Observed and shannon plot generation")) -->
<!--       richness.plot <- plot_richness(phyloseq, x=pheno.group, measures=c("Shannon", "Observed")) + -->
<!--                          # ggtitle(label=pheno) + -->
<!--                          geom_boxplot(outlier.size=-1) + -->
<!--                          # geom_point(position=position_jitterdodge(seed=1L), aes(fill=family)) + -->
<!--                          geom_point(position=position_jitterdodge(seed=1L), size=6, aes(color=birth_type)) + -->
<!--                          scale_color_manual(values=colors) + -->
<!--                          # guides(fill="none") + -->
<!--                          # geom_text(position=position_jitterdodge(seed=1L), aes(fill=family, label=family)) + -->
<!--                          # color="#5D73E3" -->
<!--                          # theme(text=element_text(size=25), axis.title.x=element_blank(),  -->
<!--                          #       axis.text.x=element_text(angle=45, vjust=1, hjust=1)) + -->
<!--                          # facet_grid(variable, labeller=c(measures=c("Observed number of species", "Shannon"))) + -->
<!--                          theme(text=element_text(size=35), axis.title.x=element_blank(), axis.text.x=element_blank()) + -->
<!--                          guides(color=guide_legend(title=NULL)) + -->

<!--                          stat_compare_means(method="wilcox.test", comparisons=comparisons,  -->
<!--                                             aes(label=paste0("p=", ..p.format..), ), size=5) -->
<!--       richness.plot$data$variable <- factor(richness.plot$data$variable, levels=c("Observed number of species", "Shannon")) -->
<!--       # print(richness.plot$data$variable) -->
<!--       richness.plot$data$variable[which(richness.plot$data$variable=="Observed")] <- "Observed number of species" -->
<!--       # print(richness.plot$data$variable) -->
<!--       richness.plot$data$variable <- factor(richness.plot$data$variable, levels=c("Observed number of species", "Shannon")) -->
<!--       print(which(richness.plot$data$variable=="Observed number of species")) -->
<!--       # print(richness.plot$data$variable) -->
<!--       # print(richness.plot$layers) -->
<!--       richness.plot$layers <- richness.plot$layers[-1] -->
<!--             cat("\n    ", paste("Reads boxplot plot generation")) -->
<!--       sam.data <- as.data.frame(sample_data(phyloseq)) -->
<!--       reads.boxplot <- ggplot(sam.data, aes(x=get(pheno.group), y=reads)) + -->
<!--                          geom_boxplot(outlier.size=-1) +  -->
<!--                          # geom_point(position=position_jitterdodge(seed=1L), aes(fill=family)) + -->
<!--                          geom_point(position=position_jitterdodge(seed=1L), aes(color=birth_type)) + -->
<!--                          scale_color_manual(values=colors) + -->
<!--                          # guides(fill="none") + -->
<!--                          ylab("Reads") + -->
<!--                          # geom_text(position=position_jitterdodge(seed=1L), aes(fill=family, label=family)) + -->
<!--                          theme(text=element_text(size=25), axis.title.x=element_blank(), axis.text.x=element_text(angle=45, vjust=1, hjust=1)) + -->
<!--                          stat_compare_means(method="wilcox.test", comparisons=comparisons,  -->
<!--                                             aes(label=paste0("p=", ..p.format..), ), size=5) -->
<!--       # plot <- ggarrange(richness.plot, reads.boxplot, ncol=2, nrow=1) -->
<!--       plot <- ggarrange(richness.plot, ncol=1, nrow=1) -->
<!--       if(length(unname(unlist(unique(as.data.frame(sample_data(phyloseq))[, "type"]))))==1){ -->
<!--         # plot <- annotate_figure(plot, top=text_grob(label=paste(unname(unlist(unique(as.data.frame(sample_data(phyloseq))[, "type"]))),  -->
<!--                                                                   # taxa.level, str_to_title(pheno.group)), color="black", size=25)) + -->
<!--                                 # bgcolor("white") -->
<!--       } -->

<!--       # assign(paste0(pheno, ".plot"), plot) -->
<!--     } -->
<!--     if(pheno=="All"){ -->
<!--       if (phyloseq.name=="NULL"){ -->
<!--         fileName <- paste0(output.path, "/alpha.diversity.", pheno, ".", tolower(pheno.group), ".png") -->
<!--       } else { -->
<!--         fileName <- paste0(output.path, "/alpha.diversity.", phyloseq.name, ".", tolower(taxa.level), ".", tolower(pheno.group), ".", pheno, ".png") -->
<!--       } -->
<!--       # ggsave(filename=fileName, plot=plot, width=18, height=6.5) -->
<!--       ggsave(filename=fileName, plot=plot, width=10, height=6.5) -->
<!--     }else{ -->
<!--       return(plot) -->
<!--     } -->
<!--   } -->

<!--   if(pheno.group!="type"){ -->
<!--     if(length(unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"]))))==1){ -->
<!--       pheno("All") -->
<!--     }else{ -->
<!--       phenos <- unique(unname(unlist(as.data.frame(sample_data(phyloseq))[, "type"]))) -->
<!--       pheno("All") -->
<!--       plot.list <- list() -->
<!--       for(a in 1:length(phenos)){ -->
<!--         if(pheno.group=="time"||pheno.group=="birth_type"||pheno.group=="sex"||pheno.group=="breastmilk"){ -->
<!--           if(phenos[a]=="BNA"||phenos[a]=="BSA"||phenos[a]=="BST"){ -->
<!--             plot.list[[a]] <-pheno(phenos[a]) -->
<!--           } -->
<!--         }else{ -->
<!--           plot.list[[a]] <- pheno(phenos[a]) -->
<!--         } -->
<!--       } -->
<!--       if(length(phenos)==2){ -->
<!--         all.plot <- ggarrange(plotlist=plot.list, ncol=1, nrow=(2)) -->
<!--         FileName <- paste0(output.path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".png") -->
<!--         ggsave(filename=FileName, plot=all.plot, width=(18*1), height=(6.5*2)) -->
<!--       }else if(length(phenos)==3){ -->
<!--         all.plot <- ggarrange(plotlist=plot.list, ncol=1, nrow=(length(phenos)/1)) -->
<!--         FileName <- paste0(output.path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".png") -->
<!--         ggsave(filename=FileName, plot=all.plot, width=(18*1), height=(6.5*(length(phenos)/1))) -->
<!--       }else{ -->
<!--         all.plot <- ggarrange(plotlist=plot.list, ncol=2, nrow=(floor(length(phenos)/2))) -->
<!--         FileName <- paste0(output.path, "/alpha.diversity.", phyloseq.name, ".", pheno.group, ".png") -->
<!--         ggsave(filename=FileName, plot=all.plot, width=(18*2), height=(6.5*(length(phenos)/2))) -->
<!--       } -->
<!--     } -->
<!--   } -->
<!-- } -->


<!-- simple beta diversity  -->
<!-- #element_blank()) -->
<!--                   # if(length(uuln(sample_data(phyloseq)[, pheno]))==2){ -->
<!--     #   colors <- color.list(5)[c(2, 6)] -->
<!--     # }else if(length(uuln(sample_data(phyloseq)[, pheno]))==3){ -->
<!--     #   colors <- color.list(5)[c(2, 5, 8)] -->
<!--     # }else if(length(uuln(sample_data(phyloseq)[, pheno]))==4){ -->
<!--     #   colors <- color.list(5)[c(2, 4, 6, 8)] -->
<!--     # }else{ -->
<!--   # if(text==TRUE){ -->
<!--   #   sample_data(phyloseq)[, "raw.reads"] <- as.numeric(data.frame(sample_data(phyloseq))[, "raw.reads"]) -->
<!--   # } -->

<!--   # plot() -->
<!--           # if(phyloseq.name=="family.phyloseq" || phyloseq.name=="family.phyloseq.initial" || phyloseq.name=="family.phyloseq.followup"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Infant nasal", "Infant saliva", "Infant stool",  -->
<!--           #                                                  "Mother nasal", "Mother saliva", "Mother milk", "Mother skin")) -->
<!--           #   height <- 0.3 + (0.3 * 7) -->
<!--           #   width <- 1.75 -->
<!--           # }else if(phyloseq.name=="family.phyloseq.infant" || phyloseq.name=="family.phyloseq.initial.infant" || -->
<!--           #          phyloseq.name=="family.phyloseq.followup.infant"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Infant nasal", "Infant saliva", "Infant stool")) -->
<!--           #   height <- 0.3 + (0.3 * 3) -->
<!--           #   width <- 1.75 -->
<!--           # }else if(phyloseq.name=="family.phyloseq.mother" || phyloseq.name=="family.phyloseq.initial.mother" || -->
<!--           #          phyloseq.name=="family.phyloseq.followup.mother"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin")) -->
<!--           #   height <- 0.3 + (0.3 * 4) -->
<!--           #   width <- 1.75 -->
<!--           # }else if(phyloseq.name=="type.phyloseq"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Infant nasal", "Infant saliva", "Infant stool",  -->
<!--           #                                                  "Mother nasal", "Mother saliva", "Mother milk", "Mother skin")) -->
<!--           #   height <- 0.3 + (0.3 * 7) -->
<!--           #   width <- 1.75 -->
<!--           # }else if(phyloseq.name=="type.phyloseq.infant"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Infant nasal", "Infant saliva", "Infant stool")) -->
<!--           #   height <- 0.3 + (0.3 * 3) -->
<!--           #   width <- 1.75 -->
<!--           # }else if(phyloseq.name=="type.phyloseq.mother"){ -->
<!--           #   plot <- plot + scale_color_discrete(limits=c("Mother nasal", "Mother saliva", "Mother milk", "Mother skin")) -->
<!--           #   height <- 0.3 + (0.3 * 4) -->
<!--           #   width <- 1.75 -->
<!--           # } -->
<!--   # legend <- plot + theme(#legend.box="horizontal",  -->
<!--   #                        legend.title=element_blank(), legend.box.spacing=unit(0, "pt"), legend.spacing=unit(0, "pt"), -->
<!--   #                        legend.text=element_text(margin=margin(r=15)), legend.background=element_rect(color="#e1e1e1"), -->
<!--   #                        legend.box.just="center", plot.background=element_blank()) -->
<!--   # print(legend)  -->